// @lcu/common/esm5/lib/routing/router-helpers.service.js
import { Injectable } from '@angular/core';
import { ActivatedRoute, Router, NavigationCancel, NavigationEnd, NavigationError, NavigationStart } from '@angular/router';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
var RouterHelpersService = /** @class */ (function () {
    // 	Constructors
    function RouterHelpersService(route, router) {
        this.route = route;
        this.router = router;
    }
    // 	API Methods
    RouterHelpersService.prototype.RunOnNavigation = function (action, navEventType) {
        if (navEventType === void 0) { navEventType = null; }
        return this.router.events.subscribe(function (event) {
            if (navEventType == null || event instanceof navEventType) {
                action(event);
            }
        });
    };
    RouterHelpersService.prototype.RunOnNavigationCancel = function (action) {
        return this.RunOnNavigation(action, NavigationCancel);
    };
    RouterHelpersService.prototype.RunOnNavigationEnd = function (action) {
        return this.RunOnNavigation(action, NavigationEnd);
    };
    RouterHelpersService.prototype.RunOnNavigationError = function (action) {
        return this.RunOnNavigation(action, NavigationError);
    };
    RouterHelpersService.prototype.RunOnNavigationStart = function (action) {
        return this.RunOnNavigation(action, NavigationStart);
    };
    RouterHelpersService.prototype.RunOnQueryParam = function (param, action, checkVal) {
        if (checkVal === void 0) { checkVal = null; }
        return this.route.queryParams.subscribe(function (params) {
            if ((checkVal == null && params[param] != null) || (checkVal != null && params[param] === checkVal)) {
                action(params);
            }
        });
    };
    RouterHelpersService.prototype.RunOnRouteParam = function (param, action, checkVal) {
        if (checkVal === void 0) { checkVal = null; }
        return this.route.params.subscribe(function (params) {
            if ((checkVal == null && params[param] != null) || (checkVal != null && params[param] === checkVal)) {
                action(params);
            }
        });
    };
    RouterHelpersService.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: Router }
    ]; };
    RouterHelpersService.ɵprov = i0.ɵɵdefineInjectable({ factory: function RouterHelpersService_Factory() { return new RouterHelpersService(i0.ɵɵinject(i1.ActivatedRoute), i0.ɵɵinject(i1.Router)); }, token: RouterHelpersService, providedIn: "root" });
    RouterHelpersService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    RouterHelpersService.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: Router }
    ]; };
    return RouterHelpersService;
}());
export { RouterHelpersService };
