// @lcu/common/esm5/lib/registry/enterprise/application.js
var Application = /** @class */ (function () {
    function Application() {
    }
    return Application;
}());
export { Application };
var DAFApplication = /** @class */ (function () {
    function DAFApplication() {
    }
    return DAFApplication;
}());
export { DAFApplication };
var DAFAPIApplicationDetails = /** @class */ (function () {
    function DAFAPIApplicationDetails() {
    }
    return DAFAPIApplicationDetails;
}());
export { DAFAPIApplicationDetails };
var DAFViewApplicationDetails = /** @class */ (function () {
    function DAFViewApplicationDetails() {
    }
    return DAFViewApplicationDetails;
}());
export { DAFViewApplicationDetails };
var DAFRedirectApplicationDetails = /** @class */ (function () {
    function DAFRedirectApplicationDetails() {
    }
    return DAFRedirectApplicationDetails;
}());
export { DAFRedirectApplicationDetails };
var DAFAppPointerApplicationDetails = /** @class */ (function () {
    function DAFAppPointerApplicationDetails() {
    }
    return DAFAppPointerApplicationDetails;
}());
export { DAFAppPointerApplicationDetails };
