// @lcu/common/esm2015/lib/utils/conversion/temperature.conversion.js
// @dynamic
/**
 * @dynamic is used because this class contains static properties
 */
export class TemperatureConversion {
    /**
     * Convert fahrenheit to celsius
     * @param val fahrenheit value
     */
    static FahrenheitToCelsius(val) {
        return (val - 32) / 1.8;
    }
    /**
     * Convert fahrenheit to kelvin
     * @param val fahrenheit value
     */
    static FahrenheitToKelvin(val) {
        return ((val - 32) / 1.8) + 273.15;
    }
    /**
     * Convert celsius to fahrenheit
     * @param val celsius value
     */
    static CelsiusToFahrenheit(val) {
        return (val * 1.8) + 32;
    }
    /**
     * Convert celsius to kelvin
     * @param val celsius value
     */
    static CelsiusToKelvin(val) {
        return val + 273.15;
    }
    /**
     * Convert kelvin to fahrenheit
     * @param val kelvin value
     */
    static KelvinToFahrenheit(val) {
        return ((val - 273.15) * 1.8) + 32;
    }
    /**
     * Convert kelvin to celsius
     * @param val kelvin value
     */
    static KelvinToCelsius(val) {
        return val - 273.15;
    }
}
