// @lcu/common/esm5/lib/guid.js
var Guid = /** @class */ (function () {
    // 	Constructors
    function Guid(guid) {
        if (!guid) {
            throw new TypeError('Invalid argument; `value` has no value.');
        }
        this.value = Guid.Empty;
        if (guid && Guid.IsGuid(guid)) {
            this.value = guid;
        }
    }
    // 	API Methods
    Guid.Create = function () {
        return Guid.Parse(Guid.CreateRaw());
    };
    Guid.CreateEmpty = function () {
        return Guid.Parse('emptyguid');
    };
    Guid.CreateRaw = function () {
        return [Guid.gen(2), Guid.gen(1), Guid.gen(1), Guid.gen(1), Guid.gen(3)].join('-');
    };
    Guid.IsGuid = function (guid) {
        var value = guid.toString();
        return guid && (guid instanceof Guid || Guid.GuidValidator.test(value));
    };
    Guid.Parse = function (guid) {
        return new Guid(guid);
    };
    // 	Helpers
    Guid.gen = function (count) {
        var out = '';
        for (var i = 0; i < count; i++) {
            // tslint:disable-next-line:no-bitwise
            out += (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return out;
    };
    Guid.prototype.Equals = function (other) {
        return Guid.IsGuid(other) && this.value === other.toString();
    };
    Guid.prototype.IsEmpty = function () {
        return this.value === Guid.Empty;
    };
    Guid.prototype.ToJSON = function () {
        return {
            Value: this.value
        };
    };
    Guid.prototype.ToString = function () {
        return this.value;
    };
    // 	Properties
    Guid.GuidValidator = new RegExp('^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$', 'i');
    Guid.Empty = '00000000-0000-0000-0000-000000000000';
    return Guid;
}());
export { Guid };
