// @lcu/common/esm5/lib/loading.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var Loading = /** @class */ (function () {
    // 	Constructors
    function Loading() {
        this.Clear();
    }
    Object.defineProperty(Loading.prototype, "Value", {
        // 	Properties
        get: function () {
            return this.loaders && this.loaders.length > 0 && this.loaders.every(function (loader) { return loader; });
        },
        enumerable: true,
        configurable: true
    });
    // 	API Methods
    Loading.prototype.Clear = function () {
        this.loaders = [];
    };
    Loading.prototype.Set = function (loading) {
        if (loading) {
            this.loaders.push(loading);
        }
        else if (this.loaders && this.loaders.length > 0) {
            this.loaders.splice(0, 1);
        }
    };
    Loading.ɵprov = i0.ɵɵdefineInjectable({ factory: function Loading_Factory() { return new Loading(); }, token: Loading, providedIn: "root" });
    Loading = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __metadata("design:paramtypes", [])
    ], Loading);
    return Loading;
}());
export { Loading };
