// @lowcodeunit/lcu-documentation-common/esm5/lib/state/docs-manager.context.js
import { __decorate, __extends } from "tslib";
import { Injectable, Injector } from '@angular/core';
import { StateContext } from '@lcu/common';
import * as i0 from "@angular/core";
var DocsManagerContext = /** @class */ (function (_super) {
    __extends(DocsManagerContext, _super);
    function DocsManagerContext(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        return _this;
    }
    DocsManagerContext.prototype.GetDocsById = function (id) {
        this.State.Loading = true;
        this.Execute({
            Arguments: {
                DocsId: id
            },
            Type: 'get-docs-by-id'
        });
    };
    DocsManagerContext.prototype.loadStateKey = function () {
        return 'main';
    };
    DocsManagerContext.prototype.loadStateName = function () {
        return 'docs';
    };
    DocsManagerContext.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    DocsManagerContext.ɵprov = i0.ɵɵdefineInjectable({ factory: function DocsManagerContext_Factory() { return new DocsManagerContext(i0.ɵɵinject(i0.INJECTOR)); }, token: DocsManagerContext, providedIn: "root" });
    DocsManagerContext = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], DocsManagerContext);
    return DocsManagerContext;
}(StateContext));
export { DocsManagerContext };
