// @lcu/cli/dist/src/cli/commands/starter.command.service.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var BaseCommandService_1 = require("./BaseCommandService");
var _3rdparty_async_1 = require("../../helpers/3rdparty-async");
var chalk_1 = __importDefault(require("chalk"));
var StarterCommandService = /** @class */ (function (_super) {
    __extends(StarterCommandService, _super);
    function StarterCommandService() {
        return _super.call(this) || this;
    }
    StarterCommandService.prototype.Setup = function (program) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, program
                        .command('starter')
                        .alias('str')
                        .description('Initialize, generate, and establish an entire LCU project structure.')
                        .option('-s|--scope <scope>', 'The @scope to use for package names.')
                        .option('-w|--workspace <workspace>', 'The name of the workspace to use.')
                        .option('-sn|--slnName <slnName>', 'The name of the project solution.')
                        .option('-d|--docs <docs>', 'Whether or not to add docs to the base of the project.')
                        .action(function (options) { return __awaiter(_this, void 0, void 0, function () {
                        var context, slnContext, err_1;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.isLcuInitialized()];
                                case 1:
                                    if (!_a.sent()) return [3 /*break*/, 2];
                                    this.establishSectionHeader('LCU Already Initialized', 'yellow');
                                    return [3 /*break*/, 8];
                                case 2:
                                    context = {
                                        projectsPath: 'projects',
                                        repo: 'lowcodeunit-devkit/lcu-cli-templates-core',
                                        scope: options.scope,
                                        workspace: options.workspace,
                                        slnName: options.slnName,
                                        docs: options.docs || false,
                                        projectName: 'lcu',
                                        template: 'LCU',
                                        initWith: null
                                    };
                                    _a.label = 3;
                                case 3:
                                    _a.trys.push([3, 7, , 8]);
                                    // Initialize
                                    return [4 /*yield*/, this.initialize(context)];
                                case 4:
                                    // Initialize
                                    _a.sent();
                                    // Project
                                    return [4 /*yield*/, this.setupProject(context)];
                                case 5:
                                    // Project
                                    _a.sent();
                                    slnContext = {
                                        name: this.slnName,
                                        disableLcuBootstrap: false,
                                        export: 'src/lcu.api.ts',
                                        module: 'app.module.ts',
                                        path: 'lib/elements',
                                        projectName: 'common',
                                        template: 'Core'
                                    };
                                    return [4 /*yield*/, this.setupSolution(slnContext)];
                                case 6:
                                    _a.sent();
                                    this.Ora.succeed('Completed initialization and creation of the LCU');
                                    return [3 /*break*/, 8];
                                case 7:
                                    err_1 = _a.sent();
                                    this.Ora.fail("Issue establishing LCU: \r\n" + err_1);
                                    process.exit(1);
                                    return [3 /*break*/, 8];
                                case 8: return [2 /*return*/];
                            }
                        });
                    }); })];
            });
        });
    };
    StarterCommandService.prototype.initialize = function (context) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, repoTempPath, answers;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        this.establishSectionHeader('LCU Initializing');
                        _a = context;
                        return [4 /*yield*/, this.ensureInquired(context.scope, 'scope')];
                    case 1:
                        _a.scope = _d.sent();
                        _b = context;
                        return [4 /*yield*/, this.ensureInquired(context.workspace, 'workspace')];
                    case 2:
                        _b.workspace = _d.sent();
                        _c = context;
                        return [4 /*yield*/, this.ensureInquired(context.slnName, 'slnName', 'What is the solution name?')];
                    case 3:
                        _c.slnName = _d.sent();
                        this.slnName = context.slnName;
                        repoTempPath = this.pathJoin(this.tempFiles, 'repos', context.repo);
                        return [4 /*yield*/, this.establishTemplatesRepo(repoTempPath, context.repo)];
                    case 4:
                        _d.sent();
                        answers = this.inquir(this.pathJoin(repoTempPath, 'initialize'));
                        context = Object.assign(context, answers);
                        return [4 /*yield*/, this.processTemplates(context, 'initialize')];
                    case 5:
                        _d.sent();
                        this.Ora.succeed('Completed initialization step');
                        return [2 /*return*/];
                }
            });
        });
    };
    StarterCommandService.prototype.setupProject = function (context) {
        return __awaiter(this, void 0, void 0, function () {
            var lcuConfig, templateRepoPath, projAnswers;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.establishSectionHeader('LCU Project Setup');
                        return [4 /*yield*/, this.loadLCUConfig()];
                    case 1:
                        lcuConfig = _a.sent();
                        templateRepoPath = this.pathJoin(this.tempFiles, 'repos', lcuConfig.templates.repository, 'project');
                        projAnswers = { "template": "LCU" };
                        return [4 /*yield*/, this.mergeObjects(context, projAnswers)];
                    case 2:
                        context = _a.sent();
                        return [4 /*yield*/, this.processTemplateCommands(this.pathJoin(templateRepoPath, context.template), context)];
                    case 3:
                        _a.sent();
                        this.Ora.succeed("Completed setup for project " + context.projectName + ".");
                        return [2 /*return*/];
                }
            });
        });
    };
    StarterCommandService.prototype.setupSolution = function (context) {
        return __awaiter(this, void 0, void 0, function () {
            var lcuConfig, templateRepoPath, slnAnswers;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.establishSectionHeader('LCU Solution Setup');
                        return [4 /*yield*/, this.loadLCUConfig()];
                    case 1:
                        lcuConfig = _a.sent();
                        templateRepoPath = this.pathJoin(this.tempFiles, 'repos', lcuConfig.templates.repository, 'solution');
                        slnAnswers = { "template": "Core" };
                        return [4 /*yield*/, this.mergeObjects(context, slnAnswers)];
                    case 2:
                        context = _a.sent();
                        return [4 /*yield*/, this.processTemplateCommands(this.pathJoin(templateRepoPath, context.template), context)];
                    case 3:
                        _a.sent();
                        this.Ora.succeed("Completed setup for solution " + context.projectName + ".");
                        return [2 /*return*/];
                }
            });
        });
    };
    //  Helpers
    StarterCommandService.prototype.ensureName = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.ensureInquired(name, 'name')];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    StarterCommandService.prototype.establishTemplatesRepo = function (repoTempPath, repo) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
                        var ora;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    ora = this.Ora.start("Loading Templates Repository '" + repo + "'");
                                    return [4 /*yield*/, _3rdparty_async_1.AsyncHelpers.rimraf(repoTempPath).catch(function (err) {
                                            ora.fail("Issue cleaning temp path @ '" + chalk_1.default.yellow(repoTempPath) + "': " + chalk_1.default.red(err));
                                            process.exit(1);
                                            reject();
                                        })];
                                case 1:
                                    _a.sent();
                                    return [4 /*yield*/, _3rdparty_async_1.AsyncHelpers.downloadGit(repo, repoTempPath).catch(function (err) {
                                            ora.fail("Template loading failed with: \n\t" + chalk_1.default.red(err));
                                            process.exit(1);
                                            reject();
                                        })];
                                case 2:
                                    _a.sent();
                                    ora.succeed("Loaded Templates to '" + repoTempPath + "'");
                                    resolve();
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
            });
        });
    };
    StarterCommandService.prototype.processTemplates = function (context, subPath) {
        return __awaiter(this, void 0, void 0, function () {
            var templatesRepoPath, source;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        templatesRepoPath = this.pathJoin(this.tempFiles, 'repos', context.repo);
                        source = this.pathJoin(templatesRepoPath, subPath);
                        return [4 /*yield*/, this.processTemplateCommands(source, context)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return StarterCommandService;
}(BaseCommandService_1.BaseCommandService));
exports.StarterCommandService = StarterCommandService;
