// @lowcodeunit/lcu-documentation-common/esm2015/lib/lcu-documentation.module.js
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { FathymSharedModule, MaterialModule } from '@lcu/common';
import { DocsService } from './services/docs.service';
import { LcuDocsComponent, markedOptionsFactory } from './controls/docs/docs.component';
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { MarkdownModule, MarkedOptions } from 'ngx-markdown';
const ɵ0 = markedOptionsFactory;
export class LcuDocumentationModule {
    static forRoot() {
        return {
            ngModule: LcuDocumentationModule,
            providers: [DocsService]
        };
    }
}
LcuDocumentationModule.decorators = [
    { type: NgModule, args: [{
                declarations: [LcuDocsComponent],
                imports: [
                    FathymSharedModule,
                    FormsModule,
                    ReactiveFormsModule,
                    FlexLayoutModule,
                    MaterialModule,
                    HttpClientModule,
                    MarkdownModule.forRoot({
                        loader: HttpClient,
                        markedOptions: {
                            provide: MarkedOptions,
                            useFactory: ɵ0
                        }
                    })
                ],
                exports: [LcuDocsComponent],
                entryComponents: []
            },] }
];
export { ɵ0 };
