// @lcu/common/esm2015/lib/directives/themes/theme-color.directive.js
import { Directive, Input, ElementRef, Renderer2 } from '@angular/core';
/**
 * Directive for setting theme color to a element, color comes from
 * a local stylesheet
 *
 * @use as an attribute: lcuColor="primary-A700", sets the class
 * .primary-A700 {} to the host element
 */
/**
 * SCSS class types
 */
// type ColorClasses = 'primary-500' | 'accent-500' | 'warn-500';
export class ThemeColorDirective {
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
    }
    /**
     * Set the current theme class to the element
     */
    set lcuThemeColor(val) {
        // this.renderer.removeClass(this.element.nativeElement, `${val}`);
        this.renderer.addClass(this.element.nativeElement, `${val}`);
        // this.readProperty('background-color');
    }
    /**
     * Get the value of a specific style property
     *
     * @param property style property
     */
    readProperty(property) {
        console.log('element', window.getComputedStyle(this.element.nativeElement).getPropertyValue(property));
        // console.log('element', window.getComputedStyle(this.element.nativeElement));
    }
}
ThemeColorDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
ThemeColorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[lcuThemeColor]'
            },] }
];
ThemeColorDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
ThemeColorDirective.propDecorators = {
    lcuThemeColor: [{ type: Input }]
};
