// @lcu/common/esm5/lib/services/theming/theme-color-picker.utils.js
// https://ajonp.com/courses/angularmaterial/angular-material-theming/
import { OverlayContainer } from '@angular/cdk/overlay';
import { BehaviorSubject } from 'rxjs/internal/BehaviorSubject';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
var ThemeColorPickerService = /** @class */ (function () {
    function ThemeColorPickerService(overlayContainer) {
        this.overlayContainer = overlayContainer;
        this.initialClass = 'fathym-contrast-theme';
        this.colorClass = new BehaviorSubject(this.initialClass);
        var storageClass = localStorage.getItem('theme-picker');
        console.log('color picker service', storageClass);
        if (storageClass) {
            overlayContainer.getContainerElement().classList.add(storageClass);
            this.colorClass.next(storageClass);
        }
        else {
            overlayContainer.getContainerElement().classList.add(this.initialClass);
        }
    }
    ThemeColorPickerService.prototype.GetColorClass = function () {
        return this.colorClass;
    };
    ThemeColorPickerService.prototype.SetColorClass = function (className) {
        var _this = this;
        this.overlayContainer.getContainerElement().classList.forEach(function (css) {
            if (css !== 'cdk-overlay-container') {
                _this.overlayContainer.getContainerElement().classList.remove(css);
            }
        });
        this.overlayContainer.getContainerElement().classList.add(className);
        this.colorClass.next(className);
        localStorage.setItem('theme-picker', className);
    };
    ThemeColorPickerService.ctorParameters = function () { return [
        { type: OverlayContainer }
    ]; };
    ThemeColorPickerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThemeColorPickerService_Factory() { return new ThemeColorPickerService(i0.ɵɵinject(i1.OverlayContainer)); }, token: ThemeColorPickerService, providedIn: "root" });
    ThemeColorPickerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    ThemeColorPickerService.ctorParameters = function () { return [
        { type: OverlayContainer }
    ]; };
    return ThemeColorPickerService;
}());
export { ThemeColorPickerService };
