// @lowcodeunit/lcu-documentation-common/esm2015/lib/state/docs-manager.context.js
import { __decorate } from "tslib";
import { Injectable, Injector } from '@angular/core';
import { StateContext } from '@lcu/common';
import * as i0 from "@angular/core";
let DocsManagerContext = class DocsManagerContext extends StateContext {
    constructor(injector) {
        super(injector);
        this.injector = injector;
    }
    GetDocsById(id) {
        this.State.Loading = true;
        this.Execute({
            Arguments: {
                DocsId: id
            },
            Type: 'get-docs-by-id'
        });
    }
    loadStateKey() {
        return 'main';
    }
    loadStateName() {
        return 'docs';
    }
};
DocsManagerContext.ctorParameters = () => [
    { type: Injector }
];
DocsManagerContext.ɵprov = i0.ɵɵdefineInjectable({ factory: function DocsManagerContext_Factory() { return new DocsManagerContext(i0.ɵɵinject(i0.INJECTOR)); }, token: DocsManagerContext, providedIn: "root" });
DocsManagerContext = __decorate([
    Injectable({
        providedIn: 'root'
    })
], DocsManagerContext);
export { DocsManagerContext };
