// @lcu/common/esm2015/lib/guid.js
export class Guid {
    // 	Constructors
    constructor(guid) {
        if (!guid) {
            throw new TypeError('Invalid argument; `value` has no value.');
        }
        this.value = Guid.Empty;
        if (guid && Guid.IsGuid(guid)) {
            this.value = guid;
        }
    }
    // 	API Methods
    static Create() {
        return Guid.Parse(Guid.CreateRaw());
    }
    static CreateEmpty() {
        return Guid.Parse('emptyguid');
    }
    static CreateRaw() {
        return [Guid.gen(2), Guid.gen(1), Guid.gen(1), Guid.gen(1), Guid.gen(3)].join('-');
    }
    static IsGuid(guid) {
        const value = guid.toString();
        return guid && (guid instanceof Guid || Guid.GuidValidator.test(value));
    }
    static Parse(guid) {
        return new Guid(guid);
    }
    // 	Helpers
    static gen(count) {
        let out = '';
        for (let i = 0; i < count; i++) {
            // tslint:disable-next-line:no-bitwise
            out += (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return out;
    }
    Equals(other) {
        return Guid.IsGuid(other) && this.value === other.toString();
    }
    IsEmpty() {
        return this.value === Guid.Empty;
    }
    ToJSON() {
        return {
            Value: this.value
        };
    }
    ToString() {
        return this.value;
    }
}
// 	Properties
Guid.GuidValidator = new RegExp('^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$', 'i');
Guid.Empty = '00000000-0000-0000-0000-000000000000';
