// @lcu/common/esm2015/lib/services/theming/theme-color-picker.utils.js
import { __decorate, __metadata } from "tslib";
// https://ajonp.com/courses/angularmaterial/angular-material-theming/
import { OverlayContainer } from '@angular/cdk/overlay';
import { BehaviorSubject } from 'rxjs/internal/BehaviorSubject';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
let ThemeColorPickerService = class ThemeColorPickerService {
    constructor(overlayContainer) {
        this.overlayContainer = overlayContainer;
        this.initialClass = 'fathym-contrast-theme';
        this.colorClass = new BehaviorSubject(this.initialClass);
        const storageClass = localStorage.getItem('theme-picker');
        console.log('color picker service', storageClass);
        if (storageClass) {
            overlayContainer.getContainerElement().classList.add(storageClass);
            this.colorClass.next(storageClass);
        }
        else {
            overlayContainer.getContainerElement().classList.add(this.initialClass);
        }
    }
    GetColorClass() {
        return this.colorClass;
    }
    SetColorClass(className) {
        this.overlayContainer.getContainerElement().classList.forEach(css => {
            if (css !== 'cdk-overlay-container') {
                this.overlayContainer.getContainerElement().classList.remove(css);
            }
        });
        this.overlayContainer.getContainerElement().classList.add(className);
        this.colorClass.next(className);
        localStorage.setItem('theme-picker', className);
    }
};
ThemeColorPickerService.ctorParameters = () => [
    { type: OverlayContainer }
];
ThemeColorPickerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThemeColorPickerService_Factory() { return new ThemeColorPickerService(i0.ɵɵinject(i1.OverlayContainer)); }, token: ThemeColorPickerService, providedIn: "root" });
ThemeColorPickerService = __decorate([
    Injectable({
        providedIn: 'root'
    }),
    __metadata("design:paramtypes", [OverlayContainer])
], ThemeColorPickerService);
export { ThemeColorPickerService };
