// @lcu/common/fesm2015/lcu-common.js
import { throwError, BehaviorSubject, Observable, Subject, ReplaySubject } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { HttpClient, HttpHeaders, HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { __awaiter } from 'tslib';
import { HttpTransportType, HubConnectionBuilder, LogLevel, HubConnectionState } from '@aspnet/signalr';
import { EventEmitter, NgZone, Injector, ɵɵdefineInjectable, ɵɵinject, INJECTOR, Injectable, Input, Pipe, NgModule, ElementRef, Renderer2, Directive, InjectionToken, Inject } from '@angular/core';
import { CommonModule, DatePipe, DecimalPipe, PercentPipe } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { RouterModule, NavigationCancel, NavigationEnd, NavigationError, NavigationStart, ActivatedRoute, Router } from '@angular/router';
import { DomSanitizer } from '@angular/platform-browser';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { CdkTableModule } from '@angular/cdk/table';
import { MatTreeModule } from '@angular/material/tree';
import { MatBottomSheet } from '@angular/material/bottom-sheet';
import { MatBadgeModule } from '@angular/material/badge';
import { MatDividerModule } from '@angular/material/divider';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { MatButtonModule } from '@angular/material/button';
import { MatInputModule } from '@angular/material/input';
import { MatAutocompleteModule } from '@angular/material/autocomplete';
import { MatDatepickerModule } from '@angular/material/datepicker';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatRadioModule } from '@angular/material/radio';
import { MatSelectModule } from '@angular/material/select';
import { MatSliderModule } from '@angular/material/slider';
import { MatSlideToggleModule } from '@angular/material/slide-toggle';
import { MatMenuModule } from '@angular/material/menu';
import { MatSidenavModule } from '@angular/material/sidenav';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatListModule } from '@angular/material/list';
import { MatGridListModule } from '@angular/material/grid-list';
import { MatCardModule } from '@angular/material/card';
import { MatStepperModule } from '@angular/material/stepper';
import { MatTabsModule } from '@angular/material/tabs';
import { MatExpansionModule } from '@angular/material/expansion';
import { MatButtonToggleModule } from '@angular/material/button-toggle';
import { MatChipsModule } from '@angular/material/chips';
import { MatIconModule } from '@angular/material/icon';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatProgressBarModule } from '@angular/material/progress-bar';
import { MatDialogModule } from '@angular/material/dialog';
import { MatTooltipModule } from '@angular/material/tooltip';
import { MatSnackBarModule } from '@angular/material/snack-bar';
import { MatTableModule, MatTableDataSource } from '@angular/material/table';
import { MatSortModule, MatSort } from '@angular/material/sort';
import { MatPaginatorModule, MatPaginator } from '@angular/material/paginator';
import { OverlayContainer } from '@angular/cdk/overlay';
import { BehaviorSubject as BehaviorSubject$1 } from 'rxjs/internal/BehaviorSubject';

class Address {
}

class Amount {
}

class LCUServiceSettings {
}
class LCUApplicationConfig {
}
class LCUStateConfig {
}

class DAFService {
    //  Constructors
    constructor(injector) {
        this.injector = injector;
        this.http = injector.get(HttpClient);
        try {
            this.settings = injector.get(LCUServiceSettings);
        }
        catch (err) { }
    }
    //  Helpers
    catchError() {
        return catchError(this.handleError);
    }
    delete(path) {
        return this.http.delete(this.resolvePath(path)).pipe(this.catchError());
    }
    get(path) {
        return this.http.get(this.resolvePath(path)).pipe(this.catchError());
    }
    handleError(error) {
        // In a real world app, we might use a remote logging infrastructure
        // We'd also dig deeper into the error to get a better message
        const errMsg = error.message ? error.message : error.status ? `${error.status} - ${error.statusText}` : 'Server error';
        console.error(errMsg); // log to console instead
        return throwError(errMsg);
    }
    patch(data, path) {
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.patch(this.resolvePath(path), data, options).pipe(this.catchError());
    }
    post(data, path) {
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.post(this.resolvePath(path), data, options).pipe(this.catchError());
    }
    put(data, path) {
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.put(this.resolvePath(path), data, options).pipe(this.catchError());
    }
    resolvePath(path) {
        if (this.settings && this.settings.APIRoot) {
            if (this.settings.APIRoot.endsWith('/')) {
                this.settings.APIRoot = this.settings.APIRoot.substring(this.settings.APIRoot.length - 1);
            }
            if (path.startsWith('/')) {
                path = path.substring(1);
            }
            return `${this.settings.APIRoot}/${path}`;
        }
        else {
            return path;
        }
    }
}

class ObservableContextService {
    //  Constructors
    constructor() {
        this.loader = new BehaviorSubject(false);
        this.subject = new BehaviorSubject(this.defaultValue());
        this.Context = this.subject.asObservable();
        this.Loading = this.loader.asObservable();
    }
    //  Helpers
    loading(loading) {
        this.loader.next(loading);
    }
    defaultValue() {
        return null;
    }
}

class RealTimeConnection {
    //  Constructors
    constructor(http, rtUrl, actionUrl, maxConnectionRetryAttempts = 10) {
        this.http = http;
        this.actionUrl = actionUrl;
        this.connectionAttempts = 0;
        this.rtUrl = rtUrl;
        this.ConnectionError = new EventEmitter();
        this.ReconnectionAttempt = new EventEmitter();
        this.MaxConnectionRetryAttempts = maxConnectionRetryAttempts;
        this.Started = new EventEmitter();
    }
    //  API Methods
    Start(transport = HttpTransportType.WebSockets) {
        this.buildHub(transport).then((hub) => {
            this.Hub = hub;
            this.Hub.serverTimeoutInMilliseconds = 600000;
            this.Hub.onclose(err => {
                console.log('onclose: ' + err);
                this.retryConnection();
            });
            try {
                this.Hub.start()
                    .then(() => {
                    this.connectionAttempts = 0;
                    console.log(`Connection started`);
                    this.Started.emit(this.Hub);
                })
                    .catch(err => {
                    console.log('Error while starting connection: ' + err);
                    this.ConnectionError.emit(err);
                    this.retryConnection();
                });
            }
            catch (err) {
                console.log('Error while starting connection: ' + err);
                this.retryConnection();
            }
        });
    }
    RegisterHandler(methodName) {
        return new Observable(obs => {
            if (this.Hub) {
                try {
                    this.Hub.on(methodName, req => {
                        obs.next(req);
                    });
                }
                catch (err) {
                    console.log(`Error while handling ${methodName}: ` + err);
                    obs.error(err);
                }
            }
            else {
                obs.error('The hub must be started and configured before registering a handler.');
            }
        });
    }
    InvokeAction(methodName, headers, args) {
        const url = `${this.actionUrl}/${methodName}`;
        return this.http.post(url, args, {
            headers,
            withCredentials: true
        });
    }
    Invoke(methodName, ...args) {
        return new Observable(obs => {
            if (this.Hub) {
                try {
                    this.Hub.invoke(methodName, ...args)
                        .then(res => {
                        obs.next(res);
                    })
                        .catch(e => {
                        obs.error(e);
                    });
                }
                catch (err) {
                    console.log(`Error while invoking ${methodName}: ` + err);
                    obs.error(err);
                }
            }
            else {
                obs.error('The hub must be started and configured before invoking.');
            }
        });
    }
    //  Helpers
    buildHub(transport) {
        return __awaiter(this, void 0, void 0, function* () {
            return new HubConnectionBuilder()
                .configureLogging(LogLevel.Information)
                .withUrl(this.rtUrl, {
                transport
            })
                .build();
        });
    }
    stop() {
        return this.Hub.stop();
    }
    /**
     * Retry connection
     */
    retryConnection() {
        if (this.connectionAttempts < this.MaxConnectionRetryAttempts) {
            console.log(`Retrying connection attempt ${this.connectionAttempts}`);
            this.connectionAttempts += 1;
            setTimeout(() => {
                this.reconnect();
            }, 1000);
        }
        else if (this.connectionAttempts >= this.MaxConnectionRetryAttempts) {
            this.stop().then();
            this.ConnectionError.emit('The maximum number of connection retries has been met.');
        }
    }
    /**
     * Attempt to reconnect
     */
    reconnect() {
        this.ReconnectionAttempt.emit(this.connectionAttempts);
        this.Start();
    }
}

//  TODO:  Need to manage reconnection to hub scenarios here
class RealTimeService {
    //  Constructors
    constructor(injector) {
        this.injector = injector;
        this.ReconnectionAttempt = new Subject();
        this.connectionAttempts = 0;
        try {
            this.Settings = injector.get(LCUServiceSettings);
            this.zone = injector.get(NgZone);
        }
        catch (err) { }
        this.started = new ReplaySubject();
        this.Started = this.started.asObservable();
        this.start();
    }
    //  API Methods
    Start() {
        return new Promise((resolve, reject) => {
            this.buildHub('').then((hub) => {
                this.hub = hub;
                this.hub.serverTimeoutInMilliseconds = 600000;
                this.hub.onclose(err => {
                    console.log('onclose: ' + err);
                    this.retryConnection();
                });
                try {
                    this.hub
                        .start()
                        .then(() => {
                        // this.connectionAttempts = 0;
                        console.log(`Connection started`);
                        resolve(this.hub);
                    })
                        .catch(err => {
                        this.retryConnection();
                        if (this.showConnectionError) {
                            reject(err);
                            console.log('Error while starting connection: ' + err);
                            this.showConnectionError = false;
                        }
                    });
                }
                catch (err) {
                    console.log('Error while starting connection 02: ' + err);
                    this.retryConnection();
                }
            });
        });
    }
    RegisterHandler(methodName) {
        return this.WithHub(hub => {
            return Observable.create((obs) => {
                try {
                    hub.on(methodName, req => {
                        obs.next(req);
                        this.zone.run(() => { });
                    });
                }
                catch (err) {
                    console.log(`Error while handling ${methodName}: ` + err);
                    obs.error(err);
                    this.retryConnection();
                }
            });
        });
    }
    Invoke(methodName, ...args) {
        return this.WithHub(hub => {
            return Observable.create((obs) => {
                try {
                    hub
                        .invoke(methodName, ...args)
                        .then(res => {
                        obs.next(res);
                        this.zone.run(() => { });
                    })
                        .catch(e => {
                        obs.error(e);
                    });
                }
                catch (err) {
                    console.log(`Error while invoking ${methodName}: ` + err);
                    obs.error(err);
                    this.retryConnection();
                }
            });
        });
    }
    WithHub(action) {
        try {
            return Observable.create((obs) => {
                if (this.hub.state !== HubConnectionState.Connected) {
                    this.Start().then(hub => {
                        console.log('Restarting connection in flight...');
                        this.runWithHub(obs, action);
                    });
                }
                else {
                    this.runWithHub(obs, action);
                }
            });
        }
        catch (err) {
            return Observable.create((obs) => {
                obs.error(err);
                obs.complete();
            });
        }
    }
    //  Helpers
    buildHub(urlRoot) {
        return __awaiter(this, void 0, void 0, function* () {
            this.url = this.buildHubUrl(urlRoot);
            return (new HubConnectionBuilder()
                .configureLogging(LogLevel.Information)
                .withUrl(this.url)
                // .withUrl(this.url, {
                //   transport: signalR.HttpTransportType.LongPolling
                // })
                .build());
        });
    }
    buildHubUrl(urlRoot) {
        const url = this.loadHubUrl(urlRoot);
        return url;
    }
    loadHubPath() {
        const stateRoot = this.loadStateRoot();
        return `${stateRoot}?lcu-app-id=${this.Settings.AppConfig.ID}&lcu-app-ent-api-key=${this.Settings.AppConfig.EnterpriseAPIKey}&lcu-environment=${this.Settings.StateConfig.Environment || ''}`;
    }
    loadHubUrl(urlRoot) {
        const apiRoot = this.Settings ? this.Settings.APIRoot || '' : '';
        const hubPath = this.loadHubPath();
        return `${apiRoot}${urlRoot || ''}${hubPath}`;
    }
    loadStateRoot() {
        // return this.Settings.StateConfig &&
        //   this.Settings.StateConfig.Root !== undefined
        //   ? this.Settings.StateConfig.Root
        //   : '/state';
        return '/state';
    }
    runWithHub(obs, action) {
        const res = action(this.hub);
        if (res) {
            res.subscribe(r => {
                obs.next(r);
                this.zone.run(() => { });
            }, e => {
                obs.error(e);
            });
        }
    }
    start() {
        setTimeout(() => {
            this.Start().then(hub => this.started.next(hub));
        }, 50);
    }
    stop() {
        // this.hub.stop();
        this.showConnectionError = true;
    }
    /**
     * Retry connection
     */
    retryConnection() {
        if (this.connectionAttempts < 5) {
            console.log(`Retrying connection attempt ${this.connectionAttempts}`);
            this.connectionAttempts += 1;
            this.reconnect();
        }
        else if (this.connectionAttempts >= 5) {
            this.stopReconnection();
        }
    }
    /**
     * Attempt to reconnect
     */
    reconnect() {
        this.attemptingToReconnect = true;
        this.reconnectionMessage();
        this.start();
    }
    /**
     * Stop trying to reconnect
     */
    stopReconnection() {
        this.attemptingToReconnect = false;
        this.reconnectionMessage();
        this.stop();
    }
    /**
     * Notify user of reconnection attempt(s)
     */
    reconnectionMessage() {
        this.ReconnectionAttempt.next(this.attemptingToReconnect);
    }
}
RealTimeService.ctorParameters = () => [
    { type: Injector }
];
RealTimeService.ɵprov = ɵɵdefineInjectable({ factory: function RealTimeService_Factory() { return new RealTimeService(ɵɵinject(INJECTOR)); }, token: RealTimeService, providedIn: "root" });
RealTimeService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
RealTimeService.ctorParameters = () => [
    { type: Injector }
];

class Status {
    get Metadata() {
        const meta = {};
        const self = this;
        for (const prop in self) {
            if (prop !== 'Code' && prop !== 'Message' && prop !== 'Metadata') {
                meta[prop] = self[prop];
            }
        }
        return meta;
    }
}
function isStatusSuccess(status) {
    return status && status.Code === 0;
}

class BaseResponse {
}
function isResultSuccess(result) {
    return result && isStatusSuccess(result.Status);
}

class BaseModeledResponse extends BaseResponse {
}

class LCUElementContext {
}

// @Directive()
class LcuElementComponent {
    //  Constructors
    constructor(injector) {
        this.injector = injector;
    }
    //  Life Cycle
    ngOnInit() { }
    //  API Methods
    SetContext(ctxt) {
        this.context = ctxt;
    }
}
LcuElementComponent.propDecorators = {
    context: [{ type: Input }],
    SetContext: [{ type: Input }]
};

class Guid {
    // 	Constructors
    constructor(guid) {
        if (!guid) {
            throw new TypeError('Invalid argument; `value` has no value.');
        }
        this.value = Guid.Empty;
        if (guid && Guid.IsGuid(guid)) {
            this.value = guid;
        }
    }
    // 	API Methods
    static Create() {
        return Guid.Parse(Guid.CreateRaw());
    }
    static CreateEmpty() {
        return Guid.Parse('emptyguid');
    }
    static CreateRaw() {
        return [Guid.gen(2), Guid.gen(1), Guid.gen(1), Guid.gen(1), Guid.gen(3)].join('-');
    }
    static IsGuid(guid) {
        const value = guid.toString();
        return guid && (guid instanceof Guid || Guid.GuidValidator.test(value));
    }
    static Parse(guid) {
        return new Guid(guid);
    }
    // 	Helpers
    static gen(count) {
        let out = '';
        for (let i = 0; i < count; i++) {
            // tslint:disable-next-line:no-bitwise
            out += (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return out;
    }
    Equals(other) {
        return Guid.IsGuid(other) && this.value === other.toString();
    }
    IsEmpty() {
        return this.value === Guid.Empty;
    }
    ToJSON() {
        return {
            Value: this.value
        };
    }
    ToString() {
        return this.value;
    }
}
// 	Properties
Guid.GuidValidator = new RegExp('^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$', 'i');
Guid.Empty = '00000000-0000-0000-0000-000000000000';

class LCUInterceptor {
    //  Constuctors
    constructor(settings) {
        this.settings = settings;
    }
    //  API Methods
    intercept(req, next) {
        const clonedRequest = req.clone({
            headers: req.headers
                .set('lcu-app-id', this.settings.AppConfig.ID)
                .set('lcu-app-ent-api-key', this.settings.AppConfig.EnterpriseAPIKey)
                .set('lcu-environment', this.settings.StateConfig.Environment || '')
                .set('lcu-username-mock', this.settings.StateConfig.UsernameMock || '')
        });
        // Pass the cloned request instead of the original request to the next handle
        return next.handle(clonedRequest);
    }
}

class SafePipe {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    transform(url) {
        return this.sanitizer.bypassSecurityTrustResourceUrl(url);
    }
}
SafePipe.ctorParameters = () => [
    { type: DomSanitizer }
];
SafePipe.decorators = [
    { type: Pipe, args: [{
                name: 'safe',
            },] }
];
SafePipe.ctorParameters = () => [
    { type: DomSanitizer }
];

const winAny = window;
class FathymSharedModule {
    //  Fields
    //  API Methods
    static DefaultServiceSettings(env, apiRoot) {
        return {
            APIRoot: winAny.LCU && winAny.LCU.APIRoot ? winAny.LCU.APIRoot : '',
            AppConfig: {
                ID: winAny.LCU && winAny.LCU.Application && winAny.LCU.Application.ID ? winAny.LCU.Application.ID : 'test-app',
                EnterpriseAPIKey: winAny.LCU && winAny.LCU.Application.EnterpriseAPIKey ? winAny.LCU.Application.EnterpriseAPIKey : 'test-app'
            },
            StateConfig: {
                Environment: winAny.LCU.State ? winAny.LCU.State.Environment : '',
                ActionRoot: winAny.LCU.State ? winAny.LCU.State.Root : '',
                Root: winAny.LCU.State ? winAny.LCU.State.Root : '',
                UsernameMock: winAny.LCU.State ? winAny.LCU.State.UsernameMock : ''
            },
            Settings: winAny.LCU.Settings
        };
    }
    static forRoot() {
        return {
            ngModule: FathymSharedModule,
            providers: [
                RealTimeService,
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: LCUInterceptor,
                    multi: true,
                    deps: [LCUServiceSettings]
                }
            ]
        };
    }
}
FathymSharedModule.decorators = [
    { type: NgModule, args: [{
                declarations: [SafePipe],
                imports: [CommonModule, FormsModule, HttpClientModule, RouterModule],
                exports: [CommonModule, FormsModule, HttpClientModule, RouterModule, SafePipe],
                providers: [],
                entryComponents: []
            },] }
];

class Icon {
}

class JSONSchema {
}

class Loading {
    // 	Constructors
    constructor() {
        this.Clear();
    }
    // 	Properties
    get Value() {
        return this.loaders && this.loaders.length > 0 && this.loaders.every(loader => loader);
    }
    // 	API Methods
    Clear() {
        this.loaders = [];
    }
    Set(loading) {
        if (loading) {
            this.loaders.push(loading);
        }
        else if (this.loaders && this.loaders.length > 0) {
            this.loaders.splice(0, 1);
        }
    }
}
Loading.ɵprov = ɵɵdefineInjectable({ factory: function Loading_Factory() { return new Loading(); }, token: Loading, providedIn: "root" });
Loading.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
Loading.ctorParameters = () => [];

class Pageable {
}

class Application {
}
class DAFApplication {
}
class DAFAPIApplicationDetails {
}
class DAFViewApplicationDetails {
}
class DAFRedirectApplicationDetails {
}
class DAFAppPointerApplicationDetails {
}

class RouterHelpersService {
    // 	Constructors
    constructor(route, router) {
        this.route = route;
        this.router = router;
    }
    // 	API Methods
    RunOnNavigation(action, navEventType = null) {
        return this.router.events.subscribe((event) => {
            if (navEventType == null || event instanceof navEventType) {
                action(event);
            }
        });
    }
    RunOnNavigationCancel(action) {
        return this.RunOnNavigation(action, NavigationCancel);
    }
    RunOnNavigationEnd(action) {
        return this.RunOnNavigation(action, NavigationEnd);
    }
    RunOnNavigationError(action) {
        return this.RunOnNavigation(action, NavigationError);
    }
    RunOnNavigationStart(action) {
        return this.RunOnNavigation(action, NavigationStart);
    }
    RunOnQueryParam(param, action, checkVal = null) {
        return this.route.queryParams.subscribe((params) => {
            if ((checkVal == null && params[param] != null) || (checkVal != null && params[param] === checkVal)) {
                action(params);
            }
        });
    }
    RunOnRouteParam(param, action, checkVal = null) {
        return this.route.params.subscribe((params) => {
            if ((checkVal == null && params[param] != null) || (checkVal != null && params[param] === checkVal)) {
                action(params);
            }
        });
    }
}
RouterHelpersService.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: Router }
];
RouterHelpersService.ɵprov = ɵɵdefineInjectable({ factory: function RouterHelpersService_Factory() { return new RouterHelpersService(ɵɵinject(ActivatedRoute), ɵɵinject(Router)); }, token: RouterHelpersService, providedIn: "root" });
RouterHelpersService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
RouterHelpersService.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: Router }
];

class StateAction {
    constructor() {
        this.Arguments = {};
    }
}
class ImageMessage {
}

//  TODO:  Need to manage reconnection to hub scenarios here
class StateContext extends ObservableContextService {
    //  Constructors
    constructor(injector) {
        super();
        this.injector = injector;
        this.connectedToState = new BehaviorSubject({
            Code: -1,
            Message: 'Initialized',
        });
        this.ConnectedToState = this.connectedToState.asObservable();
        this.http = injector.get(HttpClient);
        this.ReconnectionAttempt = new Subject();
        this.Settings = injector.get(LCUServiceSettings);
        const rtUrl = this.buildHubUrl('');
        const actionUrl = this.loadActionUrl('');
        this.rt = new RealTimeConnection(this.http, rtUrl, actionUrl);
        this.rt.ReconnectionAttempt.subscribe((val) => {
            this.ReconnectionAttempt.next(val);
        });
        this.setup();
    }
    //  API Methods
    Execute(action) {
        return this.executeAction(action);
    }
    $Refresh(args = {}) {
        this.Execute({
            Arguments: args,
            Type: 'Refresh',
        });
    }
    Start(shouldUpdate) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.startSub) {
                this.startSub = this.rt.Started.subscribe(() => __awaiter(this, void 0, void 0, function* () {
                    const groupName = yield this.connectToState(shouldUpdate);
                    this.setupReceiveState(groupName);
                    this.connectedToState.next({ Code: 0, Message: 'Success' });
                    this.callRefresh();
                }));
                this.rt.Start();
            }
        });
    }
    //  Helpers
    buildActionUrl(urlRoot) {
        const url = this.loadActionUrl(urlRoot);
        return url;
    }
    buildHubUrl(urlRoot) {
        const url = this.loadHubUrl(urlRoot);
        return url;
    }
    callRefresh() {
        this.$Refresh();
    }
    connectToState(shouldUpdate) {
        return __awaiter(this, void 0, void 0, function* () {
            const stateKey = this.loadStateKey();
            const stateName = this.loadStateName();
            const env = this.loadEnvironment();
            return new Promise((resolve, reject) => {
                this.rt
                    .InvokeAction('ConnectToState', this.loadHeaders(), {
                    ShouldSend: shouldUpdate,
                    Key: stateKey,
                    State: stateName,
                    Environment: env,
                })
                    .subscribe({
                    next: (req) => {
                        if (req.status && req.status.code === 0) {
                            resolve(req.groupName);
                        }
                        else {
                            reject(req.status
                                ? req.status.message
                                : 'Unknown issue connecting to state.');
                        }
                    },
                    error: (err) => reject(err),
                });
            });
        });
    }
    defaultValue() {
        return {};
    }
    executeAction(action) {
        return __awaiter(this, void 0, void 0, function* () {
            const stateKey = this.loadStateKey();
            const stateName = this.loadStateName();
            return this.rt
                .InvokeAction(action.Type, this.loadHeaders(), Object.assign(Object.assign({}, action), { Key: stateKey, State: stateName }))
                .subscribe();
        });
    }
    loadActionPath() {
        const actionRoot = this.loadStateActionRoot();
        return `${actionRoot}`; // ?lcu-app-id=${this.Settings.AppConfig.ID}&lcu-app-ent-api-key=${this.Settings.AppConfig.EnterpriseAPIKey}`;
    }
    loadActionUrl(urlRoot) {
        const apiRoot = this.Settings ? this.Settings.APIRoot || '' : '';
        const actionPath = this.loadActionPath();
        return `${apiRoot}${urlRoot || ''}${actionPath}`;
    }
    loadEnvironment() {
        let env = this.Settings.StateConfig
            ? this.Settings.StateConfig.Environment
            : null;
        if (!env) {
            env = '';
        }
        return env;
    }
    loadHeaders() {
        return {
            'lcu-ent-api-key': this.Settings.AppConfig.EnterpriseAPIKey,
            'lcu-hub-name': this.loadStateName(),
            'lcu-state-key': this.loadStateKey(),
            'lcu-environment': this.loadEnvironment(),
            'lcu-username-mock': this.loadUsernameMock(),
        };
    }
    loadHubPath() {
        const stateRoot = this.loadStateRoot();
        const env = this.loadEnvironment();
        const unmock = this.loadUsernameMock();
        return `${stateRoot}?lcu-app-ent-api-key=${this.Settings.AppConfig.EnterpriseAPIKey}&lcu-environment=${env}&lcu-username-mock=${unmock}`;
    }
    loadHubUrl(urlRoot) {
        const apiRoot = this.Settings ? this.Settings.APIRoot || '' : '';
        const hubPath = this.loadHubPath();
        return `${apiRoot}${urlRoot || ''}${hubPath}`;
    }
    loadStateRoot() {
        const stateRoot = this.Settings.StateConfig && this.Settings.StateConfig.Root !== undefined
            ? this.Settings.StateConfig.Root
            : '';
        return `${stateRoot}/${this.loadStateName()}`;
    }
    loadStateActionRoot() {
        const stateActinRoot = this.Settings.StateConfig &&
            this.Settings.StateConfig.ActionRoot !== undefined
            ? this.Settings.StateConfig.ActionRoot
            : '';
        return `${stateActinRoot}/${this.loadStateName()}`;
    }
    loadUsernameMock() {
        return this.Settings.StateConfig && this.Settings.StateConfig.UsernameMock
            ? this.Settings.StateConfig.UsernameMock
            : '';
    }
    setup() {
        this.Start(false).then();
    }
    setupReceiveState(groupName) {
        this.rt.RegisterHandler(`ReceiveState=>${groupName}`).subscribe((req) => {
            console.log(`Handled state from ${groupName}`);
            this.subject.next(req);
        });
    }
}

const materialModules = [
    MatAutocompleteModule,
    MatBadgeModule,
    MatButtonModule,
    MatButtonToggleModule,
    MatCardModule,
    MatCheckboxModule,
    MatChipsModule,
    MatDatepickerModule,
    MatDialogModule,
    MatDividerModule,
    MatExpansionModule,
    MatFormFieldModule,
    MatGridListModule,
    MatIconModule,
    MatInputModule,
    MatListModule,
    MatMenuModule,
    MatPaginatorModule,
    MatProgressBarModule,
    MatProgressSpinnerModule,
    MatRadioModule,
    MatSelectModule,
    MatSidenavModule,
    MatSlideToggleModule,
    MatSliderModule,
    MatSnackBarModule,
    MatSortModule,
    MatStepperModule,
    MatTableModule,
    MatTabsModule,
    MatToolbarModule,
    MatTooltipModule,
    MatTreeModule,
    // CDK Modules
    CdkTableModule,
    DragDropModule,
];
const providers = [
    MatBottomSheet,
    MatPaginator,
    MatSort,
    MatTableDataSource
];
class MaterialModule {
    static forRoot() {
        return {
            ngModule: MaterialModule,
            providers: [providers]
        };
    }
}
MaterialModule.decorators = [
    { type: NgModule, args: [{
                declarations: [],
                imports: [materialModules],
                exports: [materialModules]
            },] }
];

// @dynamic
/**
 * @dynamic need this because there are static memebers
 *
 * Data pipe constants
 */
class DataPipeConstants {
}
DataPipeConstants.DATE_FMT = 'dd/MMM/yyyy';
DataPipeConstants.DATE_SHORTDATE = 'M/d/yy';
DataPipeConstants.DATE_MEDIUMDATE = 'MMM d, y';
DataPipeConstants.DATE_FULLDATE = 'EEEE, MMMM d, y';
DataPipeConstants.DATE_TIME_FMT = `${DataPipeConstants.DATE_FMT} hh:mm:ss`;
DataPipeConstants.DATE_DAY_TIME = 'EEE h a';
DataPipeConstants.PIPE_DATE = 'date';
DataPipeConstants.PIPE_SHORTDATE = 'shortdate';
DataPipeConstants.PIPE_MEDIUMDATE = 'mediumdate';
DataPipeConstants.PIPE_FULLDATE = 'fulldate';
DataPipeConstants.PIPE_DECIMAL_TWO = 'decimalTwo';
DataPipeConstants.PIPE_DECIMAL_FOUR = 'decimalfour';
DataPipeConstants.PIPE_EPOCH = 'epoch';
DataPipeConstants.PIPE_NUMBER = 'number';
DataPipeConstants.PIPE_PERCENTAGE = 'percentage';
DataPipeConstants.PIPE_PERCENTAGE_DECIMAL = 'percentagedecimal';
DataPipeConstants.PIPE_TEMP_FAHRENHEIT = 'fahrenheit';
DataPipeConstants.PIPE_TEMP_CELSIUS = 'celsius';
DataPipeConstants.PIPE_TEMP_KELVIN = 'kelvin';
DataPipeConstants.PIPE_MPH = 'mph';

// @dynamic
/**
 * @dynamic is used because this class contains static properties
 */
class TemperatureConversion {
    /**
     * Convert fahrenheit to celsius
     * @param val fahrenheit value
     */
    static FahrenheitToCelsius(val) {
        return (val - 32) / 1.8;
    }
    /**
     * Convert fahrenheit to kelvin
     * @param val fahrenheit value
     */
    static FahrenheitToKelvin(val) {
        return ((val - 32) / 1.8) + 273.15;
    }
    /**
     * Convert celsius to fahrenheit
     * @param val celsius value
     */
    static CelsiusToFahrenheit(val) {
        return (val * 1.8) + 32;
    }
    /**
     * Convert celsius to kelvin
     * @param val celsius value
     */
    static CelsiusToKelvin(val) {
        return val + 273.15;
    }
    /**
     * Convert kelvin to fahrenheit
     * @param val kelvin value
     */
    static KelvinToFahrenheit(val) {
        return ((val - 273.15) * 1.8) + 32;
    }
    /**
     * Convert kelvin to celsius
     * @param val kelvin value
     */
    static KelvinToCelsius(val) {
        return val - 273.15;
    }
}

/**
 * Pipes for datagrid values
 */
class DataPipes {
    transform(value, args) {
        if (!args) {
            return value;
        }
        /**
         * Return date M/d/YYYY
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_SHORTDATE) {
            const pipe = new DatePipe('en-US');
            const transformed = pipe.transform(value, DataPipeConstants.DATE_SHORTDATE);
            return transformed;
        }
        /**
         * Return date
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_DATE) {
            const pipe = new DatePipe('en-US');
            const transformed = pipe.transform(value, DataPipeConstants.DATE_FMT);
            return transformed;
        }
        /**
         * Return date MMM d, y
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_MEDIUMDATE) {
            const pipe = new DatePipe('en-US');
            const transformed = pipe.transform(value, 'MMM d, y');
            return transformed;
        }
        /**
         * Return date 'EEEE, MMMM d, y'
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_FULLDATE) {
            const pipe = new DatePipe('en-US');
            const transformed = pipe.transform(value, DataPipeConstants.DATE_FULLDATE);
            return transformed;
        }
        /**
         * Return date from epoch value
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_EPOCH) {
            const pipe = new DatePipe('en-US');
            const transformed = pipe.transform(value * 1000, DataPipeConstants.DATE_DAY_TIME);
            return transformed;
        }
        /**
         * Return number with decimal
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_NUMBER) {
            const pipe = new DecimalPipe('en-US');
            const transformed = pipe.transform(value, '1.0-0');
            return transformed;
        }
        /**
         * Return return number with mph appended
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_MPH) {
            const pipe = new DecimalPipe('en-US');
            const transformed = pipe.transform(value, '1.0-0');
            return transformed + ' mph';
        }
        /**
         * Return percentage
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_PERCENTAGE) {
            const pipe = new PercentPipe('en-US');
            const transformed = pipe.transform(value);
            return transformed;
        }
        /**
         * Return percentage
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_PERCENTAGE_DECIMAL) {
            const pipe = new PercentPipe('en-US');
            const transformed = pipe.transform(value, '2.2-2');
            return transformed;
        }
        /**
         * Return percentage with two decimals
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_DECIMAL_TWO) {
            const pipe = new DecimalPipe('en-US');
            const transformed = pipe.transform(value, '1.2-2');
            return transformed;
        }
        /**
         * Return percentage with four decimals
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_DECIMAL_FOUR) {
            const pipe = new DecimalPipe('en-US');
            const transformed = pipe.transform(value, '1.4-4');
            return transformed;
        }
        /**
         * Return temperature in fahrenheit, append °F
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_TEMP_FAHRENHEIT) {
            const pipe = new DecimalPipe('en-US');
            // const temperature = (value * 32) + 1.8;
            const transformed = pipe.transform(value, '1.0-0');
            return transformed + ' °F';
        }
        /**
         * Return temperature in kelvin, append °K
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_TEMP_KELVIN) {
            const pipe = new DecimalPipe('en-US');
            // const temperature = (value * 32) + 1.8;
            const temperature = TemperatureConversion.FahrenheitToKelvin(value);
            const transformed = pipe.transform(temperature, '1.0-0');
            return transformed + ' °K';
        }
        /**
         * Return temperature in celsius, append °C
         */
        if (args.toLowerCase() === DataPipeConstants.PIPE_TEMP_CELSIUS) {
            const pipe = new DecimalPipe('en-US');
            const temperature = TemperatureConversion.FahrenheitToCelsius(value);
            const transformed = pipe.transform(temperature, '1.0-0');
            return transformed + ' °C';
        }
        /**
         * IF none of the above work, return the original value
         */
        return value;
    }
}
DataPipes.decorators = [
    { type: Pipe, args: [{
                name: 'dataPipes'
            },] }
];

/**
 * Pipe used to sanitze html
 */
class SanitizeHtmlPipe {
    constructor(sanitized) {
        this.sanitized = sanitized;
    }
    transform(value) {
        return this.sanitized.bypassSecurityTrustHtml(value);
    }
}
SanitizeHtmlPipe.ctorParameters = () => [
    { type: DomSanitizer }
];
SanitizeHtmlPipe.decorators = [
    { type: Pipe, args: [{
                name: 'sanitizeHtml'
            },] }
];
SanitizeHtmlPipe.ctorParameters = () => [
    { type: DomSanitizer }
];

class DecamelizePipe {
    transform(value, separator) {
        const sep = separator ? separator : ' ';
        return value
            .charAt(0).toUpperCase() + value.slice(1)
            .replace(/([a-z\d])([A-Z])/g, '$1' + sep + '$2')
            .replace(/([A-Z]+)([A-Z][a-z\d]+)/g, '$1' + sep + '$2');
    }
}
DecamelizePipe.decorators = [
    { type: Pipe, args: [{
                name: 'decamelize'
            },] }
];

// @dynamic
/**
 * @dynamic need this because there are static members
 *
 * String pipe constants
 */
class StringPipeConstants {
}
StringPipeConstants.PIPE_CAMEL_CASE = 'camelcase';
StringPipeConstants.PIPE_KEBAB_CASE = 'kebabcase';
StringPipeConstants.PIPE_PASCAL_CASE = 'pascalcase';
StringPipeConstants.PIPE_SNAKE_CASE = 'snakecase';

/**
 * Pipes for string values
 */
class StringPipes {
    transform(value, args) {
        if (!args) {
            return value;
        }
        /**
         * Return given string formatted as 'thisIsCamelCase'.
         *
         * For example, 'test-string' or 'Test String' would be converted to 'testString'.
         */
        if (args.toLowerCase() === StringPipeConstants.PIPE_CAMEL_CASE) {
            value = value.toString();
            return value.toLowerCase()
                .replace(/[^a-zA-Z0-9]+(.)/g, (_, chr) => chr.toUpperCase());
        }
        /**
         * Return given string formatted as 'this-is-kebab-case'.
         *
         * For example, 'TestString' would be converted to 'test-string'.
         */
        if (args.toLowerCase() === StringPipeConstants.PIPE_KEBAB_CASE) {
            return value.match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
                .map((x) => x.toLowerCase())
                .join('-');
        }
        /**
         * Return given string formatted as 'ThisIsPascalCase'.
         *
         * For example, 'testString' would be converted to 'Teststring'.
         * For example, 'test string' would be converted to 'Test String'.
         */
        if (args.toLowerCase() === StringPipeConstants.PIPE_PASCAL_CASE) {
            return value.replace(/\w\S*/g, (m) => m.charAt(0).toUpperCase() + m.substr(1).toLowerCase());
        }
        /**
         * Return given string formatted as 'this_is_snake_case'.
         *
         * For example, 'TestString' would be converted to 'test_string'.
         */
        if (args.toLowerCase() === StringPipeConstants.PIPE_SNAKE_CASE) {
            return value.match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
                .map((x) => x.toLowerCase())
                .join('_');
        }
        /**
         * If none of the above work, return the original value.
         */
        return value;
    }
}
StringPipes.decorators = [
    { type: Pipe, args: [{
                name: 'stringPipes'
            },] }
];

class PipeModule {
    static forRoot() {
        return {
            ngModule: PipeModule,
            providers: [],
        };
    }
}
PipeModule.decorators = [
    { type: NgModule, args: [{
                declarations: [DataPipes, SanitizeHtmlPipe, DecamelizePipe, StringPipes],
                imports: [CommonModule],
                exports: [DataPipes, SanitizeHtmlPipe, DecamelizePipe, StringPipes]
            },] }
];

/**
 * Directive for setting theme color to a element, color comes from
 * a local stylesheet
 *
 * @use as an attribute: lcuColor="primary-A700", sets the class
 * .primary-A700 {} to the host element
 */
/**
 * SCSS class types
 */
// type ColorClasses = 'primary-500' | 'accent-500' | 'warn-500';
class ThemeColorDirective {
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
    }
    /**
     * Set the current theme class to the element
     */
    set lcuThemeColor(val) {
        // this.renderer.removeClass(this.element.nativeElement, `${val}`);
        this.renderer.addClass(this.element.nativeElement, `${val}`);
        // this.readProperty('background-color');
    }
    /**
     * Get the value of a specific style property
     *
     * @param property style property
     */
    readProperty(property) {
        console.log('element', window.getComputedStyle(this.element.nativeElement).getPropertyValue(property));
        // console.log('element', window.getComputedStyle(this.element.nativeElement));
    }
}
ThemeColorDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
ThemeColorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[lcuThemeColor]'
            },] }
];
ThemeColorDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
ThemeColorDirective.propDecorators = {
    lcuThemeColor: [{ type: Input }]
};

class DirectiveModule {
    static forRoot() {
        return {
            ngModule: DirectiveModule,
            providers: [],
        };
    }
}
DirectiveModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    ThemeColorDirective
                ],
                imports: [CommonModule],
                exports: [ThemeColorDirective]
            },] }
];

class Base64Model {
    constructor(blob, file) {
        this.Blob = blob;
        this.File = file;
    }
}

class DataFlow {
}

class DataFlowAction {
}

class DataFlowActionEvent {
}

class DataFlowLayout {
}

class DataFlowModule {
}

var DataFlowModuleControlTypes;
(function (DataFlowModuleControlTypes) {
    DataFlowModuleControlTypes["Direct"] = "Direct";
    DataFlowModuleControlTypes["Flow"] = "Flow";
    DataFlowModuleControlTypes["Gate"] = "Gate";
    DataFlowModuleControlTypes["Helper"] = "Helper";
})(DataFlowModuleControlTypes || (DataFlowModuleControlTypes = {}));

class DataFlowModuleDisplay {
}

class DataFlowModuleOption {
}

class DataFlowModulePack {
}

class DataFlowModulePackSetup {
}

var DataFlowModuleShapeTypes;
(function (DataFlowModuleShapeTypes) {
    DataFlowModuleShapeTypes["Rectangle"] = "Rectangle";
    DataFlowModuleShapeTypes["Circle"] = "Circle";
    DataFlowModuleShapeTypes["Ellipse"] = "Ellipse";
    DataFlowModuleShapeTypes["Custom"] = "Custom";
})(DataFlowModuleShapeTypes || (DataFlowModuleShapeTypes = {}));

class DataFlowOutput {
}

class DataFlowStream {
}

class IdeActivity {
}

class IdeEditor {
}

class IdePanel {
}

class IdeSideBar {
}
class IdeSideBarAction {
}

class NapkinIDESolution {
}

class ValidationMessageModel {
}

// @dynamic
/**
 * @dynamic need this because there are static memebers
 *
 * Email constants
 */
class EmailValidationMessagesConstants {
}
/**
 * Constant for email pattern
 */
EmailValidationMessagesConstants.PATTERN = 'Enter a valid email';
/**
 * Constant for email required
 */
EmailValidationMessagesConstants.REQUIRED = 'Email is required';

// @dynamic
/**
 * @dynamic need this because there are static memebers
 *
 * Username constants
 */
class UserValidationMessagesConstants {
    /**
     * Constant for username max length
     *
     * @param length charcter length
     */
    static MAX_LENGTH(length) {
        return 'Username cannon be more than ' + length.toString() + ' characters long.';
    }
    /**
     * Constant for username min length
     *
     * @param length charcter length
     */
    static MIN_LENGTH(length) {
        return 'Username must be at least ' + length.toString() + ' characters long.';
    }
}
/**
 * Constant for username pattern
 */
UserValidationMessagesConstants.PATTERN = 'Username cannot end or begin with _ or .';
/**
 * Constant for username required
 */
UserValidationMessagesConstants.REQUIRED = 'Username is required';
/**
 * Constant for valid username
 */
UserValidationMessagesConstants.VALID = 'Your username has already been taken';

// @dynamic
/**
 * @dynamic need this because there are static memebers
 */
class Constants {
}
Constants.REQUIRED = 'required';
Constants.MIN_LENGTH = 'minlength';
Constants.MAX_LENGTH = 'maxlength';
Constants.PATTERN = 'pattern';
Constants.VALID_USERNAME = 'ValidUsername';
Constants.PASSWORDS_MATCH = 'PasswordsMatch';

// @dynamic
/**
 * @dynamic need this because there are static memebers
 *
 * Password / Confirm password constants
 */
class PasswordValidationMessagesConstants {
    /**
     * Constant for password min length
     *
     * @param length charcter length
     */
    static MIN_LENGTH(length) {
        return 'Password must be at least ' + length.toString() + ' characters long.';
    }
}
/**
 * Constant for confirm password equal
 */
PasswordValidationMessagesConstants.PASSWORDS_MATCH = 'Password mismatch';
/**
 * Constant for confirm password required
 */
PasswordValidationMessagesConstants.CONFIRM_REQUIRED = 'Confirm password is required';
/**
 * Constant for password pattern
 */
PasswordValidationMessagesConstants.PATTERN = 'At least one uppercase, one lowercase, one special character, and one number';
/**
 * Constant for password required
 */
PasswordValidationMessagesConstants.REQUIRED = 'Password is required';

// @dynamic
/**
 * @dynamic need this because there are static memebers
 *
 * Email constants
 */
class ZipcodeValidationMessagesConstants {
}
/**
 * Constant for email pattern
 */
ZipcodeValidationMessagesConstants.PATTERN = 'Please enter a valid zipcode';
/**
 * Constant for email required
 */
ZipcodeValidationMessagesConstants.REQUIRED = 'Email is required';

// @dynamic
/**
 * @dynamic need this because there are static memebers
 */
class ValidationMessages {
}
ValidationMessages.UserName = [
    { Type: Constants.REQUIRED, Message: UserValidationMessagesConstants.REQUIRED },
    { Type: Constants.MIN_LENGTH, Message: UserValidationMessagesConstants.MIN_LENGTH(5) },
    { Type: Constants.MAX_LENGTH, Message: UserValidationMessagesConstants.MAX_LENGTH(25) },
    { Type: Constants.PATTERN, Message: UserValidationMessagesConstants.PATTERN },
    { Type: Constants.VALID_USERNAME, Message: UserValidationMessagesConstants.VALID }
];
ValidationMessages.Email = [
    { Type: Constants.REQUIRED, Message: EmailValidationMessagesConstants.REQUIRED },
    { Type: Constants.PATTERN, Message: EmailValidationMessagesConstants.PATTERN }
];
ValidationMessages.Password = [
    { Type: Constants.REQUIRED, Message: PasswordValidationMessagesConstants.REQUIRED },
    { Type: Constants.MIN_LENGTH, Message: PasswordValidationMessagesConstants.MIN_LENGTH(5) },
    { Type: Constants.PATTERN, Message: PasswordValidationMessagesConstants.PATTERN }
];
ValidationMessages.ConfirmPassword = [
    { Type: Constants.REQUIRED, Message: PasswordValidationMessagesConstants.CONFIRM_REQUIRED },
    { Type: Constants.PASSWORDS_MATCH, Message: PasswordValidationMessagesConstants.PASSWORDS_MATCH }
];
ValidationMessages.Zipcode = [
    { Type: Constants.PATTERN, Message: ZipcodeValidationMessagesConstants.PATTERN }
];

// @dynamic
/**
 * @dynamic need this because there are static memebers
 */
class EmailValidator {
}
/**
 * Email regular expression pattern, requires domain
 */
// public static readonly EmailPatternDomain: RegExp = /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{3,4}$/gmi;
EmailValidator.EmailPatternDomain = '\[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-z]{2,4}';

// @dynamic
/**
 * @dynamic is used because this class contains static properties
 */
class PasswordValidator {
    /**
     * Check if password and confirm password match
     *
     * @param password password control
     * @param confirm confirm password control
     */
    static PasswordsMatch(password, confirm) {
        if (!password || !confirm) {
            return;
        }
        return (formGroup) => {
            if (password.value !== confirm.value) {
                confirm.setErrors({ PasswordsMatch: true });
            }
            else {
                confirm.setErrors(null);
            }
            return;
        };
    }
}
/**
 * Password regular expression pattern:
 *
 * At least one upper case English letter
 *
 * At least one lower case Englis letter
 *
 * At least one digit
 *
 * At least one special character
 *
 * Minimum of eight characters
 */
PasswordValidator.StrongPassword = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,}$)/gm;
// public static readonly StrongPassword: string = '\(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,}$)';
PasswordValidator.MediumPassword = /^(((?=.*[a-z])(?=.*[A-Z]))|((?=.*[a-z])(?=.*[0-9]))|((?=.*[A-Z])(?=.*[0-9])))(?=.{6,})$/gm;

// import * as libphonenumber from 'google-libphonenumber';
// @dynamic
/**
 * @dynamic need this because there are static memebers
 */
class PhoneValidator {
}

// @dynamic
/**
 * @dynamic need this because there are static memebers
 */
class UserNameValidator {
    static ValidUsername(control) {
        if (control.value.toUpperCase() === 'ABC123' || control.value.toUpperCase() === '123ABC') {
            return ({ ValidUsername: true });
        }
        else {
            return (null);
        }
    }
}
/**
 * Username regular expression pattern
 *
 * letters and numbers only
 */
// public static readonly UsernamePattern: RegExp = /^[a-z0-9]*$/gmi;
UserNameValidator.UsernamePattern = '\[a-z0-9]*';
/**
 * Username regular expression pattern
 *
 * letters and numbers
 *
 * Doesn't allow underscore or . at the beginning or end of value
 */
UserNameValidator.UsernameNoUnderscoreDotPattern = /^[a-zA-Z0-9]([._](?![._])|[a-zA-Z0-9]){6,18}[a-zA-Z0-9]$/gmi;

// @dynamic
/**
 * @dynamic need this because there are static memebers
 */
class ZipcodeValidator {
    /**
     * Return postal code regular expression, based off country code
     *
     * @param iso country code
     */
    static ZipcodeExpression(iso) {
        const arr = ZipcodeValidator.globalZipCodes.filter(itm => itm.ISO === iso.toUpperCase());
        let regex;
        arr.find((obj) => {
            return regex = obj['Regex'];
        });
        return regex;
    }
}
/**
 * Regualr expression patterns for global postal codes
 *
 * credit : matthewbednarski
 *
 * https://gist.github.com/jamesbar2/1c677c22df8f21e869cca7e439fc3f5b
 */
ZipcodeValidator.globalZipCodes = [{
        Note: 'The first two digits (ranging from 10–43) correspond to the province, while the last two digits correspond either to the city/delivery zone (range 01–50) or to the district/delivery zone (range 51–99). Afghanistan Postal code lookup',
        Country: 'Afghanistan',
        ISO: 'AF',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'With Finland, first two numbers are 22.',
        Country: 'Åland Islands',
        ISO: 'AX',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Introduced in 2006, gradually implemented throughout 2007.',
        Country: 'Albania',
        ISO: 'AL',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'First two as in ISO 3166-2:DZ',
        Country: 'Algeria',
        ISO: 'DZ',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'U.S. ZIP codes (range 96799)',
        Country: 'American Samoa',
        ISO: 'AS',
        Format: 'NNNNN (optionally NNNNN-NNNN or NNNNN-NNNNNN)',
        Regex: /^\d{5}(-{1}\d{4,6})$/
    }, {
        Note: 'Each parish now has its own post code.',
        Country: 'Andorra',
        ISO: 'AD',
        Format: 'CCNNN',
        Regex: /^[Aa][Dd]\d{3}$/
    }, {
        Note: '',
        Country: 'Angola',
        ISO: 'AO',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Single code used for all addresses.',
        Country: 'Anguilla',
        ISO: 'AI',
        Format: 'AI-2640',
        Regex: /^[Aa][I][-][2][6][4][0]$/
    }, {
        Note: '',
        Country: 'Antigua and Barbuda',
        ISO: 'AG',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Codigo Postal Argentino (CPA), where the first A is the province code as in ISO 3166-2:AR, the four numbers are the old postal codes, the three last letters indicate a side of the block. Previously NNNN which &#10;o the minimum requirement as of 2006.',
        Country: 'Argentina',
        ISO: 'AR',
        Format: '1974-1998 NNNN; From 1999 ANNNNAAA',
        Regex: /^\d{4}|[A-Za-z]\d{4}[a-zA-Z]{3}$/
    }, {
        Note: 'Previously used NNNNNN system inherited from former Soviet Union.',
        Country: 'Armenia',
        ISO: 'AM',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Aruba',
        ISO: 'AW',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Single code used for all addresses. Part of UK system.',
        Country: 'Ascension island',
        ISO: 'AC',
        Format: 'AAAANAA one code: ASCN 1ZZ',
        Regex: /^[Aa][Ss][Cc][Nn]\s{0,1}[1][Zz][Zz]$/
    }, {
        Note: 'In general, the first digit identifies the state or territory.',
        Country: 'Australia',
        ISO: 'AU',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'The first digit denotes regions, which are partly identical to one of the nine provinces—called Bundesländer; the last the nearest post office in the area.',
        Country: 'Austria',
        ISO: 'AT',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Previously used NNNNNN system inherited from former Soviet Union.',
        Country: 'Azerbaijan',
        ISO: 'AZ',
        Format: 'CCNNNN',
        Regex: /^[Aa][Zz]\d{4}$/
    }, {
        Note: '',
        Country: 'Bahamas',
        ISO: 'BS',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Valid post code numbers are 101 to 1216 with gaps in the range. Known as block number (Arabic: رقم المجمع‎) formally. The first digit in NNN format and the first two digits in NNNN format refer to one of the 12 municipalities of the country. PO Box address doesn\'t need a block number or city name, just the PO Box number followed by the name of the country, Bahrain.',
        Country: 'Bahrain',
        ISO: 'BH',
        Format: 'NNN or NNNN',
        Regex: /^\d{3,4}$/
    }, {
        Note: '',
        Country: 'Bangladesh',
        ISO: 'BD',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Only one postal code currently assigned. 11000 applies to the General Post Office building in Cheapside, Bridgetown, to enable delivery to Barbados by global package delivery companies whose software requires a postal code.',
        Country: 'Barbados',
        ISO: 'BB',
        Format: 'CCNNNNN',
        Regex: /^[Aa][Zz]\d{5}$/
    }, {
        Note: 'Retained system inherited from former Soviet Union.',
        Country: 'Belarus',
        ISO: 'BY',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'In general, the first digit gives the province.',
        Country: 'Belgium',
        ISO: 'BE',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Belize',
        ISO: 'BZ',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Benin',
        ISO: 'BJ',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'AA NN for street addresses, AA AA for P.O. Box addresses. The second half of the postcode identifies the street delivery walk (e.g.: Hamilton HM 12) or the PO Box number range (e.g.: Hamilton HM BX). See Postal codes in Bermuda.',
        Country: 'Bermuda',
        ISO: 'BM',
        Format: 'AA NN or AA AA',
        Regex: /^[A-Za-z]{2}\s([A-Za-z]{2}|\d{2})$/
    }, {
        Note: '',
        Country: 'Bhutan',
        ISO: 'BT',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Bolivia',
        ISO: 'BO',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Bonaire, Sint Eustatius and Saba',
        ISO: 'BQ',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Bosnia and Herzegovina',
        ISO: 'BA',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Botswana',
        ISO: 'BW',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Código de Endereçamento Postal (CEP): -000 to -899 are used for streets, roads, avenues, boulevards; -900 to -959 are used for buildings with a high postal use; -960 to -969 are for promotional use; -970 to -989 are post offices and regular P.O. boxes; and -990 to -998 are used for community P.O. boxes. -999 is used for special services.',
        Country: 'Brazil',
        ISO: 'BR',
        Format: 'NNNNN-NNN (NNNNN from 1971 to 1992)',
        Regex: /^\d{5}-\d{3}$/
    }, {
        Note: 'Single code used for all addresses.',
        Country: 'British Antarctic Territory',
        ISO: '',
        Format: 'BIQQ 1ZZ',
        Regex: /^[Bb][Ii][Qq]{2}\s{0,1}[1][Zz]{2}$/
    }, {
        Note: 'UK territory, but not UK postcode.',
        Country: 'British Indian Ocean Territory',
        ISO: 'IO',
        Format: 'AAAANAA one code: BBND 1ZZ',
        Regex: /^[Bb]{2}[Nn][Dd]\s{0,1}[1][Zz]{2}$/
    }, {
        Note: 'Specifically, VG1110 through VG1160[1]',
        Country: 'British Virgin Islands',
        ISO: 'VG',
        Format: 'CCNNNN',
        Regex: /^[Vv][Gg]\d{4}$/
    }, {
        Note: '',
        Country: 'Brunei',
        ISO: 'BN',
        Format: 'AANNNN',
        Regex: /^[A-Za-z]{2}\d{4}$/
    }, {
        Note: '',
        Country: 'Bulgaria',
        ISO: 'BG',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Burkina Faso',
        ISO: 'BF',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Burundi',
        ISO: 'BI',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Cambodia',
        ISO: 'KH',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Cameroon',
        ISO: 'CM',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'The system was gradually introduced starting in April 1971 in Ottawa. The letters D, F, I, O, Q, and U are not used to avoid confusion with other letters or numbers.',
        Country: 'Canada',
        ISO: 'CA',
        Format: 'ANA NAN',
        Regex: /^(?=[^DdFfIiOoQqUu\d\s])[A-Za-z]\d(?=[^DdFfIiOoQqUu\d\s])[A-Za-z]\s{0,1}\d(?=[^DdFfIiOoQqUu\d\s])[A-Za-z]\d$/
    }, {
        Note: 'The first digit indicates the island.',
        Country: 'Cape Verde',
        ISO: 'CV',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Cayman Islands',
        ISO: 'KY',
        Format: 'CCN-NNNN',
        Regex: /^[Kk][Yy]\d[-\s]{0,1}\d{4}$/
    }, {
        Note: '',
        Country: 'Central African Republic',
        ISO: 'CF',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Chad',
        ISO: 'TD',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'May only be required for bulk mail.',
        Country: 'Chile',
        ISO: 'CL',
        Format: 'NNNNNNN (NNN-NNNN)',
        Regex: /^\d{7}\s\(\d{3}-\d{4}\)$/
    }, {
        Note: 'A postal code or youbian (邮编) in a subordinate division will have the same first two digits as its governing one (see Political divisions of China. The postal services in Macau or Hong Kong Special Administrative Regions remain separate from Mainland China, with no post code system currently used.',
        Country: 'China',
        ISO: 'CN',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Part of the Australian postal code system.',
        Country: 'Christmas Island',
        ISO: 'CX',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Part of the Australian postal code system.',
        Country: 'Cocos (Keeling) Island',
        ISO: 'CC',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'First NN = 32 departments Códigos Postales | 4-72',
        Country: 'Colombia',
        ISO: 'CO',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: '',
        Country: 'Comoros',
        ISO: 'KM',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Congo (Brazzaville)',
        ISO: 'CG',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Congo, Democratic Republic',
        ISO: 'CD',
        Format: '- no codes -',
        Regex: /^[Cc][Dd]$/
    }, {
        Note: '',
        Country: 'Cook Islands',
        ISO: 'CK',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'First codes the provinces, next two the canton, last two the district.',
        Country: 'Costa Rica',
        ISO: 'CR',
        Format: 'NNNNN (NNNN until 2007)',
        Regex: /^\d{4,5}$/
    }, {
        Note: '',
        Country: 'Côte d\'Ivoire (Ivory Coast)',
        ISO: 'CI',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Croatia',
        ISO: 'HR',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'May only be required for bulk mail. The letters CP are frequently used before the postal code. This is not a country code, but an abbreviation for \'codigo postal\' or postal code.',
        Country: 'Cuba',
        ISO: 'CU',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Curaçao',
        ISO: 'CW',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Post code system covers whole island, but not used in Northern Cyprus where \'Mersin 10, Turkey\' is used instead.',
        Country: 'Cyprus',
        ISO: 'CY',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'With Slovak Republic, Poštovní směrovací číslo (PSČ) - postal routing number.',
        Country: 'Czech Republic',
        ISO: 'CZ',
        Format: 'NNNNN (NNN NN)',
        Regex: /^\d{5}\s\(\d{3}\s\d{2}\)$/
    }, {
        Note: 'Numbering follows the dispatch of postal trains from Copenhagen.[3] Also used by Greenland, e.g.: DK-3900 Nuuk.',
        Country: 'Denmark',
        ISO: 'DK',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Djibouti',
        ISO: 'DJ',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Dominica',
        ISO: 'DM',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Dominican Republic',
        ISO: 'DO',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'No postal code system in use since Indonesian withdrawal in 1999.',
        Country: 'East Timor',
        ISO: 'TL',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Ecuador',
        ISO: 'EC',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Used for all inbound mail to El Salvador. The postal office then distributes the mail internally depending on their destination.',
        Country: 'El Salvador',
        ISO: 'SV',
        Format: '1101',
        Regex: /^1101$/
    }, {
        Note: '',
        Country: 'Egypt',
        ISO: 'EG',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Equatorial Guinea',
        ISO: 'GQ',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Eritrea',
        ISO: 'ER',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Estonia',
        ISO: 'EE',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'The code is only used on a trial basis for Addis Ababa addresses.',
        Country: 'Ethiopia',
        ISO: 'ET',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'UK territory, but not UK postcode',
        Country: 'Falkland Islands',
        ISO: 'FK',
        Format: 'AAAANAA one code: FIQQ 1ZZ',
        Regex: /^[Ff][Ii][Qq]{2}\s{0,1}[1][Zz]{2}$/
    }, {
        Note: 'Self-governing territory within the Kingdom of Denmark, but not Danish postcode.',
        Country: 'Faroe Islands',
        ISO: 'FO',
        Format: 'NNN',
        Regex: /^\d{3}$/
    }, {
        Note: '',
        Country: 'Fiji',
        ISO: 'FJ',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'A lower first digit indicates a place in south (for example 00100 Helsinki), a higher indicates a place further to north (99800 in Ivalo). The last digit is usually 0, except for postal codes for PO Box number ranges, in which case it is 1. Country code for Finland: \'FI\'. In the Åland Islands, the postal code is prefixed with \'AX\', not \'FI\'. Some postal codes for rural settlements may end with 5, and there are some unique postal codes for large companies and institutions, e.g. 00014 HELSINGIN YLIOPISTO (university), 00102 EDUSKUNTA (parliament), 00020 NORDEA (a major Scandinavian bank).',
        Country: 'Finland',
        ISO: 'FI',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'The first two digits give the département number, while in Paris, Lyon and Marseille, the last two digits of the postal code indicates the arrondissement. Both of the 2 corsican départements use \'20\' as the first two digits. Also used by French overseas departments and territories. Monaco is also part of the French postal code system, but the country code MC- is used for Monegasque addresses.',
        Country: 'France',
        ISO: 'FR',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Overseas Department of France. French codes used. Range 97300 - 97390.',
        Country: 'French Guiana',
        ISO: 'GF',
        Format: '973NN',
        Regex: /^973\d{2}$/
    }, {
        Note: 'Overseas Department of France. French codes used. Range 98700 - 98790.',
        Country: 'French Polynesia',
        ISO: 'PF',
        Format: '987NN',
        Regex: /^987\d{2}$/
    }, {
        Note: 'French codes in the 98400 range have been reserved.',
        Country: 'French Southern and Antarctic Territories',
        ISO: 'TF',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Two digit postal zone goes after city name.',
        Country: 'Gabon',
        ISO: 'GA',
        Format: 'NN [city name] NN',
        Regex: /^\d{2}\s[a-zA-Z-_ ]\s\d{2}$/
    }, {
        Note: '',
        Country: 'Gambia',
        ISO: 'GM',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Georgia',
        ISO: 'GE',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Postleitzahl (PLZ)',
        Country: 'Germany',
        ISO: 'DE',
        Format: 'NN',
        Regex: /^\d{2}$/
    }, {
        Note: 'Postleitzahl (PLZ)',
        Country: 'Germany',
        ISO: 'DE',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Postleitzahl (PLZ), introduced after the German reunification. Between 1989 and 1993 the old separate 4-digit postal codes of former West- and East-Germany were distinguished by preceding \'W-\' or \'O-\' (\'Ost\' for East).',
        Country: 'Germany',
        ISO: 'DE',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '[citation needed]',
        Country: 'Ghana',
        ISO: 'GH',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Single code used for all addresses.',
        Country: 'Gibraltar',
        ISO: 'GI',
        Format: 'GX11 1AA',
        Regex: /^[Gg][Xx][1]{2}\s{0,1}[1][Aa]{2}$/
    }, {
        Note: '',
        Country: 'Greece',
        ISO: 'GR',
        Format: 'NNN NN',
        Regex: /^\d{3}\s{0,1}\d{2}$/
    }, {
        Note: 'Part of the Danish postal code system.',
        Country: 'Greenland',
        ISO: 'GL',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Grenada',
        ISO: 'GD',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Overseas Department of France. French codes used. Range 97100 - 97190.',
        Country: 'Guadeloupe',
        ISO: 'GP',
        Format: '971NN',
        Regex: /^971\d{2}$/
    }, {
        Note: 'U.S. ZIP codes. Range 96910 - 96932.',
        Country: 'Guam',
        ISO: 'GU',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'The first two numbers identify the department, the third number the route and the last two the office.',
        Country: 'Guatemala',
        ISO: 'GT',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'UK-format postcode (first two letters are always GY not GG)',
        Country: 'Guernsey',
        ISO: 'GG',
        Format: 'AAN NAA, AANN NAA',
        Regex: /^[A-Za-z]{2}\d\s{0,1}\d[A-Za-z]{2}$/
    }, {
        Note: '',
        Country: 'Guinea',
        ISO: 'GN',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Guinea Bissau',
        ISO: 'GW',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Guyana',
        ISO: 'GY',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Haiti',
        ISO: 'HT',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Part of the Australian postcode system.',
        Country: 'Heard and McDonald Islands',
        ISO: 'HM',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Honduras',
        ISO: 'HN',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '[1] The dummy postal code of Hong Kong is 999077 but it is unnecessary in fact',
        Country: 'Hong Kong',
        ISO: 'HK',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'The code defines an area, usually one code per settlement except the six largest towns. One code can identify more (usually) small settlements as well.',
        Country: 'Hungary',
        ISO: 'HU',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Iceland',
        ISO: 'IS',
        Format: 'NNN',
        Regex: /^\d{3}$/
    }, {
        Note: 'Postal Index Number (PIN)',
        Country: 'India',
        ISO: 'IN',
        Format: 'NNNNNN,&#10;NNN NNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Kode Pos. Included East Timor (ranges 88xxx and 89xxx) until 1999, no longer used. For Indonesia postal code information visit [2]',
        Country: 'Indonesia',
        ISO: 'ID',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '(Persian: کد پستی)',
        Country: 'Iran',
        ISO: 'IR',
        Format: 'NNNNN-NNNNN',
        Regex: /^\d{5}-\d{5}$/
    }, {
        Note: '',
        Country: 'Iraq',
        ISO: 'IQ',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Currently no postal codes; however, Dublin is divided into postal districts on syntax Dublin 9. A national post code system is planned. See also Republic of Ireland postal addresses.',
        Country: 'Ireland',
        ISO: 'IE',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'UK-format postcode. The first two letters are always IM.',
        Country: 'Isle of Man',
        ISO: 'IM',
        Format: 'CCN NAA, CCNN NAA',
        Regex: /^[Ii[Mm]\d{1,2}\s\d\[A-Z]{2}$/
    }, {
        Note: 'Postcode is always written BEFORE the city/place name, i.e. to the Right in Hebrew or Arabic script - to the Left in Latin script.  This also allows the legacy postal code version (even though deprecated) since it\'s still in high use.',
        Country: 'Israel',
        ISO: 'IL',
        Format: 'NNNNNNN, NNNNN',
        Regex: /^\b\d{5}(\d{2})?$/
    }, {
        Note: 'Codice di Avviamento Postale (CAP). Also used by San Marino and Vatican City. First two digits identify province with some exceptions, because there are more than 100 provinces.',
        Country: 'Italy',
        ISO: 'IT',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Jamaica currently has no national postal code system, except for Kingston and Lower St. Andrew, which are divided into postal districts numbered 1-20[4] &#10;Before the 2007 suspension, the first two letters of a national post code were always \'JM\' (for Jamaica) while the third was for one of the four zones (A-D) into which the island was divided. The last two letters were for the parish, while the two digits were for the local post office.[5]',
        Country: 'Jamaica',
        ISO: 'JM',
        Format: 'Before suspension: CCAAANN &#10;After suspension: NN',
        Regex: /^\d{2}$/
    }, {
        Note: 'See also Japanese addressing system.',
        Country: 'Japan',
        ISO: 'JP',
        Format: 'NNNNNNN (NNN-NNNN)',
        Regex: /^\d{7}\s\(\d{3}-\d{4}\)$/
    }, {
        Note: 'UK-format postcode.',
        Country: 'Jersey',
        ISO: 'JE',
        Format: 'CCN NAA',
        Regex: /^[Jj][Ee]\d\s{0,1}\d[A-Za-z]{2}$/
    }, {
        Note: 'Deliveries to PO Boxes only.',
        Country: 'Jordan',
        ISO: 'JO',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '[6]',
        Country: 'Kazakhstan',
        ISO: 'KZ',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Deliveries to PO Boxes only. The postal code refers to the post office at which the receiver\'s P. O. Box is located.',
        Country: 'Kenya',
        ISO: 'KE',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Kiribati',
        ISO: 'KI',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Korea, North',
        ISO: 'KP',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Korea, South',
        ISO: 'KR',
        Format: 'NNNNNN (NNN-NNN)(1988~2015)',
        Regex: /^\d{6}\s\(\d{3}-\d{3}\)$/
    }, {
        Note: 'A separate postal code for Kosovo was introduced by the UNMIK postal administration in 2004. Serbian postcodes are still widely used in the Serbian enclaves. No country code has been assigned.',
        Country: 'Kosovo',
        ISO: 'XK',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'The first two digits represent the sector and the last three digits represents the post office.',
        Country: 'Kuwait',
        ISO: 'KW',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Kyrgyzstan',
        ISO: 'KG',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: '',
        Country: 'Latvia',
        ISO: 'LV',
        Format: 'LV-NNNN',
        Regex: /^[Ll][Vv][- ]{0,1}\d{4}$/
    }, {
        Note: '',
        Country: 'Laos',
        ISO: 'LA',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'The first four digits represent the region or postal zone,the last four digits represent the building see also Lebanon Postal code website.',
        Country: 'Lebanon',
        ISO: 'LB',
        Format: 'NNNN NNNN',
        Regex: /^\d{4}\s{0,1}\d{4}$/
    }, {
        Note: '',
        Country: 'Lesotho',
        ISO: 'LS',
        Format: 'NNN',
        Regex: /^\d{3}$/
    }, {
        Note: 'Two digit postal zone after city name.',
        Country: 'Liberia',
        ISO: 'LR',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Libya',
        ISO: 'LY',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'With Switzerland, ordered from west to east. Range 9485 - 9498.',
        Country: 'Liechtenstein',
        ISO: 'LI',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'References: http://www.post.lt/en/help/postal-code-search. Previously 9999 which was actually the old Soviet 999999 format code with the first 2 digits dropped.',
        Country: 'Lithuania',
        ISO: 'LT',
        Format: 'LT-NNNNN',
        Regex: /^[Ll][Tt][- ]{0,1}\d{5}$/
    }, {
        Note: 'References: http://www.upu.int/post_code/en/countries/LUX.pdf',
        Country: 'Luxembourg',
        ISO: 'LU',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '[2]',
        Country: 'Macau',
        ISO: 'MO',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Macedonia',
        ISO: 'MK',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Madagascar',
        ISO: 'MG',
        Format: 'NNN',
        Regex: /^\d{3}$/
    }, {
        Note: '',
        Country: 'Malawi',
        ISO: 'MW',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Maldives',
        ISO: 'MV',
        Format: 'NNNN, NNNNN',
        Regex: /^\d{4,5}$/
    }, {
        Note: '',
        Country: 'Malaysia',
        ISO: 'MY',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Mali',
        ISO: 'ML',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Kodiċi Postali',
        Country: 'Malta',
        ISO: 'MT',
        Format: 'AAANNNN (AAA NNNN)',
        Regex: /^[A-Za-z]{3}\s{0,1}\d{4}$/
    }, {
        Note: 'U.S. ZIP codes. Range 96960 - 96970.',
        Country: 'Marshall Islands',
        ISO: 'MH',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Mauritania',
        ISO: 'MR',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Mauritius',
        ISO: 'MU',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Overseas Department of France. French codes used. Range 97200 - 97290.',
        Country: 'Martinique',
        ISO: 'MQ',
        Format: '972NN',
        Regex: /^972\d{2}$/
    }, {
        Note: 'Overseas Department of France. French codes used. Range 97600 - 97690.',
        Country: 'Mayotte',
        ISO: 'YT',
        Format: '976NN',
        Regex: /^976\d{2}$/
    }, {
        Note: 'US ZIP Code. Range 96941 - 96944.',
        Country: 'Micronesia',
        ISO: 'FM',
        Format: 'NNNNN or NNNNN-NNNN',
        Regex: /^\d{5}(-{1}\d{4})$/
    }, {
        Note: 'The first two digits identify the state (or a part thereof), except for Nos. 00 to 16, which indicate delegaciones (boroughs) of the Federal District (Mexico City).',
        Country: 'Mexico',
        ISO: 'MX',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'U.S. ZIP codes. Range 96941 - 96944.',
        Country: 'Micronesia',
        ISO: 'FM',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Moldova',
        ISO: 'MD',
        Format: 'CCNNNN (CC-NNNN)',
        Regex: /^[Mm][Dd][- ]{0,1}\d{4}$/
    }, {
        Note: 'Uses the French Postal System, but with an \'MC\' Prefix for Monaco.',
        Country: 'Monaco',
        ISO: 'MC',
        Format: '980NN',
        Regex: /^980\d{2}$/
    }, {
        Note: 'First digit: region / zone&#10;Second digit: province / district&#10;Last three digits: locality / delivery block[7]',
        Country: 'Mongolia',
        ISO: 'MN',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Montenegro',
        ISO: 'ME',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Montserrat',
        ISO: 'MS',
        Format: 'MSR 1110-1350',
        Regex: /^[Mm][Ss][Rr]\s{0,1}\d{4}$/
    }, {
        Note: '',
        Country: 'Morocco',
        ISO: 'MA',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Mozambique',
        ISO: 'MZ',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Myanmar',
        ISO: 'MM',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Postal Code ranges from 9000-9299 (NOTE: mainly 9000 is used)',
        Country: 'Namibia',
        ISO: 'NA',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Nauru',
        ISO: 'NR',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Nepal',
        ISO: 'NP',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'The combination of the postcode and the housenumber gives a unique identifier of the address.',
        Country: 'Netherlands',
        ISO: 'NL',
        Format: 'NNNN AA',
        Regex: /^\d{4}\s{0,1}[A-Za-z]{2}$/
    }, {
        Note: 'Overseas Department of France. French codes used. Range 98800 - 98890.',
        Country: 'New Caledonia',
        ISO: 'NC',
        Format: '988NN',
        Regex: /^988\d{2}$/
    }, {
        Note: 'Postcodes were originally intended for bulk mailing and were not needed for addressing individual items. However, new post codes for general use were phased in from June 2006 and came into force by July 2008.',
        Country: 'New Zealand',
        ISO: 'NZ',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Nicaragua',
        ISO: 'NI',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Niger',
        ISO: 'NE',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Nigeria',
        ISO: 'NG',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: '',
        Country: 'Niue',
        ISO: 'NU',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Part of the Australian postal code system.',
        Country: 'Norfolk Island',
        ISO: 'NF',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'U.S. ZIP codes. Range 96950 - 96952.',
        Country: 'Northern Mariana Islands',
        ISO: 'MP',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'From south to north',
        Country: 'Norway',
        ISO: 'NO',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Deliveries to P.O. Boxes only.',
        Country: 'Oman',
        ISO: 'OM',
        Format: 'NNN',
        Regex: /^\d{3}$/
    }, {
        Note: 'Pakistan postal codes list',
        Country: 'Pakistan',
        ISO: 'PK',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'U.S. ZIP codes. All locations 96940.',
        Country: 'Palau',
        ISO: 'PW',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Panama',
        ISO: 'PA',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: '',
        Country: 'Papua New Guinea',
        ISO: 'PG',
        Format: 'NNN',
        Regex: /^\d{3}$/
    }, {
        Note: '',
        Country: 'Paraguay',
        ISO: 'PY',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Peru',
        ISO: 'PE',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Philippines',
        ISO: 'PH',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'UK territory, but not UK postcode',
        Country: 'Pitcairn Islands',
        ISO: 'PN',
        Format: 'AAAANAA one code: PCRN 1ZZ',
        Regex: /^[Pp][Cc][Rr][Nn]\s{0,1}[1][Zz]{2}$/
    }, {
        Note: '',
        Country: 'Poland',
        ISO: 'PL',
        Format: 'NNNNN (NN-NNN)',
        Regex: /^\d{2}[- ]{0,1}\d{3}$/
    }, {
        Note: '',
        Country: 'Portugal',
        ISO: 'PT',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: '',
        Country: 'Portugal',
        ISO: 'PT',
        Format: 'NNNN-NNN (NNNN NNN)',
        Regex: /^\d{4}[- ]{0,1}\d{3}$/
    }, {
        Note: 'U.S. ZIP codes. ZIP codes 006XX for NW PR, 007XX for SE PR, in which XX designates the town or post office and 009XX for the San Juan Metropolitan Area, in which XX designates the area or borough of San Juan. The last four digits identify an area within the post office. For example 00716-2604: 00716-for the east section of the city of Ponce and 2604 for Aceitillo St. in the neighborhood of Los Caobos. US Post office is changing the PR address format to the American one: 1234 No Name Avenue, San Juan, PR 00901.',
        Country: 'Puerto Rico',
        ISO: 'PR',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Qatar',
        ISO: 'QA',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Overseas Department of France. French codes used. Range 97400 - 97490.',
        Country: 'Réunion',
        ISO: 'RE',
        Format: '974NN',
        Regex: /^974\d{2}$/
    }, {
        Note: 'Previously 99999 in Bucharest and 9999 in rest of country.',
        Country: 'Romania',
        ISO: 'RO',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Placed on a line of its own.',
        Country: 'Russia',
        ISO: 'RU',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Overseas Collectivity of France. French codes used.',
        Country: 'Saint Barthélemy',
        ISO: 'BL',
        Format: '97133',
        Regex: /^97133$/
    }, {
        Note: 'Single code used for all addresses.',
        Country: 'Saint Helena',
        ISO: 'SH',
        Format: 'STHL 1ZZ',
        Regex: /^[Ss][Tt][Hh][Ll]\s{0,1}[1][Zz]{2}$/
    }, {
        Note: '',
        Country: 'Saint Kitts and Nevis',
        ISO: 'KN',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Saint Lucia',
        ISO: 'LC',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Overseas Collectivity of France. French codes used.',
        Country: 'Saint Martin',
        ISO: 'MF',
        Format: '97150',
        Regex: /^97150$/
    }, {
        Note: 'Overseas Collectivity of France. French codes used.',
        Country: 'Saint Pierre and Miquelon',
        ISO: 'PM',
        Format: '97500',
        Regex: /^97500$/
    }, {
        Note: '',
        Country: 'Saint Vincent and the Grenadines',
        ISO: 'VC',
        Format: 'CCNNNN',
        Regex: /^[Vv][Cc]\d{4}$/
    }, {
        Note: 'With Italy, uses a five-digit numeric CAP of Emilia Romagna. Range 47890 and 47899',
        Country: 'San Marino',
        ISO: 'SM',
        Format: '4789N',
        Regex: /^4789\d$/
    }, {
        Note: '',
        Country: 'Sao Tome and Principe',
        ISO: 'ST',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'A complete 13-digit code has 5-digit number representing region, sector, city, and zone; 4-digit X between 2000 and 5999; 4-digit Y between 6000 and 9999 [3]. Digits of 5-digit code may represent postal region, sector, branch, section, and block respectively [4].',
        Country: 'Saudi Arabia',
        ISO: 'SA',
        Format: 'NNNNN for PO Boxes. NNNNN-NNNN for home delivery.',
        Regex: /^\d{5}(-{1}\d{4})?$/
    }, {
        Note: 'The letters CP or C.P. are often written in front of the postcode. This is not a country code, but simply an abbreviation for \'code postal\'.',
        Country: 'Senegal',
        ISO: 'SN',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Poštanski adresni kod (PAK)',
        Country: 'Serbia',
        ISO: 'RS',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Serbia',
        ISO: 'RS',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Seychelles',
        ISO: 'SC',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Sint Maarten',
        ISO: 'SX',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Sierra Leone',
        ISO: 'SL',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Singapore',
        ISO: 'SG',
        Format: 'NN',
        Regex: /^\d{2}$/
    }, {
        Note: '',
        Country: 'Singapore',
        ISO: 'SG',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Each building has its own unique postcode.',
        Country: 'Singapore',
        ISO: 'SG',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'with Czech Republic from west to east, Poštové smerovacie číslo (PSČ) - postal routing number.',
        Country: 'Slovakia',
        ISO: 'SK',
        Format: 'NNNNN (NNN NN)',
        Regex: /^\d{5}\s\(\d{3}\s\d{2}\)$/
    }, {
        Note: '',
        Country: 'Slovenia',
        ISO: 'SI',
        Format: 'NNNN (CC-NNNN)',
        Regex: /^([Ss][Ii][- ]{0,1}){0,1}\d{4}$/
    }, {
        Note: '',
        Country: 'Solomon Islands',
        ISO: 'SB',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'A 5 digit code has been publicized, but never taken into use.',
        Country: 'Somalia',
        ISO: 'SO',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Postal codes are allocated to individual Post Office branches, some have two codes to differentiate between P.O. Boxes and street delivery addresses. Included Namibia (ranges 9000-9299) until 1992, no longer used.',
        Country: 'South Africa',
        ISO: 'ZA',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'One code for all addresses.',
        Country: 'South Georgia and the South Sandwich Islands',
        ISO: 'GS',
        Format: 'SIQQ 1ZZ',
        Regex: /^[Ss][Ii][Qq]{2}\s{0,1}[1][Zz]{2}$/
    }, {
        Note: '',
        Country: 'South Korea',
        ISO: 'KR',
        Format: 'NNNNNN (NNN-NNN)',
        Regex: /^\d{6}\s\(\d{3}-\d{3}\)$/
    }, {
        Note: 'First two indicate the province, range 01-52',
        Country: 'Spain',
        ISO: 'ES',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Reference: http://mohanjith.net/ZIPLook/ Incorporates Colombo postal districts, e.g.: Colombo 1 is \'00100\'. You can search for specific postal codes here.',
        Country: 'Sri Lanka',
        ISO: 'LK',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Sudan',
        ISO: 'SD',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Suriname',
        ISO: 'SR',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'The letter identifies one of the country\'s four districts.',
        Country: 'Swaziland',
        ISO: 'SZ',
        Format: 'ANNN',
        Regex: /^[A-Za-z]\d{3}$/
    }, {
        Note: '',
        Country: 'Sweden',
        ISO: 'SE',
        Format: 'NNNNN (NNN NN)',
        Regex: /^\d{3}\s*\d{2}$/
    }, {
        Note: 'With Liechtenstein, ordered from west to east. In Geneva and other big cities, like Basel, Bern, Zurich, there may be one or two digits after the name of the city when the generic City code (1211) is used instead of the area-specific code (1201, 1202...), e.g.: 1211 Geneva 13. The digit identifies the post office. This addressing is generally used for P.O. box deliveries.',
        Country: 'Switzerland',
        ISO: 'CH',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'Norway postal codes',
        Country: 'Svalbard and Jan Mayen',
        ISO: 'SJ',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'A 4-digit system has been announced. Status unknown.',
        Country: 'Syria',
        ISO: 'SY',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'The first three digits of the postal code are required; the last two digits are optional. Codes are known as youdi quhao (郵遞區號), and are also assigned to Senkaku Islands (Diaoyutai), though Japanese-administered, the Pratas Islands and the Spratly Islands. See List of postal codes in Taiwan.',
        Country: 'Taiwan',
        ISO: 'TW',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Retained system from former Soviet Union.',
        Country: 'Tajikistan',
        ISO: 'TJ',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: '',
        Country: 'Tanzania',
        ISO: 'TZ',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'The first two specify the province, numbers as in ISO 3166-2:TH, the third and fourth digits specify a district (amphoe)',
        Country: 'Thailand',
        ISO: 'TH',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'Togo',
        ISO: 'TG',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Tokelau',
        ISO: 'TK',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Tonga',
        ISO: 'TO',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'First two digits specify a postal district (one of 72), next two digits a carrier route, last two digits a building or zone along that route',
        Country: 'Trinidad and Tobago',
        ISO: 'TT',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Single code used for all addresses.',
        Country: 'Tristan da Cunha',
        ISO: 'SH',
        Format: 'TDCU 1ZZ',
        Regex: /^[Tt][Dd][Cc][Uu]\s{0,1}[1][Zz]{2}$/
    }, {
        Note: '',
        Country: 'Tunisia',
        ISO: 'TN',
        Format: 'NNNN',
        Regex: /^\d{4}$/
    }, {
        Note: 'First two digits are the city numbers.[8]',
        Country: 'Turkey',
        ISO: 'TR',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Retained system from former Soviet Union.',
        Country: 'Turkmenistan',
        ISO: 'TM',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Single code used for all addresses.',
        Country: 'Turks and Caicos Islands',
        ISO: 'TC',
        Format: 'TKCA 1ZZ',
        Regex: /^[Tt][Kk][Cc][Aa]\s{0,1}[1][Zz]{2}$/
    }, {
        Note: '',
        Country: 'Tuvalu',
        ISO: 'TV',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Uganda',
        ISO: 'UG',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Ukraine',
        ISO: 'UA',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: '',
        Country: 'United Arab Emirates',
        ISO: 'AE',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Known as the postcode. The first letter(s) indicate the postal area, such as the town or part of London. Placed on a separate line below the city (or county, if used). The UK postcode is made up of two parts separated by a space. These are known as the outward postcode and the inward postcode. The outward postcode is always one of the following formats: AN, ANN, AAN, AANN, ANA, AANA, AAA. The inward postcode is always formatted as NAA. A valid inward postcode never contains the letters: C, I, K, M, O or V. The British Forces Post Office has a different system, but as of 2012 has also adopted UK-style postcodes that begin with \'BF1\' for electronic compatibility.',
        Country: 'United Kingdom',
        ISO: 'GB',
        Format: 'A(A)N(A/N)NAA (A[A]N[A/N] NAA)',
        Regex: /^[A-Z]{1,2}[0-9R][0-9A-Z]?\s*[0-9][A-Z-[CIKMOV]]{2}$/
    }, {
        Note: 'Known as the ZIP Code with five digits 99999* or the ZIP+4 Code with nine digits 99999-9999* (while the minimum requirement is the first five digits, the U.S. Postal Service encourages everyone to use all nine). Also used by the former US Pacific Territories: Federated States of Micronesia; Palau; and the Marshall Islands, as well as in current US territories American Samoa, Guam, Northern Mariana Islands, Puerto Rico, and the United States Virgin Islands. An individual delivery point may be represented as an 11-digit number, but these are usually represented by Intelligent Mail barcode or formerly POSTNET bar code.',
        Country: 'United States',
        ISO: 'US',
        Format: 'NNNNN (optionally NNNNN-NNNN)',
        Regex: /^\b\d{5}\b(?:[- ]{1}\d{4})?$/
    }, {
        Note: '',
        Country: 'Uruguay',
        ISO: 'UY',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'U.S. ZIP codes. Range 00801 - 00851.',
        Country: 'U.S. Virgin Islands',
        ISO: 'VI',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'Почтовые индексы',
        Country: 'Uzbekistan',
        ISO: 'UZ',
        Format: 'NNN NNN',
        Regex: /^\d{3} \d{3}$/
    }, {
        Note: '',
        Country: 'Vanuatu',
        ISO: 'VU',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: 'Single code used for all addresses. Part of the Italian postal code system.',
        Country: 'Vatican',
        ISO: 'VA',
        Format: '120',
        Regex: /^120$/
    }, {
        Note: '',
        Country: 'Venezuela',
        ISO: 'VE',
        Format: 'NNNN or NNNN A',
        Regex: /^\d{4}(\s[a-zA-Z]{1})?$/
    }, {
        Note: 'First two indicate a province.',
        Country: 'Vietnam',
        ISO: 'VN',
        Format: 'NNNNNN',
        Regex: /^\d{6}$/
    }, {
        Note: 'Overseas Department of France. French codes used. Range 98600 - 98690.',
        Country: 'Wallis and Futuna',
        ISO: 'WF',
        Format: '986NN',
        Regex: /^986\d{2}$/
    }, {
        Note: 'System for Sana\'a Governorate using geocoding \'عنواني\' based on the OpenPostcode algorithm is inaugurated in 2014.[9]',
        Country: 'Yemen',
        ISO: 'YE',
        Format: '- no codes -',
        Regex: ''
    }, {
        Note: '',
        Country: 'Zambia',
        ISO: 'ZM',
        Format: 'NNNNN',
        Regex: /^\d{5}$/
    }, {
        Note: 'System is being planned.',
        Country: 'Zimbabwe',
        ISO: 'ZW',
        Format: '- no codes -',
        Regex: ''
    }];

let BROWSER_FAVICONS_CONFIG = new InjectionToken('Favicons Configuration');
/**
 * Abstract class that acts as both interface for implementation and as the dependency-injection
 */
class FaviconsService {
}
/**
 * Provide the browser-oriented implementation of the Favicons class
 */
class BrowserFavicons {
    /**
     * initialize the Favicons service.
     */
    constructor(config) {
        this.elementId = 'favicons-service-injected-node';
        this.icons = Object.assign(Object.create(null), config.icons);
        this.useCacheBusting = (config.cacheBusting || false);
        // Since the document may have a static favicon definition, we want to strip out
        // any exisitng elements that are attempting to define a favicon. This way, there
        // is only one favicon element on the page at a time.
        this.removeExternalLinkElements();
    }
    /**
     * Active favicon with given identifier
     *
     * @param name favicon name
     */
    activate(name) {
        if (!this.icons[name]) {
            throw (new Error(`Favicon for [ ${name} ] not found.`));
        }
        this.setNode(this.icons[name].type, this.icons[name].href);
    }
    /**
     * Activate the default favicon (with isDefault set to True)
     */
    reset() {
        for (const name of Object.keys(this.icons)) {
            const icon = this.icons[name];
            if (icon.isDefault) {
                this.setNode(icon.type, icon.href);
                return;
            }
        }
        // If we made it this far, none of the favicons were flagged as default. In that
        // case, let's just remove the favicon node altogether.
        this.removeNode();
    }
    /**
     * Inject the favicon element into the document header
     *
     * @param type type of icon (.png, .jpeg, etc)
     *
     * @param href location
     */
    addNode(type, href) {
        const linkElement = document.createElement('link');
        linkElement.setAttribute('id', this.elementId);
        linkElement.setAttribute('rel', 'icon');
        linkElement.setAttribute('type', type);
        linkElement.setAttribute('href', href);
        document.head.appendChild(linkElement);
    }
    /**
     * Return an augmented HREF value with a cache-busting query-string parameter
     *
     * @param href location
     */
    cacheBustHref(href) {
        const augmentedHref = (href.indexOf('?') === -1)
            ? `${href}?faviconCacheBust=${Date.now()}`
            : `${href}&faviconCacheBust=${Date.now()}`;
        return (augmentedHref);
    }
    /**
     * I remove any favicon nodes that are not controlled by this service
     */
    removeExternalLinkElements() {
        // tslint:disable-next-line:prefer-const
        let linkElements = document.querySelectorAll(`link[ rel ~= 'icon' i]`);
        for (const linkElement of Array.from(linkElements)) {
            linkElement.parentNode.removeChild(linkElement);
        }
    }
    /**
     * Remove the favicon node from the document header
     */
    removeNode() {
        const linkElement = document.head.querySelector('#' + this.elementId);
        if (linkElement) {
            document.head.removeChild(linkElement);
        }
    }
    /**
     * Remove existing favicon node and inject a new favicon node with the give
     * elemet settings
     *
     * @param type type of favicon
     * @param href location
     */
    setNode(type, href) {
        const augmentedHref = this.useCacheBusting
            ? this.cacheBustHref(href)
            : href;
        this.removeNode();
        this.addNode(type, augmentedHref);
    }
}
BrowserFavicons.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [BROWSER_FAVICONS_CONFIG,] }] }
];

// @dynamic
class ConvertToBase64Util {
    /**
     *
     * @param event either type File or Blob
     *
     * Converts to Base64 and returns an observable of Base64Model
     */
    static GetBase64(files) {
        const base64Observable = new Subject();
        const baseArray = [];
        let file;
        for (file of files) {
            const reader = new FileReader();
            // tslint:disable-next-line:no-shadowed-variable
            ((file) => {
                reader.onload = () => {
                    baseArray.push(new Base64Model(reader.result.toString(), file));
                    if (baseArray.length === files.length) {
                        base64Observable.next(baseArray);
                    }
                };
            })(file);
            reader.readAsDataURL(file['file'].rawFile);
            reader.onerror = (error) => {
                console.error('Error: ', error);
            };
        }
        return base64Observable;
    }
}

class ToggleThemeUtil {
    /**
     * Toggle fathym themes // could probably add themes to local storage
     *
     * @param classList classlist that holds the theme class
     *
     * @param val theme to change to
     */
    static Toggle(classList, val) {
        let theme;
        if (classList.contains('arctic-theme')) {
            classList.remove('arctic-theme');
        }
        if (classList.contains('contrast-theme')) {
            classList.remove('contrast-theme');
        }
        if (classList.contains('cool-candy-theme')) {
            classList.remove('cool-candy-theme');
        }
        if (classList.contains('flipper-theme')) {
            classList.remove('flipper-theme');
        }
        if (classList.contains('ice-theme')) {
            classList.remove('ice-theme');
        }
        if (classList.contains('sea-green-theme')) {
            classList.remove('sea-green-theme');
        }
        if (classList.contains('white-mint')) {
            classList.remove('white-mint');
        }
        switch (val.toUpperCase()) {
            case 'ARCTIC-THEME':
                theme = 'arctic-theme';
                classList.add('arctic-theme');
                break;
            case 'CONTRAST-THEME':
                theme = 'contrast-theme';
                classList.add('contrast-theme');
                break;
            case 'COOL-CANDY-THEME':
                theme = 'cool-candy-theme';
                classList.add('cool-candy-theme');
                break;
            case 'FLIPPER-THEME':
                theme = 'flipper-theme';
                classList.add('flipper-theme');
                break;
            case 'ICE-THEME':
                theme = 'ice-theme';
                classList.add('ice-theme');
                break;
            case 'SEA-GREEN-THEME':
                theme = 'sea-green-theme';
                classList.add('sea-green-theme');
                break;
            case 'WHITE-MINT-THEME':
                theme = 'white-mint-theme';
                classList.add('white-mint-theme');
                break;
        }
        return theme;
    }
}

// @dynamic
/**
 * @dynamic is used because this class contains static properties
 */
/**
 * For checking data types
 */
class IsDataTypeUtil {
    /**
     * Check for string
     *
     * (new string) typeof will return object, so check for that with instanceof
     *
     * @param value value to test
     */
    static IsString(value) {
        return typeof value === 'string' || value instanceof String;
    }
    /**
     * Check for number
     *
     * More than 'number' will be returned, like: 'NaN' and 'Infinity,' so to be sure
     * a value is a number use the function isFinite()
     *
     * @param value value to test
     */
    static IsNumber(value) {
        return typeof value === 'number' && isFinite(value);
    }
    /**
     * Check for array
     *
     * An array is returned as an object, to check if it's really an array
     * its constructor can be compared to 'Array.'
     *
     * ES5 has a method for this: Array.isArray(value)
     *
     * @param value value to test
     */
    static IsArray(value) {
        return value && typeof value === 'object' && value.constructor === Array;
    }
    /**
     * Check for function
     *
     * @param value value to test
     */
    static IsFunction(value) {
        return typeof value === 'function';
    }
    /**
     * Check for Object
     *
     * A lot of items can be objects, so to test that an item can have
     * properties and be looped through, compare its constructor with 'Object.'
     *
     * @param value value to test
     */
    static IsObject(value) {
        return value && typeof value === 'object' && value.constructor === Object;
    }
    /**
     * Check for null
     *
     * @param value value to test
     */
    static IsNull(value) {
        return value === null;
    }
    /**
     * Check for undefined
     *
     * @param value value to test
     */
    static IsUndefined(value) {
        return value === 'undefined';
    }
    /**
     * Check for boolean
     *
     * @param value value to test
     */
    static IsBoolean(value) {
        return typeof value === 'boolean';
    }
    /**
     * Check for regex
     *
     * RegEx's are object, so check against its constructor
     *
     * @param value value to test
     */
    static IsRegEx(value) {
        return value && typeof value === 'object' && value.constructor === RegExp;
    }
    /**
     * Check for Date
     *
     * Date ins't really a data type in JS, so to check a Date Object use 'instanceof.'
     *
     * @param value value to test
     */
    static IsData(value) {
        return value instanceof Date;
    }
    /**
     * Check for Symbol
     *
     * New datatype in ES6
     *
     * @param value value to test
     */
    static IsSymbol(value) {
        return typeof value === 'symbol';
    }
    /**
     * Find out what kind of datatype you have
     *
     * @param value value to test
     */
    static GetDataType(value) {
        if (this.IsArray(value)) {
            return 'array';
        }
        else if (this.IsBoolean(value)) {
            return 'boolean';
        }
        else if (this.IsData(value)) {
            return 'data';
        }
        else if (this.IsFunction(value)) {
            return 'function';
        }
        else if (this.IsNull(value)) {
            return 'null';
        }
        else if (this.IsNumber(value)) {
            return 'number';
        }
        else if (this.IsObject(value)) {
            return 'object';
        }
        else if (this.IsRegEx(value)) {
            return 'regex';
        }
        else if (this.IsString(value)) {
            return 'string';
        }
        else if (this.IsSymbol(value)) {
            return 'symbol';
        }
        else if (this.IsUndefined(value)) {
            return 'undefined';
        }
        else {
            throw new Error('No Datatype found');
        }
    }
}

// @dynamic
/**
 * @dynamic is used because this class contains static properties
 */
/**
 * Flatten and Unflatten JSON
 *
 * Original Code Example: http://jsfiddle.net/crl/WSzec/154/
 */
class JSONFlattenUnflatten {
    /**
     * Flatten a standard JSON Object
     *
     * @param json JSON object to be flattened
     *
     * @param map recursive map data
     *
     * @param prefix dotnotated path
     */
    static Flatten(json, map = {}, prefix = '') {
        if (Array.isArray(json)) {
            json.forEach((itm, idx) => {
                if (IsDataTypeUtil.IsObject(json[itm]) && json[itm]) {
                    this.Flatten(json[itm], map, prefix + '[' + itm + ']');
                }
                else {
                    map[prefix + '[' + itm + ']'] = json[itm];
                }
            });
            return map;
        }
        Object.entries(json).forEach(([key, value]) => {
            if (IsDataTypeUtil.IsObject(json[key]) && json[key]) {
                this.Flatten(json[key], map, (prefix ? prefix + '.' : '') + key);
            }
            else {
                map[(prefix ? prefix + '.' : '') + key] = json[key];
            }
        });
        return map;
    }
    /**
     * Flatten JSON into a map
     *
     * @param json JSON object to be flattened
     *
     * @param map recursive map data
     *
     * @param prefix dotnotated path
     */
    static FlattenMap(json, map = new Map(), prefix = '') {
        if (Array.isArray(json)) {
            json.forEach((itm, idx) => {
                if (IsDataTypeUtil.IsObject(json[itm]) && json[itm]) {
                    this.FlattenMap(json[itm], map, prefix + '[' + itm + ']');
                }
                else {
                    map.set(prefix + '[' + itm + ']', json[itm]);
                }
            });
            return map;
        }
        Object.entries(json).forEach(([key, value]) => {
            if (IsDataTypeUtil.IsObject(json[key]) && json[key]) {
                if (Object.entries(json[key]).length === 0) {
                    return;
                }
                this.FlattenMap(json[key], map, (prefix ? prefix + '.' : '') + key);
            }
            else {
                map.set((prefix ? prefix + '.' : '') + key, json[key]);
            }
        });
        return map;
    }
    /**
     * Flatten JSON into a map
     *
     * @param json JSON object to be flattened
     *
     * @param map recursive map data
     *
     * @param prefix dotnotated path
     */
    static FlattenMapTest(json, map = new Map(), prefix = '') {
        if (Array.isArray(json)) {
            json.forEach((itm, idx) => {
                if (IsDataTypeUtil.IsObject(json[itm]) && json[itm]) {
                    this.FlattenMap(json[itm], map, prefix + '[' + itm + ']');
                }
                else {
                    map.set(prefix + '[' + itm + ']', json[itm]);
                }
            });
            return map;
        }
        Object.entries(json).forEach(([key, value]) => {
            if (IsDataTypeUtil.IsObject(json[key]) && json[key]) {
                if (Object.entries(json[key]).length === 0) {
                    return;
                }
                this.FlattenMapTest(json[key], map, (prefix ? prefix + '.' : '') + key);
            }
            else {
                map.set((prefix ? prefix + '.' : '') + key, json[key]);
            }
        });
        return map;
    }
    /**
     * Flatten JSON into a map
     *
     * @param json JSON object to be flattened
     *
     * @param map recursive map data
     *
     * @param prefix dotnotated path
     */
    static FlattenMapJSONSchemaItem(json, map = new Map(), prefix = '') {
        if (Array.isArray(json)) {
            json.forEach((itm, idx) => {
                if (IsDataTypeUtil.IsObject(json[itm]) && json[itm]) {
                    this.FlattenMap(json[itm], map, prefix + '[' + itm + ']');
                }
                else {
                    map.set(prefix + '[' + itm + ']', json[itm]);
                }
            });
            return map;
        }
        Object.entries(json).forEach(([key, value]) => {
            if (IsDataTypeUtil.IsObject(json[key]) && json[key]) {
                this.FlattenMap(json[key], map, (prefix ? prefix + '.' : '') + key);
            }
            else {
                map.set((prefix ? prefix + '.' : '') + key, json[key]);
            }
        });
        return map;
    }
    /**
     * Unflatten an already flatten JSON object
     *
     * @param map object to unflatten
     */
    static Unflatten(map) {
        const returnObject = {};
        Object.entries(map).forEach(([key, value]) => {
            const keysRegEx = /(?:\.?([^.[]+))|(\[\d+\])/g;
            let match;
            let oi = returnObject; // ref to returnObject, modified in loop below
            // tslint:disable-next-line:no-conditional-assignment
            while (match = keysRegEx.exec(key)) { // method executes a search for a match in a specified string. Returns a result array,
                const prefix = key[match.index + match[0].length];
                let keyName = match[1];
                if (!prefix) { // nothing more, store value in oi[keyName]
                    oi[keyName] = value;
                }
                else {
                    if (keyName) {
                        if (!oi.hasOwnProperty(keyName)) {
                            oi[keyName] = prefix === '[' ? [] : {};
                        }
                    }
                    else {
                        keyName = match[2].slice(1, -1);
                        if (!oi.hasOwnProperty(keyName)) {
                            oi[keyName] = prefix === '[' ? [] : {};
                        }
                    }
                }
                oi = oi[keyName];
            }
        });
        return returnObject;
    }
    /**
     * Unflatten map data
     *
     * @param map map data
     */
    static UnflattenMap(map) {
        const returnObject = {};
        map.forEach((value, key) => {
            const keysRegEx = /(?:\.?([^.[]+))|(\[\d+\])/g;
            let match;
            let oi = returnObject; // ref to returnObject, modified in loop below
            // tslint:disable-next-line:no-conditional-assignment
            while (match = keysRegEx.exec(key)) { // method executes a search for a match in a specified string. Returns a result array,
                const prefix = key[match.index + match[0].length];
                let keyName = match[1];
                if (!prefix) { // nothing more, store value in oi[keyName]
                    oi[keyName] = value;
                }
                else {
                    if (keyName) {
                        if (!oi.hasOwnProperty(keyName)) {
                            oi[keyName] = prefix === '[' ? [] : {};
                        }
                    }
                    else {
                        keyName = match[2].slice(1, -1);
                        if (!oi.hasOwnProperty(keyName)) {
                            oi[keyName] = prefix === '[' ? [] : {};
                        }
                    }
                }
                oi = oi[keyName];
            }
        });
        return returnObject;
    }
}

// @dynamic
/**
 * @dynamic is used because this class contains static properties
 */
class DotNotationUtil {
    /**
     * Using dot notation, iterate the object and change the key's value
     *
     * @param obj object to test
     *
     * @param propertyPath Dot notation path
     *
     * @param newVal Changed value
     */
    static SetValue(obj, propertyPath, newVal) {
        /**
         * split propertyPath string into an array of strings and iterate each itm with reduce
         *
         * @param acc(accumulator) returned object to iterate
         *
         * @param curr(current value) current element being processed in the array
         *
         * @param idx index position of the current element being processed in the array
         *
         * @param arr array we created with the split function above
         *
         */
        return propertyPath.split('.').reduce((acc, curr, idx, arr) => {
            if (idx === (arr.length - 1) && acc) {
                acc[curr] = newVal;
            }
            // return a new accumulator to the reduce callback(starts the loop with the next curr value)
            return acc ? acc[curr] : null;
            // inital value to use as the first argument(this is the item to start the iteration with)
        }, obj);
    }
    /**
     * Using dot notation, iterate the object and return the key's value
     *
     * @param obj object to test
     *
     * @param pathArr array of names used to drill into objects
     */
    static GetValue(obj, propertyPath) {
        /**
         * split propertyPath string into an array of strings and iterate each itm with reduce
         *
         * @param acc(accumulator) returned object to iterate
         *
         * @param curr(current value) current element being processed in the array
         *
         * @param idx index position of the current element being processed in the array
         *
         * @param arr array we created with the split function above
         *
         */
        return propertyPath.split('.').reduce((acc, curr, idx, arr) => {
            if (idx === (arr.length - 1) && acc) {
                return acc[curr];
            }
            // return a new accumulator to the reduce callback(starts the loop with the next curr value)
            return acc ? acc[curr] : null;
            // inital value to use as the first argument(this is the item to start the iteration with)
        }, obj);
    }
    static SetKeyValue(json, propertyPath, oldKey, newKey, objToChange) {
        // for top level properties
        if (propertyPath.split('.').length === 1) {
            return Object.keys(json).reduce((acc, curr, idx, arr) => curr === oldKey ? (Object.assign(Object.assign({}, acc), { [newKey]: json[oldKey] })) : (Object.assign(Object.assign({}, acc), { [curr]: json[curr] })), {});
        }
        // for nested properties
        if (propertyPath.split('.').length > 1) {
            return propertyPath.split('.').reduce((acc, curr, idx, arr) => {
                if (idx === (arr.length - 1) && acc) {
                    const idxPos = arr.indexOf(objToChange);
                    const idxPosArr = [...arr.splice(0, idxPos + 1)];
                    DotNotationUtil.SetValue(json, idxPosArr.join('.'), this.RenameKeys({ [oldKey]: newKey }, acc));
                    return json;
                    // return {
                    //   ...schema, [objToChange]: this.renameKeys({[oldKey]: newKey}, acc)
                    // };
                }
                return acc ? acc[curr] : null; // if acc, then start additional iterations with acc[curr]
            }, json); // first item to start the loop
        }
    }
    /**
     * Rename property keys
     */
    static RenameKeys(keysMap, obj) {
        return Object
            .keys(obj)
            .reduce((acc, key) => {
            const renamedObject = {
                [keysMap[key] || key]: obj[key]
            };
            return Object.assign(Object.assign({}, acc), renamedObject);
        }, {});
    }
}

// @dynamic
/**
 * @dynamic need this because there are static memebers
 */
class CountryConstants {
}
CountryConstants.Countries = [
    {
        "country": "Albania",
        "alpha2": "AL",
        "alpha3": "ALB",
        "numeric": 8,
        "latitude": 41,
        "longitude": 20
    },
    {
        "country": "Algeria",
        "alpha2": "DZ",
        "alpha3": "DZA",
        "numeric": 12,
        "latitude": 28,
        "longitude": 3
    },
    {
        "country": "American Samoa",
        "alpha2": "AS",
        "alpha3": "ASM",
        "numeric": 16,
        "latitude": -14.3333,
        "longitude": -170
    },
    {
        "country": "Andorra",
        "alpha2": "AD",
        "alpha3": "AND",
        "numeric": 20,
        "latitude": 42.5,
        "longitude": 1.6
    },
    {
        "country": "Angola",
        "alpha2": "AO",
        "alpha3": "AGO",
        "numeric": 24,
        "latitude": -12.5,
        "longitude": 18.5
    },
    {
        "country": "Anguilla",
        "alpha2": "AI",
        "alpha3": "AIA",
        "numeric": 660,
        "latitude": 18.25,
        "longitude": -63.1667
    },
    {
        "country": "Antarctica",
        "alpha2": "AQ",
        "alpha3": "ATA",
        "numeric": 10,
        "latitude": -90,
        "longitude": 0
    },
    {
        "country": "Antigua and Barbuda",
        "alpha2": "AG",
        "alpha3": "ATG",
        "numeric": 28,
        "latitude": 17.05,
        "longitude": -61.8
    },
    {
        "country": "Argentina",
        "alpha2": "AR",
        "alpha3": "ARG",
        "numeric": 32,
        "latitude": -34,
        "longitude": -64
    },
    {
        "country": "Armenia",
        "alpha2": "AM",
        "alpha3": "ARM",
        "numeric": 51,
        "latitude": 40,
        "longitude": 45
    },
    {
        "country": "Aruba",
        "alpha2": "AW",
        "alpha3": "ABW",
        "numeric": 533,
        "latitude": 12.5,
        "longitude": -69.9667
    },
    {
        "country": "Australia",
        "alpha2": "AU",
        "alpha3": "AUS",
        "numeric": 36,
        "latitude": -27,
        "longitude": 133
    },
    {
        "country": "Austria",
        "alpha2": "AT",
        "alpha3": "AUT",
        "numeric": 40,
        "latitude": 47.3333,
        "longitude": 13.3333
    },
    {
        "country": "Azerbaijan",
        "alpha2": "AZ",
        "alpha3": "AZE",
        "numeric": 31,
        "latitude": 40.5,
        "longitude": 47.5
    },
    {
        "country": "Bahamas",
        "alpha2": "BS",
        "alpha3": "BHS",
        "numeric": 44,
        "latitude": 24.25,
        "longitude": -76
    },
    {
        "country": "Bahrain",
        "alpha2": "BH",
        "alpha3": "BHR",
        "numeric": 48,
        "latitude": 26,
        "longitude": 50.55
    },
    {
        "country": "Bangladesh",
        "alpha2": "BD",
        "alpha3": "BGD",
        "numeric": 50,
        "latitude": 24,
        "longitude": 90
    },
    {
        "country": "Barbados",
        "alpha2": "BB",
        "alpha3": "BRB",
        "numeric": 52,
        "latitude": 13.1667,
        "longitude": -59.5333
    },
    {
        "country": "Belarus",
        "alpha2": "BY",
        "alpha3": "BLR",
        "numeric": 112,
        "latitude": 53,
        "longitude": 28
    },
    {
        "country": "Belgium",
        "alpha2": "BE",
        "alpha3": "BEL",
        "numeric": 56,
        "latitude": 50.8333,
        "longitude": 4
    },
    {
        "country": "Belize",
        "alpha2": "BZ",
        "alpha3": "BLZ",
        "numeric": 84,
        "latitude": 17.25,
        "longitude": -88.75
    },
    {
        "country": "Benin",
        "alpha2": "BJ",
        "alpha3": "BEN",
        "numeric": 204,
        "latitude": 9.5,
        "longitude": 2.25
    },
    {
        "country": "Bermuda",
        "alpha2": "BM",
        "alpha3": "BMU",
        "numeric": 60,
        "latitude": 32.3333,
        "longitude": -64.75
    },
    {
        "country": "Bhutan",
        "alpha2": "BT",
        "alpha3": "BTN",
        "numeric": 64,
        "latitude": 27.5,
        "longitude": 90.5
    },
    {
        "country": "Bolivia, Plurinational State of",
        "alpha2": "BO",
        "alpha3": "BOL",
        "numeric": 68,
        "latitude": -17,
        "longitude": -65
    },
    {
        "country": "Bosnia and Herzegovina",
        "alpha2": "BA",
        "alpha3": "BIH",
        "numeric": 70,
        "latitude": 44,
        "longitude": 18
    },
    {
        "country": "Botswana",
        "alpha2": "BW",
        "alpha3": "BWA",
        "numeric": 72,
        "latitude": -22,
        "longitude": 24
    },
    {
        "country": "Bouvet Island",
        "alpha2": "BV",
        "alpha3": "BVT",
        "numeric": 74,
        "latitude": -54.4333,
        "longitude": 3.4
    },
    {
        "country": "Brazil",
        "alpha2": "BR",
        "alpha3": "BRA",
        "numeric": 76,
        "latitude": -10,
        "longitude": -55
    },
    {
        "country": "British Indian Ocean Territory",
        "alpha2": "IO",
        "alpha3": "IOT",
        "numeric": 86,
        "latitude": -6,
        "longitude": 71.5
    },
    {
        "country": "Brunei Darussalam",
        "alpha2": "BN",
        "alpha3": "BRN",
        "numeric": 96,
        "latitude": 4.5,
        "longitude": 114.6667
    },
    {
        "country": "Bulgaria",
        "alpha2": "BG",
        "alpha3": "BGR",
        "numeric": 100,
        "latitude": 43,
        "longitude": 25
    },
    {
        "country": "Burkina Faso",
        "alpha2": "BF",
        "alpha3": "BFA",
        "numeric": 854,
        "latitude": 13,
        "longitude": -2
    },
    {
        "country": "Burundi",
        "alpha2": "BI",
        "alpha3": "BDI",
        "numeric": 108,
        "latitude": -3.5,
        "longitude": 30
    },
    {
        "country": "Cambodia",
        "alpha2": "KH",
        "alpha3": "KHM",
        "numeric": 116,
        "latitude": 13,
        "longitude": 105
    },
    {
        "country": "Cameroon",
        "alpha2": "CM",
        "alpha3": "CMR",
        "numeric": 120,
        "latitude": 6,
        "longitude": 12
    },
    {
        "country": "Canada",
        "alpha2": "CA",
        "alpha3": "CAN",
        "numeric": 124,
        "latitude": 60,
        "longitude": -95
    },
    {
        "country": "Cape Verde",
        "alpha2": "CV",
        "alpha3": "CPV",
        "numeric": 132,
        "latitude": 16,
        "longitude": -24
    },
    {
        "country": "Cayman Islands",
        "alpha2": "KY",
        "alpha3": "CYM",
        "numeric": 136,
        "latitude": 19.5,
        "longitude": -80.5
    },
    {
        "country": "Central African Republic",
        "alpha2": "CF",
        "alpha3": "CAF",
        "numeric": 140,
        "latitude": 7,
        "longitude": 21
    },
    {
        "country": "Chad",
        "alpha2": "TD",
        "alpha3": "TCD",
        "numeric": 148,
        "latitude": 15,
        "longitude": 19
    },
    {
        "country": "Chile",
        "alpha2": "CL",
        "alpha3": "CHL",
        "numeric": 152,
        "latitude": -30,
        "longitude": -71
    },
    {
        "country": "China",
        "alpha2": "CN",
        "alpha3": "CHN",
        "numeric": 156,
        "latitude": 35,
        "longitude": 105
    },
    {
        "country": "Christmas Island",
        "alpha2": "CX",
        "alpha3": "CXR",
        "numeric": 162,
        "latitude": -10.5,
        "longitude": 105.6667
    },
    {
        "country": "Cocos (Keeling) Islands",
        "alpha2": "CC",
        "alpha3": "CCK",
        "numeric": 166,
        "latitude": -12.5,
        "longitude": 96.8333
    },
    {
        "country": "Colombia",
        "alpha2": "CO",
        "alpha3": "COL",
        "numeric": 170,
        "latitude": 4,
        "longitude": -72
    },
    {
        "country": "Comoros",
        "alpha2": "KM",
        "alpha3": "COM",
        "numeric": 174,
        "latitude": -12.1667,
        "longitude": 44.25
    },
    {
        "country": "Congo",
        "alpha2": "CG",
        "alpha3": "COG",
        "numeric": 178,
        "latitude": -1,
        "longitude": 15
    },
    {
        "country": "Congo, the Democratic Republic of the",
        "alpha2": "CD",
        "alpha3": "COD",
        "numeric": 180,
        "latitude": 0,
        "longitude": 25
    },
    {
        "country": "Cook Islands",
        "alpha2": "CK",
        "alpha3": "COK",
        "numeric": 184,
        "latitude": -21.2333,
        "longitude": -159.7667
    },
    {
        "country": "Costa Rica",
        "alpha2": "CR",
        "alpha3": "CRI",
        "numeric": 188,
        "latitude": 10,
        "longitude": -84
    },
    {
        "country": "Côte d'Ivoire",
        "alpha2": "CI",
        "alpha3": "CIV",
        "numeric": 384,
        "latitude": 8,
        "longitude": -5
    },
    {
        "country": "Croatia",
        "alpha2": "HR",
        "alpha3": "HRV",
        "numeric": 191,
        "latitude": 45.1667,
        "longitude": 15.5
    },
    {
        "country": "Cuba",
        "alpha2": "CU",
        "alpha3": "CUB",
        "numeric": 192,
        "latitude": 21.5,
        "longitude": -80
    },
    {
        "country": "Cyprus",
        "alpha2": "CY",
        "alpha3": "CYP",
        "numeric": 196,
        "latitude": 35,
        "longitude": 33
    },
    {
        "country": "Czech Republic",
        "alpha2": "CZ",
        "alpha3": "CZE",
        "numeric": 203,
        "latitude": 49.75,
        "longitude": 15.5
    },
    {
        "country": "Denmark",
        "alpha2": "DK",
        "alpha3": "DNK",
        "numeric": 208,
        "latitude": 56,
        "longitude": 10
    },
    {
        "country": "Djibouti",
        "alpha2": "DJ",
        "alpha3": "DJI",
        "numeric": 262,
        "latitude": 11.5,
        "longitude": 43
    },
    {
        "country": "Dominica",
        "alpha2": "DM",
        "alpha3": "DMA",
        "numeric": 212,
        "latitude": 15.4167,
        "longitude": -61.3333
    },
    {
        "country": "Dominican Republic",
        "alpha2": "DO",
        "alpha3": "DOM",
        "numeric": 214,
        "latitude": 19,
        "longitude": -70.6667
    },
    {
        "country": "Ecuador",
        "alpha2": "EC",
        "alpha3": "ECU",
        "numeric": 218,
        "latitude": -2,
        "longitude": -77.5
    },
    {
        "country": "Egypt",
        "alpha2": "EG",
        "alpha3": "EGY",
        "numeric": 818,
        "latitude": 27,
        "longitude": 30
    },
    {
        "country": "El Salvador",
        "alpha2": "SV",
        "alpha3": "SLV",
        "numeric": 222,
        "latitude": 13.8333,
        "longitude": -88.9167
    },
    {
        "country": "Equatorial Guinea",
        "alpha2": "GQ",
        "alpha3": "GNQ",
        "numeric": 226,
        "latitude": 2,
        "longitude": 10
    },
    {
        "country": "Eritrea",
        "alpha2": "ER",
        "alpha3": "ERI",
        "numeric": 232,
        "latitude": 15,
        "longitude": 39
    },
    {
        "country": "Estonia",
        "alpha2": "EE",
        "alpha3": "EST",
        "numeric": 233,
        "latitude": 59,
        "longitude": 26
    },
    {
        "country": "Ethiopia",
        "alpha2": "ET",
        "alpha3": "ETH",
        "numeric": 231,
        "latitude": 8,
        "longitude": 38
    },
    {
        "country": "Falkland Islands (Malvinas)",
        "alpha2": "FK",
        "alpha3": "FLK",
        "numeric": 238,
        "latitude": -51.75,
        "longitude": -59
    },
    {
        "country": "Faroe Islands",
        "alpha2": "FO",
        "alpha3": "FRO",
        "numeric": 234,
        "latitude": 62,
        "longitude": -7
    },
    {
        "country": "Fiji",
        "alpha2": "FJ",
        "alpha3": "FJI",
        "numeric": 242,
        "latitude": -18,
        "longitude": 175
    },
    {
        "country": "Finland",
        "alpha2": "FI",
        "alpha3": "FIN",
        "numeric": 246,
        "latitude": 64,
        "longitude": 26
    },
    {
        "country": "France",
        "alpha2": "FR",
        "alpha3": "FRA",
        "numeric": 250,
        "latitude": 46,
        "longitude": 2
    },
    {
        "country": "French Guiana",
        "alpha2": "GF",
        "alpha3": "GUF",
        "numeric": 254,
        "latitude": 4,
        "longitude": -53
    },
    {
        "country": "French Polynesia",
        "alpha2": "PF",
        "alpha3": "PYF",
        "numeric": 258,
        "latitude": -15,
        "longitude": -140
    },
    {
        "country": "French Southern Territories",
        "alpha2": "TF",
        "alpha3": "ATF",
        "numeric": 260,
        "latitude": -43,
        "longitude": 67
    },
    {
        "country": "Gabon",
        "alpha2": "GA",
        "alpha3": "GAB",
        "numeric": 266,
        "latitude": -1,
        "longitude": 11.75
    },
    {
        "country": "Gambia",
        "alpha2": "GM",
        "alpha3": "GMB",
        "numeric": 270,
        "latitude": 13.4667,
        "longitude": -16.5667
    },
    {
        "country": "Georgia",
        "alpha2": "GE",
        "alpha3": "GEO",
        "numeric": 268,
        "latitude": 42,
        "longitude": 43.5
    },
    {
        "country": "Germany",
        "alpha2": "DE",
        "alpha3": "DEU",
        "numeric": 276,
        "latitude": 51,
        "longitude": 9
    },
    {
        "country": "Ghana",
        "alpha2": "GH",
        "alpha3": "GHA",
        "numeric": 288,
        "latitude": 8,
        "longitude": -2
    },
    {
        "country": "Gibraltar",
        "alpha2": "GI",
        "alpha3": "GIB",
        "numeric": 292,
        "latitude": 36.1833,
        "longitude": -5.3667
    },
    {
        "country": "Greece",
        "alpha2": "GR",
        "alpha3": "GRC",
        "numeric": 300,
        "latitude": 39,
        "longitude": 22
    },
    {
        "country": "Greenland",
        "alpha2": "GL",
        "alpha3": "GRL",
        "numeric": 304,
        "latitude": 72,
        "longitude": -40
    },
    {
        "country": "Grenada",
        "alpha2": "GD",
        "alpha3": "GRD",
        "numeric": 308,
        "latitude": 12.1167,
        "longitude": -61.6667
    },
    {
        "country": "Guadeloupe",
        "alpha2": "GP",
        "alpha3": "GLP",
        "numeric": 312,
        "latitude": 16.25,
        "longitude": -61.5833
    },
    {
        "country": "Guam",
        "alpha2": "GU",
        "alpha3": "GUM",
        "numeric": 316,
        "latitude": 13.4667,
        "longitude": 144.7833
    },
    {
        "country": "Guatemala",
        "alpha2": "GT",
        "alpha3": "GTM",
        "numeric": 320,
        "latitude": 15.5,
        "longitude": -90.25
    },
    {
        "country": "Guernsey",
        "alpha2": "GG",
        "alpha3": "GGY",
        "numeric": 831,
        "latitude": 49.5,
        "longitude": -2.56
    },
    {
        "country": "Guinea",
        "alpha2": "GN",
        "alpha3": "GIN",
        "numeric": 324,
        "latitude": 11,
        "longitude": -10
    },
    {
        "country": "Guinea-Bissau",
        "alpha2": "GW",
        "alpha3": "GNB",
        "numeric": 624,
        "latitude": 12,
        "longitude": -15
    },
    {
        "country": "Guyana",
        "alpha2": "GY",
        "alpha3": "GUY",
        "numeric": 328,
        "latitude": 5,
        "longitude": -59
    },
    {
        "country": "Haiti",
        "alpha2": "HT",
        "alpha3": "HTI",
        "numeric": 332,
        "latitude": 19,
        "longitude": -72.4167
    },
    {
        "country": "Heard Island and McDonald Islands",
        "alpha2": "HM",
        "alpha3": "HMD",
        "numeric": 334,
        "latitude": -53.1,
        "longitude": 72.5167
    },
    {
        "country": "Holy See (Vatican City State)",
        "alpha2": "VA",
        "alpha3": "VAT",
        "numeric": 336,
        "latitude": 41.9,
        "longitude": 12.45
    },
    {
        "country": "Honduras",
        "alpha2": "HN",
        "alpha3": "HND",
        "numeric": 340,
        "latitude": 15,
        "longitude": -86.5
    },
    {
        "country": "Hong Kong",
        "alpha2": "HK",
        "alpha3": "HKG",
        "numeric": 344,
        "latitude": 22.25,
        "longitude": 114.1667
    },
    {
        "country": "Hungary",
        "alpha2": "HU",
        "alpha3": "HUN",
        "numeric": 348,
        "latitude": 47,
        "longitude": 20
    },
    {
        "country": "Iceland",
        "alpha2": "IS",
        "alpha3": "ISL",
        "numeric": 352,
        "latitude": 65,
        "longitude": -18
    },
    {
        "country": "India",
        "alpha2": "IN",
        "alpha3": "IND",
        "numeric": 356,
        "latitude": 20,
        "longitude": 77
    },
    {
        "country": "Indonesia",
        "alpha2": "ID",
        "alpha3": "IDN",
        "numeric": 360,
        "latitude": -5,
        "longitude": 120
    },
    {
        "country": "Iran, Islamic Republic of",
        "alpha2": "IR",
        "alpha3": "IRN",
        "numeric": 364,
        "latitude": 32,
        "longitude": 53
    },
    {
        "country": "Iraq",
        "alpha2": "IQ",
        "alpha3": "IRQ",
        "numeric": 368,
        "latitude": 33,
        "longitude": 44
    },
    {
        "country": "Ireland",
        "alpha2": "IE",
        "alpha3": "IRL",
        "numeric": 372,
        "latitude": 53,
        "longitude": -8
    },
    {
        "country": "Isle of Man",
        "alpha2": "IM",
        "alpha3": "IMN",
        "numeric": 833,
        "latitude": 54.23,
        "longitude": -4.55
    },
    {
        "country": "Israel",
        "alpha2": "IL",
        "alpha3": "ISR",
        "numeric": 376,
        "latitude": 31.5,
        "longitude": 34.75
    },
    {
        "country": "Italy",
        "alpha2": "IT",
        "alpha3": "ITA",
        "numeric": 380,
        "latitude": 42.8333,
        "longitude": 12.8333
    },
    {
        "country": "Jamaica",
        "alpha2": "JM",
        "alpha3": "JAM",
        "numeric": 388,
        "latitude": 18.25,
        "longitude": -77.5
    },
    {
        "country": "Japan",
        "alpha2": "JP",
        "alpha3": "JPN",
        "numeric": 392,
        "latitude": 36,
        "longitude": 138
    },
    {
        "country": "Jersey",
        "alpha2": "JE",
        "alpha3": "JEY",
        "numeric": 832,
        "latitude": 49.21,
        "longitude": -2.13
    },
    {
        "country": "Jordan",
        "alpha2": "JO",
        "alpha3": "JOR",
        "numeric": 400,
        "latitude": 31,
        "longitude": 36
    },
    {
        "country": "Kazakhstan",
        "alpha2": "KZ",
        "alpha3": "KAZ",
        "numeric": 398,
        "latitude": 48,
        "longitude": 68
    },
    {
        "country": "Kenya",
        "alpha2": "KE",
        "alpha3": "KEN",
        "numeric": 404,
        "latitude": 1,
        "longitude": 38
    },
    {
        "country": "Kiribati",
        "alpha2": "KI",
        "alpha3": "KIR",
        "numeric": 296,
        "latitude": 1.4167,
        "longitude": 173
    },
    {
        "country": "Korea, Democratic People's Republic of",
        "alpha2": "KP",
        "alpha3": "PRK",
        "numeric": 408,
        "latitude": 40,
        "longitude": 127
    },
    {
        "country": "Korea, Republic of",
        "alpha2": "KR",
        "alpha3": "KOR",
        "numeric": 410,
        "latitude": 37,
        "longitude": 127.5
    },
    {
        "country": "Kuwait",
        "alpha2": "KW",
        "alpha3": "KWT",
        "numeric": 414,
        "latitude": 29.3375,
        "longitude": 47.6581
    },
    {
        "country": "Kyrgyzstan",
        "alpha2": "KG",
        "alpha3": "KGZ",
        "numeric": 417,
        "latitude": 41,
        "longitude": 75
    },
    {
        "country": "Lao People's Democratic Republic",
        "alpha2": "LA",
        "alpha3": "LAO",
        "numeric": 418,
        "latitude": 18,
        "longitude": 105
    },
    {
        "country": "Latvia",
        "alpha2": "LV",
        "alpha3": "LVA",
        "numeric": 428,
        "latitude": 57,
        "longitude": 25
    },
    {
        "country": "Lebanon",
        "alpha2": "LB",
        "alpha3": "LBN",
        "numeric": 422,
        "latitude": 33.8333,
        "longitude": 35.8333
    },
    {
        "country": "Lesotho",
        "alpha2": "LS",
        "alpha3": "LSO",
        "numeric": 426,
        "latitude": -29.5,
        "longitude": 28.5
    },
    {
        "country": "Liberia",
        "alpha2": "LR",
        "alpha3": "LBR",
        "numeric": 430,
        "latitude": 6.5,
        "longitude": -9.5
    },
    {
        "country": "Libyan Arab Jamahiriya",
        "alpha2": "LY",
        "alpha3": "LBY",
        "numeric": 434,
        "latitude": 25,
        "longitude": 17
    },
    {
        "country": "Liechtenstein",
        "alpha2": "LI",
        "alpha3": "LIE",
        "numeric": 438,
        "latitude": 47.1667,
        "longitude": 9.5333
    },
    {
        "country": "Lithuania",
        "alpha2": "LT",
        "alpha3": "LTU",
        "numeric": 440,
        "latitude": 56,
        "longitude": 24
    },
    {
        "country": "Luxembourg",
        "alpha2": "LU",
        "alpha3": "LUX",
        "numeric": 442,
        "latitude": 49.75,
        "longitude": 6.1667
    },
    {
        "country": "Macao",
        "alpha2": "MO",
        "alpha3": "MAC",
        "numeric": 446,
        "latitude": 22.1667,
        "longitude": 113.55
    },
    {
        "country": "Macedonia, the former Yugoslav Republic of",
        "alpha2": "MK",
        "alpha3": "MKD",
        "numeric": 807,
        "latitude": 41.8333,
        "longitude": 22
    },
    {
        "country": "Madagascar",
        "alpha2": "MG",
        "alpha3": "MDG",
        "numeric": 450,
        "latitude": -20,
        "longitude": 47
    },
    {
        "country": "Malawi",
        "alpha2": "MW",
        "alpha3": "MWI",
        "numeric": 454,
        "latitude": -13.5,
        "longitude": 34
    },
    {
        "country": "Malaysia",
        "alpha2": "MY",
        "alpha3": "MYS",
        "numeric": 458,
        "latitude": 2.5,
        "longitude": 112.5
    },
    {
        "country": "Maldives",
        "alpha2": "MV",
        "alpha3": "MDV",
        "numeric": 462,
        "latitude": 3.25,
        "longitude": 73
    },
    {
        "country": "Mali",
        "alpha2": "ML",
        "alpha3": "MLI",
        "numeric": 466,
        "latitude": 17,
        "longitude": -4
    },
    {
        "country": "Malta",
        "alpha2": "MT",
        "alpha3": "MLT",
        "numeric": 470,
        "latitude": 35.8333,
        "longitude": 14.5833
    },
    {
        "country": "Marshall Islands",
        "alpha2": "MH",
        "alpha3": "MHL",
        "numeric": 584,
        "latitude": 9,
        "longitude": 168
    },
    {
        "country": "Martinique",
        "alpha2": "MQ",
        "alpha3": "MTQ",
        "numeric": 474,
        "latitude": 14.6667,
        "longitude": -61
    },
    {
        "country": "Mauritania",
        "alpha2": "MR",
        "alpha3": "MRT",
        "numeric": 478,
        "latitude": 20,
        "longitude": -12
    },
    {
        "country": "Mauritius",
        "alpha2": "MU",
        "alpha3": "MUS",
        "numeric": 480,
        "latitude": -20.2833,
        "longitude": 57.55
    },
    {
        "country": "Mayotte",
        "alpha2": "YT",
        "alpha3": "MYT",
        "numeric": 175,
        "latitude": -12.8333,
        "longitude": 45.1667
    },
    {
        "country": "Mexico",
        "alpha2": "MX",
        "alpha3": "MEX",
        "numeric": 484,
        "latitude": 23,
        "longitude": -102
    },
    {
        "country": "Micronesia, Federated States of",
        "alpha2": "FM",
        "alpha3": "FSM",
        "numeric": 583,
        "latitude": 6.9167,
        "longitude": 158.25
    },
    {
        "country": "Moldova, Republic of",
        "alpha2": "MD",
        "alpha3": "MDA",
        "numeric": 498,
        "latitude": 47,
        "longitude": 29
    },
    {
        "country": "Monaco",
        "alpha2": "MC",
        "alpha3": "MCO",
        "numeric": 492,
        "latitude": 43.7333,
        "longitude": 7.4
    },
    {
        "country": "Mongolia",
        "alpha2": "MN",
        "alpha3": "MNG",
        "numeric": 496,
        "latitude": 46,
        "longitude": 105
    },
    {
        "country": "Montenegro",
        "alpha2": "ME",
        "alpha3": "MNE",
        "numeric": 499,
        "latitude": 42,
        "longitude": 19
    },
    {
        "country": "Montserrat",
        "alpha2": "MS",
        "alpha3": "MSR",
        "numeric": 500,
        "latitude": 16.75,
        "longitude": -62.2
    },
    {
        "country": "Morocco",
        "alpha2": "MA",
        "alpha3": "MAR",
        "numeric": 504,
        "latitude": 32,
        "longitude": -5
    },
    {
        "country": "Mozambique",
        "alpha2": "MZ",
        "alpha3": "MOZ",
        "numeric": 508,
        "latitude": -18.25,
        "longitude": 35
    },
    {
        "country": "Myanmar",
        "alpha2": "MM",
        "alpha3": "MMR",
        "numeric": 104,
        "latitude": 22,
        "longitude": 98
    },
    {
        "country": "Namibia",
        "alpha2": "NA",
        "alpha3": "NAM",
        "numeric": 516,
        "latitude": -22,
        "longitude": 17
    },
    {
        "country": "Nauru",
        "alpha2": "NR",
        "alpha3": "NRU",
        "numeric": 520,
        "latitude": -0.5333,
        "longitude": 166.9167
    },
    {
        "country": "Nepal",
        "alpha2": "NP",
        "alpha3": "NPL",
        "numeric": 524,
        "latitude": 28,
        "longitude": 84
    },
    {
        "country": "Netherlands",
        "alpha2": "NL",
        "alpha3": "NLD",
        "numeric": 528,
        "latitude": 52.5,
        "longitude": 5.75
    },
    {
        "country": "Netherlands Antilles",
        "alpha2": "AN",
        "alpha3": "ANT",
        "numeric": 530,
        "latitude": 12.25,
        "longitude": -68.75
    },
    {
        "country": "New Caledonia",
        "alpha2": "NC",
        "alpha3": "NCL",
        "numeric": 540,
        "latitude": -21.5,
        "longitude": 165.5
    },
    {
        "country": "New Zealand",
        "alpha2": "NZ",
        "alpha3": "NZL",
        "numeric": 554,
        "latitude": -41,
        "longitude": 174
    },
    {
        "country": "Nicaragua",
        "alpha2": "NI",
        "alpha3": "NIC",
        "numeric": 558,
        "latitude": 13,
        "longitude": -85
    },
    {
        "country": "Niger",
        "alpha2": "NE",
        "alpha3": "NER",
        "numeric": 562,
        "latitude": 16,
        "longitude": 8
    },
    {
        "country": "Nigeria",
        "alpha2": "NG",
        "alpha3": "NGA",
        "numeric": 566,
        "latitude": 10,
        "longitude": 8
    },
    {
        "country": "Niue",
        "alpha2": "NU",
        "alpha3": "NIU",
        "numeric": 570,
        "latitude": -19.0333,
        "longitude": -169.8667
    },
    {
        "country": "Norfolk Island",
        "alpha2": "NF",
        "alpha3": "NFK",
        "numeric": 574,
        "latitude": -29.0333,
        "longitude": 167.95
    },
    {
        "country": "Northern Mariana Islands",
        "alpha2": "MP",
        "alpha3": "MNP",
        "numeric": 580,
        "latitude": 15.2,
        "longitude": 145.75
    },
    {
        "country": "Norway",
        "alpha2": "NO",
        "alpha3": "NOR",
        "numeric": 578,
        "latitude": 62,
        "longitude": 10
    },
    {
        "country": "Oman",
        "alpha2": "OM",
        "alpha3": "OMN",
        "numeric": 512,
        "latitude": 21,
        "longitude": 57
    },
    {
        "country": "Pakistan",
        "alpha2": "PK",
        "alpha3": "PAK",
        "numeric": 586,
        "latitude": 30,
        "longitude": 70
    },
    {
        "country": "Palau",
        "alpha2": "PW",
        "alpha3": "PLW",
        "numeric": 585,
        "latitude": 7.5,
        "longitude": 134.5
    },
    {
        "country": "Palestinian Territory, Occupied",
        "alpha2": "PS",
        "alpha3": "PSE",
        "numeric": 275,
        "latitude": 32,
        "longitude": 35.25
    },
    {
        "country": "Panama",
        "alpha2": "PA",
        "alpha3": "PAN",
        "numeric": 591,
        "latitude": 9,
        "longitude": -80
    },
    {
        "country": "Papua New Guinea",
        "alpha2": "PG",
        "alpha3": "PNG",
        "numeric": 598,
        "latitude": -6,
        "longitude": 147
    },
    {
        "country": "Paraguay",
        "alpha2": "PY",
        "alpha3": "PRY",
        "numeric": 600,
        "latitude": -23,
        "longitude": -58
    },
    {
        "country": "Peru",
        "alpha2": "PE",
        "alpha3": "PER",
        "numeric": 604,
        "latitude": -10,
        "longitude": -76
    },
    {
        "country": "Philippines",
        "alpha2": "PH",
        "alpha3": "PHL",
        "numeric": 608,
        "latitude": 13,
        "longitude": 122
    },
    {
        "country": "Pitcairn",
        "alpha2": "PN",
        "alpha3": "PCN",
        "numeric": 612,
        "latitude": -24.7,
        "longitude": -127.4
    },
    {
        "country": "Poland",
        "alpha2": "PL",
        "alpha3": "POL",
        "numeric": 616,
        "latitude": 52,
        "longitude": 20
    },
    {
        "country": "Portugal",
        "alpha2": "PT",
        "alpha3": "PRT",
        "numeric": 620,
        "latitude": 39.5,
        "longitude": -8
    },
    {
        "country": "Puerto Rico",
        "alpha2": "PR",
        "alpha3": "PRI",
        "numeric": 630,
        "latitude": 18.25,
        "longitude": -66.5
    },
    {
        "country": "Qatar",
        "alpha2": "QA",
        "alpha3": "QAT",
        "numeric": 634,
        "latitude": 25.5,
        "longitude": 51.25
    },
    {
        "country": "Réunion",
        "alpha2": "RE",
        "alpha3": "REU",
        "numeric": 638,
        "latitude": -21.1,
        "longitude": 55.6
    },
    {
        "country": "Romania",
        "alpha2": "RO",
        "alpha3": "ROU",
        "numeric": 642,
        "latitude": 46,
        "longitude": 25
    },
    {
        "country": "Russian Federation",
        "alpha2": "RU",
        "alpha3": "RUS",
        "numeric": 643,
        "latitude": 60,
        "longitude": 100
    },
    {
        "country": "Rwanda",
        "alpha2": "RW",
        "alpha3": "RWA",
        "numeric": 646,
        "latitude": -2,
        "longitude": 30
    },
    {
        "country": "Saint Helena, Ascension and Tristan da Cunha",
        "alpha2": "SH",
        "alpha3": "SHN",
        "numeric": 654,
        "latitude": -15.9333,
        "longitude": -5.7
    },
    {
        "country": "Saint Kitts and Nevis",
        "alpha2": "KN",
        "alpha3": "KNA",
        "numeric": 659,
        "latitude": 17.3333,
        "longitude": -62.75
    },
    {
        "country": "Saint Lucia",
        "alpha2": "LC",
        "alpha3": "LCA",
        "numeric": 662,
        "latitude": 13.8833,
        "longitude": -61.1333
    },
    {
        "country": "Saint Pierre and Miquelon",
        "alpha2": "PM",
        "alpha3": "SPM",
        "numeric": 666,
        "latitude": 46.8333,
        "longitude": -56.3333
    },
    {
        "country": "Saint Vincent and the Grenadines",
        "alpha2": "VC",
        "alpha3": "VCT",
        "numeric": 670,
        "latitude": 13.25,
        "longitude": -61.2
    },
    {
        "country": "Samoa",
        "alpha2": "WS",
        "alpha3": "WSM",
        "numeric": 882,
        "latitude": -13.5833,
        "longitude": -172.3333
    },
    {
        "country": "San Marino",
        "alpha2": "SM",
        "alpha3": "SMR",
        "numeric": 674,
        "latitude": 43.7667,
        "longitude": 12.4167
    },
    {
        "country": "Sao Tome and Principe",
        "alpha2": "ST",
        "alpha3": "STP",
        "numeric": 678,
        "latitude": 1,
        "longitude": 7
    },
    {
        "country": "Saudi Arabia",
        "alpha2": "SA",
        "alpha3": "SAU",
        "numeric": 682,
        "latitude": 25,
        "longitude": 45
    },
    {
        "country": "Senegal",
        "alpha2": "SN",
        "alpha3": "SEN",
        "numeric": 686,
        "latitude": 14,
        "longitude": -14
    },
    {
        "country": "Serbia",
        "alpha2": "RS",
        "alpha3": "SRB",
        "numeric": 688,
        "latitude": 44,
        "longitude": 21
    },
    {
        "country": "Seychelles",
        "alpha2": "SC",
        "alpha3": "SYC",
        "numeric": 690,
        "latitude": -4.5833,
        "longitude": 55.6667
    },
    {
        "country": "Sierra Leone",
        "alpha2": "SL",
        "alpha3": "SLE",
        "numeric": 694,
        "latitude": 8.5,
        "longitude": -11.5
    },
    {
        "country": "Singapore",
        "alpha2": "SG",
        "alpha3": "SGP",
        "numeric": 702,
        "latitude": 1.3667,
        "longitude": 103.8
    },
    {
        "country": "Slovakia",
        "alpha2": "SK",
        "alpha3": "SVK",
        "numeric": 703,
        "latitude": 48.6667,
        "longitude": 19.5
    },
    {
        "country": "Slovenia",
        "alpha2": "SI",
        "alpha3": "SVN",
        "numeric": 705,
        "latitude": 46,
        "longitude": 15
    },
    {
        "country": "Solomon Islands",
        "alpha2": "SB",
        "alpha3": "SLB",
        "numeric": 90,
        "latitude": -8,
        "longitude": 159
    },
    {
        "country": "Somalia",
        "alpha2": "SO",
        "alpha3": "SOM",
        "numeric": 706,
        "latitude": 10,
        "longitude": 49
    },
    {
        "country": "South Africa",
        "alpha2": "ZA",
        "alpha3": "ZAF",
        "numeric": 710,
        "latitude": -29,
        "longitude": 24
    },
    {
        "country": "South Georgia and the South Sandwich Islands",
        "alpha2": "GS",
        "alpha3": "SGS",
        "numeric": 239,
        "latitude": -54.5,
        "longitude": -37
    },
    {
        "country": "Spain",
        "alpha2": "ES",
        "alpha3": "ESP",
        "numeric": 724,
        "latitude": 40,
        "longitude": -4
    },
    {
        "country": "Sri Lanka",
        "alpha2": "LK",
        "alpha3": "LKA",
        "numeric": 144,
        "latitude": 7,
        "longitude": 81
    },
    {
        "country": "Sudan",
        "alpha2": "SD",
        "alpha3": "SDN",
        "numeric": 736,
        "latitude": 15,
        "longitude": 30
    },
    {
        "country": "Suriname",
        "alpha2": "SR",
        "alpha3": "SUR",
        "numeric": 740,
        "latitude": 4,
        "longitude": -56
    },
    {
        "country": "Svalbard and Jan Mayen",
        "alpha2": "SJ",
        "alpha3": "SJM",
        "numeric": 744,
        "latitude": 78,
        "longitude": 20
    },
    {
        "country": "Swaziland",
        "alpha2": "SZ",
        "alpha3": "SWZ",
        "numeric": 748,
        "latitude": -26.5,
        "longitude": 31.5
    },
    {
        "country": "Sweden",
        "alpha2": "SE",
        "alpha3": "SWE",
        "numeric": 752,
        "latitude": 62,
        "longitude": 15
    },
    {
        "country": "Switzerland",
        "alpha2": "CH",
        "alpha3": "CHE",
        "numeric": 756,
        "latitude": 47,
        "longitude": 8
    },
    {
        "country": "Syrian Arab Republic",
        "alpha2": "SY",
        "alpha3": "SYR",
        "numeric": 760,
        "latitude": 35,
        "longitude": 38
    },
    {
        "country": "Taiwan, Province of China",
        "alpha2": "TW",
        "alpha3": "TWN",
        "numeric": 158,
        "latitude": 23.5,
        "longitude": 121
    },
    {
        "country": "Tajikistan",
        "alpha2": "TJ",
        "alpha3": "TJK",
        "numeric": 762,
        "latitude": 39,
        "longitude": 71
    },
    {
        "country": "Tanzania, United Republic of",
        "alpha2": "TZ",
        "alpha3": "TZA",
        "numeric": 834,
        "latitude": -6,
        "longitude": 35
    },
    {
        "country": "Thailand",
        "alpha2": "TH",
        "alpha3": "THA",
        "numeric": 764,
        "latitude": 15,
        "longitude": 100
    },
    {
        "country": "Timor-Leste",
        "alpha2": "TL",
        "alpha3": "TLS",
        "numeric": 626,
        "latitude": -8.55,
        "longitude": 125.5167
    },
    {
        "country": "Togo",
        "alpha2": "TG",
        "alpha3": "TGO",
        "numeric": 768,
        "latitude": 8,
        "longitude": 1.1667
    },
    {
        "country": "Tokelau",
        "alpha2": "TK",
        "alpha3": "TKL",
        "numeric": 772,
        "latitude": -9,
        "longitude": -172
    },
    {
        "country": "Tonga",
        "alpha2": "TO",
        "alpha3": "TON",
        "numeric": 776,
        "latitude": -20,
        "longitude": -175
    },
    {
        "country": "Trinidad and Tobago",
        "alpha2": "TT",
        "alpha3": "TTO",
        "numeric": 780,
        "latitude": 11,
        "longitude": -61
    },
    {
        "country": "Tunisia",
        "alpha2": "TN",
        "alpha3": "TUN",
        "numeric": 788,
        "latitude": 34,
        "longitude": 9
    },
    {
        "country": "Turkey",
        "alpha2": "TR",
        "alpha3": "TUR",
        "numeric": 792,
        "latitude": 39,
        "longitude": 35
    },
    {
        "country": "Turkmenistan",
        "alpha2": "TM",
        "alpha3": "TKM",
        "numeric": 795,
        "latitude": 40,
        "longitude": 60
    },
    {
        "country": "Turks and Caicos Islands",
        "alpha2": "TC",
        "alpha3": "TCA",
        "numeric": 796,
        "latitude": 21.75,
        "longitude": -71.5833
    },
    {
        "country": "Tuvalu",
        "alpha2": "TV",
        "alpha3": "TUV",
        "numeric": 798,
        "latitude": -8,
        "longitude": 178
    },
    {
        "country": "Uganda",
        "alpha2": "UG",
        "alpha3": "UGA",
        "numeric": 800,
        "latitude": 1,
        "longitude": 32
    },
    {
        "country": "Ukraine",
        "alpha2": "UA",
        "alpha3": "UKR",
        "numeric": 804,
        "latitude": 49,
        "longitude": 32
    },
    {
        "country": "United Arab Emirates",
        "alpha2": "AE",
        "alpha3": "ARE",
        "numeric": 784,
        "latitude": 24,
        "longitude": 54
    },
    {
        "country": "United Kingdom",
        "alpha2": "GB",
        "alpha3": "GBR",
        "numeric": 826,
        "latitude": 54,
        "longitude": -2
    },
    {
        "country": "United States",
        "alpha2": "US",
        "alpha3": "USA",
        "numeric": 840,
        "latitude": 38,
        "longitude": -97
    },
    {
        "country": "United States Minor Outlying Islands",
        "alpha2": "UM",
        "alpha3": "UMI",
        "numeric": 581,
        "latitude": 19.2833,
        "longitude": 166.6
    },
    {
        "country": "Uruguay",
        "alpha2": "UY",
        "alpha3": "URY",
        "numeric": 858,
        "latitude": -33,
        "longitude": -56
    },
    {
        "country": "Uzbekistan",
        "alpha2": "UZ",
        "alpha3": "UZB",
        "numeric": 860,
        "latitude": 41,
        "longitude": 64
    },
    {
        "country": "Vanuatu",
        "alpha2": "VU",
        "alpha3": "VUT",
        "numeric": 548,
        "latitude": -16,
        "longitude": 167
    },
    {
        "country": "Venezuela, Bolivarian Republic of",
        "alpha2": "VE",
        "alpha3": "VEN",
        "numeric": 862,
        "latitude": 8,
        "longitude": -66
    },
    {
        "country": "Viet Nam",
        "alpha2": "VN",
        "alpha3": "VNM",
        "numeric": 704,
        "latitude": 16,
        "longitude": 106
    },
    {
        "country": "Virgin Islands, British",
        "alpha2": "VG",
        "alpha3": "VGB",
        "numeric": 92,
        "latitude": 18.5,
        "longitude": -64.5
    },
    {
        "country": "Virgin Islands, U.S.",
        "alpha2": "VI",
        "alpha3": "VIR",
        "numeric": 850,
        "latitude": 18.3333,
        "longitude": -64.8333
    },
    {
        "country": "Wallis and Futuna",
        "alpha2": "WF",
        "alpha3": "WLF",
        "numeric": 876,
        "latitude": -13.3,
        "longitude": -176.2
    },
    {
        "country": "Western Sahara",
        "alpha2": "EH",
        "alpha3": "ESH",
        "numeric": 732,
        "latitude": 24.5,
        "longitude": -13
    },
    {
        "country": "Yemen",
        "alpha2": "YE",
        "alpha3": "YEM",
        "numeric": 887,
        "latitude": 15,
        "longitude": 48
    },
    {
        "country": "Zambia",
        "alpha2": "ZM",
        "alpha3": "ZMB",
        "numeric": 894,
        "latitude": -15,
        "longitude": 30
    },
    {
        "country": "Zimbabwe",
        "alpha2": "ZW",
        "alpha3": "ZWE",
        "numeric": 716,
        "latitude": -20,
        "longitude": 30
    },
    {
        "country": "Afghanistan",
        "alpha2": "AF",
        "alpha3": "AFG",
        "numeric": 4,
        "latitude": 33,
        "longitude": 65
    }
];

// https://ajonp.com/courses/angularmaterial/angular-material-theming/
class ThemeColorPickerService {
    constructor(overlayContainer) {
        this.overlayContainer = overlayContainer;
        this.initialClass = 'fathym-contrast-theme';
        this.colorClass = new BehaviorSubject$1(this.initialClass);
        const storageClass = localStorage.getItem('theme-picker');
        console.log('color picker service', storageClass);
        if (storageClass) {
            overlayContainer.getContainerElement().classList.add(storageClass);
            this.colorClass.next(storageClass);
        }
        else {
            overlayContainer.getContainerElement().classList.add(this.initialClass);
        }
    }
    GetColorClass() {
        return this.colorClass;
    }
    SetColorClass(className) {
        this.overlayContainer.getContainerElement().classList.forEach(css => {
            if (css !== 'cdk-overlay-container') {
                this.overlayContainer.getContainerElement().classList.remove(css);
            }
        });
        this.overlayContainer.getContainerElement().classList.add(className);
        this.colorClass.next(className);
        localStorage.setItem('theme-picker', className);
    }
}
ThemeColorPickerService.ctorParameters = () => [
    { type: OverlayContainer }
];
ThemeColorPickerService.ɵprov = ɵɵdefineInjectable({ factory: function ThemeColorPickerService_Factory() { return new ThemeColorPickerService(ɵɵinject(OverlayContainer)); }, token: ThemeColorPickerService, providedIn: "root" });
ThemeColorPickerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
ThemeColorPickerService.ctorParameters = () => [
    { type: OverlayContainer }
];

/**
 * Generated bundle index. Do not edit.
 */

export { Address, Amount, Application, BROWSER_FAVICONS_CONFIG, Base64Model, BaseModeledResponse, BaseResponse, BrowserFavicons, Constants, ConvertToBase64Util, CountryConstants, DAFAPIApplicationDetails, DAFAppPointerApplicationDetails, DAFApplication, DAFRedirectApplicationDetails, DAFService, DAFViewApplicationDetails, DataFlow, DataFlowAction, DataFlowActionEvent, DataFlowLayout, DataFlowModule, DataFlowModuleControlTypes, DataFlowModuleDisplay, DataFlowModuleOption, DataFlowModulePack, DataFlowModulePackSetup, DataFlowModuleShapeTypes, DataFlowOutput, DataFlowStream, DataPipeConstants, DataPipes, DecamelizePipe, DirectiveModule, DotNotationUtil, EmailValidationMessagesConstants, EmailValidator, FathymSharedModule, FaviconsService, Guid, Icon, IdeActivity, IdeEditor, IdePanel, IdeSideBar, IdeSideBarAction, ImageMessage, IsDataTypeUtil, JSONFlattenUnflatten, JSONSchema, LCUApplicationConfig, LCUElementContext, LCUServiceSettings, LCUStateConfig, LcuElementComponent, Loading, MaterialModule, NapkinIDESolution, ObservableContextService, Pageable, PasswordValidationMessagesConstants, PasswordValidator, PhoneValidator, PipeModule, RealTimeConnection, RealTimeService, RouterHelpersService, SafePipe, SanitizeHtmlPipe, StateAction, StateContext, Status, StringPipes, TemperatureConversion, ThemeColorDirective, ThemeColorPickerService, ToggleThemeUtil, UserNameValidator, UserValidationMessagesConstants, ValidationMessageModel, ValidationMessages, ZipcodeValidator, isResultSuccess, isStatusSuccess, winAny, LCUInterceptor as ɵa };
