// @lcu/common/esm2015/lib/api/daf/daf.service.js
import { throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { LCUServiceSettings } from '../lcu-service-settings';
export class DAFService {
    //  Constructors
    constructor(injector) {
        this.injector = injector;
        this.http = injector.get(HttpClient);
        try {
            this.settings = injector.get(LCUServiceSettings);
        }
        catch (err) { }
    }
    //  Helpers
    catchError() {
        return catchError(this.handleError);
    }
    delete(path) {
        return this.http.delete(this.resolvePath(path)).pipe(this.catchError());
    }
    get(path) {
        return this.http.get(this.resolvePath(path)).pipe(this.catchError());
    }
    handleError(error) {
        // In a real world app, we might use a remote logging infrastructure
        // We'd also dig deeper into the error to get a better message
        const errMsg = error.message ? error.message : error.status ? `${error.status} - ${error.statusText}` : 'Server error';
        console.error(errMsg); // log to console instead
        return throwError(errMsg);
    }
    patch(data, path) {
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.patch(this.resolvePath(path), data, options).pipe(this.catchError());
    }
    post(data, path) {
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.post(this.resolvePath(path), data, options).pipe(this.catchError());
    }
    put(data, path) {
        const options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.put(this.resolvePath(path), data, options).pipe(this.catchError());
    }
    resolvePath(path) {
        if (this.settings && this.settings.APIRoot) {
            if (this.settings.APIRoot.endsWith('/')) {
                this.settings.APIRoot = this.settings.APIRoot.substring(this.settings.APIRoot.length - 1);
            }
            if (path.startsWith('/')) {
                path = path.substring(1);
            }
            return `${this.settings.APIRoot}/${path}`;
        }
        else {
            return path;
        }
    }
}
