// @lowcodeunit-devkit/ng-lcu/src/application/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("@schematics/angular/utility/config");
const tasks_1 = require("@angular-devkit/schematics/tasks");
const schematics_1 = require("@angular-devkit/schematics");
const core_1 = require("@angular-devkit/core");
const helpers_1 = require("../utils/helpers");
function application(options) {
    return (host, context) => {
        setupOptions(host, options);
        context.logger.info(`APPLICATON OPTIONS: ${options.name}...`);
        const rule = schematics_1.chain([
            schematics_1.externalSchematic('@schematics/angular', 'application', {
                name: options.name,
                routing: options.routing,
                prefix: options.prefix,
                style: 'scss'
            }),
            processInitWith(options, context),
            addLicense(context, options),
            options.blockFavicon ? schematics_1.noop() : mergeFavIcon(options),
            options.blockDeploy ? schematics_1.noop() : addScripts(options),
            options.blockDeploy ? schematics_1.noop() : manageDeployAllScript(options),
            manageAppAssets(options, context)
        ]);
        if (!options.skipInstall)
            context.addTask(new tasks_1.NodePackageInstallTask());
        return rule(host, context);
    };
}
exports.application = application;
function mergeFavIcon(options) {
    return (host) => {
        const workspace = config_1.getWorkspace(host);
        let project = workspace.projects[options.name];
        const targetPath = core_1.normalize(project.root + '/src/');
        const solutionSource = schematics_1.apply(schematics_1.url('./files/favicon'), [
            schematics_1.template(Object.assign(Object.assign({}, core_1.strings), options)),
            schematics_1.move(targetPath),
        ]);
        return schematics_1.mergeWith(solutionSource, schematics_1.MergeStrategy.Overwrite);
    };
}
exports.mergeFavIcon = mergeFavIcon;
function mergeAppFiles(options) {
    return (host) => {
        const workspace = config_1.getWorkspace(host);
        let project = workspace.projects[options.name];
        const targetPath = core_1.normalize(project.root + '/src/app');
        const solutionSource = schematics_1.apply(schematics_1.url('./files/app'), [
            schematics_1.template(Object.assign(Object.assign({}, core_1.strings), options)),
            schematics_1.move(targetPath),
        ]);
        return schematics_1.mergeWith(solutionSource, schematics_1.MergeStrategy.Overwrite);
    };
}
exports.mergeAppFiles = mergeAppFiles;
function addScripts(options) {
    return (host) => {
        const workspace = config_1.getWorkspace(host);
        let project = workspace.projects[options.name];
        let projectSafeName = core_1.strings.dasherize(options.name);
        helpers_1.addScriptsToPackageFile(host, [
            !options.singleBundle
                ? {
                    key: `build:${projectSafeName}`,
                    value: `ng build ${projectSafeName} --prod`
                }
                : {
                    key: `build:${projectSafeName}`,
                    value: `ng build lcu --prod --single-bundle && npm run pack`
                },
            {
                key: `deploy:${projectSafeName}`,
                value: `npm version patch --prefix ${project.root} ` +
                    `&& npm run build:${projectSafeName} && npm publish ./dist/${projectSafeName} --access public`
            }
        ]);
        return host;
    };
}
exports.addScripts = addScripts;
function createPackageJson(host, options, projectName, context) {
    let workspace = config_1.getWorkspace(host);
    let project = workspace.projects[projectName];
    let packageFilePath = core_1.join(project.root, 'package.json');
    context.logger.info(`Loading package at path: ${packageFilePath}`);
    //  TODO:  Will need a way to force just using the project name and never combining with workspace
    var repoName = options.workspace == projectName ? projectName : `${options.workspace}-${projectName}`;
    let packageJson = {
        name: `${options.scope}/${repoName}`,
        version: '0.0.1',
        peerDependencies: {
            '@angular/common': '~8.1.2',
            '@angular/core': '~8.1.2'
        }
    };
    host.create(packageFilePath, JSON.stringify(packageJson, null, '\t'));
}
exports.createPackageJson = createPackageJson;
function updatePolyfills(host, options, projectName, context) {
    let workspace = config_1.getWorkspace(host);
    let project = workspace.projects[projectName];
    let polysFilePath = core_1.join(project.root, 'polyfills.ts');
    let polysFile = host.get(polysFilePath);
    let polysContent = polysFile ? polysFile.content.toString('utf8') : '';
    if (options.webCompPolys)
        polysContent += `\r\nimport '@webcomponents/custom-elements/custom-elements.min';`;
    host.create(polysFilePath, polysContent);
}
exports.updatePolyfills = updatePolyfills;
function manageAppAssets(options, context) {
    return (host) => {
        let projectSafeName = core_1.strings.dasherize(options.name);
        let packageGlob = {
            glob: 'package.json',
            input: `./projects/${projectSafeName}/`,
            output: '/'
        };
        let angularFile = host.get('angular.json');
        let angularJson = angularFile ? JSON.parse(angularFile.content.toString('utf8')) : {};
        let angularJsonText = JSON.stringify(angularJson);
        angularJson.projects[projectSafeName].architect.build.options.assets.push(packageGlob);
        if (options.isDefault)
            angularJson.defaultProject = projectSafeName;
        if (options.singleBundle) {
            angularJson.projects[projectSafeName].architect.build.options.assets = angularJson.projects[projectSafeName].architect.build.options.assets.filter((a) => a != `projects/${projectSafeName}/src/assets`);
            angularJson.projects[projectSafeName].architect.build.options.assets.push({
                glob: 'lcu.json',
                input: './',
                output: '/'
            });
        }
        if (options.es5Patch)
            delete angularJson.projects[projectSafeName].architect.build.options.es5BrowserSupport;
        if (options.es5Patch)
            delete angularJson.projects[projectSafeName].architect.build.options.es5BrowserSupport;
        if (options.webCompPolys)
            angularJson.projects[projectSafeName].architect.build.options.scripts.push('node_modules/@webcomponents/custom-elements/src/native-shim.js');
        delete angularJson.projects[projectSafeName].architect.build.configurations.production.budgets;
        host.overwrite('angular.json', JSON.stringify(angularJson, null, '\t'));
        createPackageJson(host, options, projectSafeName, context);
        return host;
    };
}
exports.manageAppAssets = manageAppAssets;
function manageDeployAllScript(options) {
    return (host) => {
        let projectSafeName = core_1.strings.dasherize(options.name);
        let deployProj = `npm run deploy:${projectSafeName}`;
        let packageFile = host.get('package.json');
        let packageJson = packageFile ? JSON.parse(packageFile.content.toString('utf8')) : {};
        let deployAll = packageJson.scripts['deploy:all'];
        if (deployAll)
            deployAll += ` && ${deployProj}`;
        else
            deployAll = deployProj;
        packageJson.scripts['deploy:all'] = deployAll;
        host.overwrite('package.json', JSON.stringify(packageJson, null, '\t'));
        return host;
    };
}
exports.manageDeployAllScript = manageDeployAllScript;
function blankOutLibrary(options, context, exceptModule, includeLCUCore) {
    return (host) => {
        let projectName = options.name;
        let workspace = config_1.getWorkspace(host);
        let project = workspace.projects[projectName];
        let srcRoot = core_1.join(project.root, 'src');
        let appRoot = core_1.join(srcRoot, 'app');
        let files = [`app.component.html`, `app.component.scss`, `app.component.spec.ts`, `app.component.ts`, `app-routing.module.ts`];
        if (!exceptModule)
            files.push(`app.module.ts`);
        helpers_1.removeFilesFromRoot(host, appRoot, files);
        if (includeLCUCore) {
            let coreFiles = [`index.html`, `styles.scss`];
            helpers_1.removeFilesFromRoot(host, srcRoot, coreFiles);
        }
        if (!options.blockFavicon) {
            let coreFiles = [`favicon.ico`];
            helpers_1.removeFilesFromRoot(host, srcRoot, coreFiles);
        }
        return host;
    };
}
function processInitWith(options, context) {
    return (host) => {
        context.logger.info(`Processing Initialization for ${options.initWith}...`);
        let rule = schematics_1.noop();
        switch (options.initWith) {
            case 'Blank':
                rule = blankOutLibrary(options, context, false, false);
                break;
            case 'Default':
                rule = schematics_1.chain([
                    blankOutLibrary(options, context, false, false),
                    mergeAppFiles(options)
                ]);
                break;
            case 'LCU-Core-App':
                rule = schematics_1.chain([
                    blankOutLibrary(options, context, false, true),
                    schematics_1.schematic('lcu-core-app', {
                        name: options.name,
                        project: options.name
                    })
                ]);
                break;
            case 'LCU-Form':
                rule = schematics_1.chain([
                    blankOutLibrary(options, context, false, true),
                    schematics_1.schematic('lcu-form', {
                        name: options.name,
                        project: options.name
                    })
                ]);
                break;
            case 'LCU-Starter-App':
                rule = schematics_1.chain([
                    blankOutLibrary(options, context, false, false),
                    schematics_1.schematic('lcu-starter-app', {
                        name: options.name,
                        project: options.name,
                        elementName: options.elementName
                    })
                ]);
                break;
            case 'Module':
                rule = blankOutLibrary(options, context, true, false);
                break;
        }
        context.logger.info(`Processing Initialized for ${options.initWith}!`);
        return rule;
    };
}
function setupOptions(host, options) {
    let lcuFile = host.get('lcu.json');
    let lcuJson = lcuFile ? JSON.parse(lcuFile.content.toString('utf8')) : {};
    options.scope = lcuJson.templates.scope;
    options.workspace = lcuJson.templates.workspace;
    options.entryFile = 'lcu.api';
    options.blockDeploy = options.blockDeploy || false;
    options.blockFavicon = options.blockFavicon || false;
    options.initWith = options.initWith || 'LCU-Core-App';
    options.isDefault = options.isDefault || false;
    options.es5Patch = options.es5Patch || false;
    options.webCompPolys = options.webCompPolys;
    options.name = options.name || 'library';
    options.prefix = options.prefix || 'lcu';
    options.routing = options.routing === undefined ? true : options.routing;
    options.singleBundle = options.singleBundle || false;
    options.skipInstall = options.skipInstall || false;
    return host;
}
exports.setupOptions = setupOptions;
/**
 * add Apache license
 */
function addLicense(context, options) {
    return (host) => {
        let workspace = config_1.getWorkspace(host);
        let projectName = options.name;
        let scope = options.scope;
        let project = workspace.projects[projectName];
        let variant = projectName ? `-${projectName}` : '';
        let fullProjectName = `${options.scope}/${options.workspace}${variant}`;
        // let newLicense: string = String(host.read('LICENSE'));
        let license = `Apache License
    Version 2.0, January 2004
  http://www.apache.org/licenses/

  TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION

  1. Definitions.

  "License" shall mean the terms and conditions for use, reproduction,
  and distribution as defined by Sections 1 through 9 of this document.

  "Licensor" shall mean the copyright owner or entity authorized by
  the copyright owner that is granting the License.

  "Legal Entity" shall mean the union of the acting entity and all
  other entities that control, are controlled by, or are under common
  control with that entity. For the purposes of this definition,
  "control" means (i) the power, direct or indirect, to cause the
  direction or management of such entity, whether by contract or
  otherwise, or (ii) ownership of fifty percent (50%) or more of the
  outstanding shares, or (iii) beneficial ownership of such entity.

  "You" (or "Your") shall mean an individual or Legal Entity
  exercising permissions granted by this License.

  "Source" form shall mean the preferred form for making modifications,
  including but not limited to software source code, documentation
  source, and configuration files.

  "Object" form shall mean any form resulting from mechanical
  transformation or translation of a Source form, including but
  not limited to compiled object code, generated documentation,
  and conversions to other media types.

  "Work" shall mean the work of authorship, whether in Source or
  Object form, made available under the License, as indicated by a
  copyright notice that is included in or attached to the work
  (an example is provided in the Appendix below).

  "Derivative Works" shall mean any work, whether in Source or Object
  form, that is based on (or derived from) the Work and for which the
  editorial revisions, annotations, elaborations, or other modifications
  represent, as a whole, an original work of authorship. For the purposes
  of this License, Derivative Works shall not include works that remain
  separable from, or merely link (or bind by name) to the interfaces of,
  the Work and Derivative Works thereof.

  "Contribution" shall mean any work of authorship, including
  the original version of the Work and any modifications or additions
  to that Work or Derivative Works thereof, that is intentionally
  submitted to Licensor for inclusion in the Work by the copyright owner
  or by an individual or Legal Entity authorized to submit on behalf of
  the copyright owner. For the purposes of this definition, "submitted"
  means any form of electronic, verbal, or written communication sent
  to the Licensor or its representatives, including but not limited to
  communication on electronic mailing lists, source code control systems,
  and issue tracking systems that are managed by, or on behalf of, the
  Licensor for the purpose of discussing and improving the Work, but
  excluding communication that is conspicuously marked or otherwise
  designated in writing by the copyright owner as "Not a Contribution."

  "Contributor" shall mean Licensor and any individual or Legal Entity
  on behalf of whom a Contribution has been received by Licensor and
  subsequently incorporated within the Work.

  2. Grant of Copyright License. Subject to the terms and conditions of
  this License, each Contributor hereby grants to You a perpetual,
  worldwide, non-exclusive, no-charge, royalty-free, irrevocable
  copyright license to reproduce, prepare Derivative Works of,
  publicly display, publicly perform, sublicense, and distribute the
  Work and such Derivative Works in Source or Object form.

  3. Grant of Patent License. Subject to the terms and conditions of
  this License, each Contributor hereby grants to You a perpetual,
  worldwide, non-exclusive, no-charge, royalty-free, irrevocable
  (except as stated in this section) patent license to make, have made,
  use, offer to sell, sell, import, and otherwise transfer the Work,
  where such license applies only to those patent claims licensable
  by such Contributor that are necessarily infringed by their
  Contribution(s) alone or by combination of their Contribution(s)
  with the Work to which such Contribution(s) was submitted. If You
  institute patent litigation against any entity (including a
  cross-claim or counterclaim in a lawsuit) alleging that the Work
  or a Contribution incorporated within the Work constitutes direct
  or contributory patent infringement, then any patent licenses
  granted to You under this License for that Work shall terminate
  as of the date such litigation is filed.

  4. Redistribution. You may reproduce and distribute copies of the
  Work or Derivative Works thereof in any medium, with or without
  modifications, and in Source or Object form, provided that You
  meet the following conditions:

  (a) You must give any other recipients of the Work or
  Derivative Works a copy of this License; and

  (b) You must cause any modified files to carry prominent notices
  stating that You changed the files; and

  (c) You must retain, in the Source form of any Derivative Works
  that You distribute, all copyright, patent, trademark, and
  attribution notices from the Source form of the Work,
  excluding those notices that do not pertain to any part of
  the Derivative Works; and

  (d) If the Work includes a "NOTICE" text file as part of its
  distribution, then any Derivative Works that You distribute must
  include a readable copy of the attribution notices contained
  within such NOTICE file, excluding those notices that do not
  pertain to any part of the Derivative Works, in at least one
  of the following places: within a NOTICE text file distributed
  as part of the Derivative Works; within the Source form or
  documentation, if provided along with the Derivative Works; or,
  within a display generated by the Derivative Works, if and
  wherever such third-party notices normally appear. The contents
  of the NOTICE file are for informational purposes only and
  do not modify the License. You may add Your own attribution
  notices within Derivative Works that You distribute, alongside
  or as an addendum to the NOTICE text from the Work, provided
  that such additional attribution notices cannot be construed
  as modifying the License.

  You may add Your own copyright statement to Your modifications and
  may provide additional or different license terms and conditions
  for use, reproduction, or distribution of Your modifications, or
  for any such Derivative Works as a whole, provided Your use,
  reproduction, and distribution of the Work otherwise complies with
  the conditions stated in this License.

  5. Submission of Contributions. Unless You explicitly state otherwise,
  any Contribution intentionally submitted for inclusion in the Work
  by You to the Licensor shall be under the terms and conditions of
  this License, without any additional terms or conditions.
  Notwithstanding the above, nothing herein shall supersede or modify
  the terms of any separate license agreement you may have executed
  with Licensor regarding such Contributions.

  6. Trademarks. This License does not grant permission to use the trade
  names, trademarks, service marks, or product names of the Licensor,
  except as required for reasonable and customary use in describing the
  origin of the Work and reproducing the content of the NOTICE file.

  7. Disclaimer of Warranty. Unless required by applicable law or
  agreed to in writing, Licensor provides the Work (and each
  Contributor provides its Contributions) on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
  implied, including, without limitation, any warranties or conditions
  of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A
  PARTICULAR PURPOSE. You are solely responsible for determining the
  appropriateness of using or redistributing the Work and assume any
  risks associated with Your exercise of permissions under this License.

  8. Limitation of Liability. In no event and under no legal theory,
  whether in tort (including negligence), contract, or otherwise,
  unless required by applicable law (such as deliberate and grossly
  negligent acts) or agreed to in writing, shall any Contributor be
  liable to You for damages, including any direct, indirect, special,
  incidental, or consequential damages of any character arising as a
  result of this License or out of the use or inability to use the
  Work (including but not limited to damages for loss of goodwill,
  work stoppage, computer failure or malfunction, or any and all
  other commercial damages or losses), even if such Contributor
  has been advised of the possibility of such damages.

  9. Accepting Warranty or Additional Liability. While redistributing
  the Work or Derivative Works thereof, You may choose to offer,
  and charge a fee for, acceptance of support, warranty, indemnity,
  or other liability obligations and/or rights consistent with this
  License. However, in accepting such obligations, You may act only
  on Your own behalf and on Your sole responsibility, not on behalf
  of any other Contributor, and only if You agree to indemnify,
  defend, and hold each Contributor harmless for any liability
  incurred by, or claims asserted against, such Contributor by reason
  of your accepting any such warranty or additional liability.

  END OF TERMS AND CONDITIONS

   ${fullProjectName}

  Copyright 2019 Fathym, Inc.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.`;
        if (!host.exists('LICENSE')) {
            host.create('LICENSE', license);
        }
        else {
            host.overwrite('LICENSE', license);
        }
    };
}
exports.addLicense = addLicense;
