// @lowcodeunit/lcu-documentation-common/esm5/lib/lcu-documentation.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { FathymSharedModule, MaterialModule } from '@lcu/common';
import { DocsService } from './services/docs.service';
import { LcuDocsComponent, markedOptionsFactory } from './controls/docs/docs.component';
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { MarkdownModule, MarkedOptions } from 'ngx-markdown';
var ɵ0 = markedOptionsFactory;
var LcuDocumentationModule = /** @class */ (function () {
    function LcuDocumentationModule() {
    }
    LcuDocumentationModule_1 = LcuDocumentationModule;
    LcuDocumentationModule.forRoot = function () {
        return {
            ngModule: LcuDocumentationModule_1,
            providers: [DocsService]
        };
    };
    var LcuDocumentationModule_1;
    LcuDocumentationModule = LcuDocumentationModule_1 = __decorate([
        NgModule({
            declarations: [LcuDocsComponent],
            imports: [
                FathymSharedModule,
                FormsModule,
                ReactiveFormsModule,
                FlexLayoutModule,
                MaterialModule,
                HttpClientModule,
                MarkdownModule.forRoot({
                    loader: HttpClient,
                    markedOptions: {
                        provide: MarkedOptions,
                        useFactory: ɵ0
                    }
                })
            ],
            exports: [LcuDocsComponent],
            entryComponents: []
        })
    ], LcuDocumentationModule);
    return LcuDocumentationModule;
}());
export { LcuDocumentationModule };
export { ɵ0 };
