// @lcu/common/esm5/lib/api/daf/daf.service.js
import { throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { LCUServiceSettings } from '../lcu-service-settings';
var DAFService = /** @class */ (function () {
    //  Constructors
    function DAFService(injector) {
        this.injector = injector;
        this.http = injector.get(HttpClient);
        try {
            this.settings = injector.get(LCUServiceSettings);
        }
        catch (err) { }
    }
    //  Helpers
    DAFService.prototype.catchError = function () {
        return catchError(this.handleError);
    };
    DAFService.prototype.delete = function (path) {
        return this.http.delete(this.resolvePath(path)).pipe(this.catchError());
    };
    DAFService.prototype.get = function (path) {
        return this.http.get(this.resolvePath(path)).pipe(this.catchError());
    };
    DAFService.prototype.handleError = function (error) {
        // In a real world app, we might use a remote logging infrastructure
        // We'd also dig deeper into the error to get a better message
        var errMsg = error.message ? error.message : error.status ? error.status + " - " + error.statusText : 'Server error';
        console.error(errMsg); // log to console instead
        return throwError(errMsg);
    };
    DAFService.prototype.patch = function (data, path) {
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.patch(this.resolvePath(path), data, options).pipe(this.catchError());
    };
    DAFService.prototype.post = function (data, path) {
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.post(this.resolvePath(path), data, options).pipe(this.catchError());
    };
    DAFService.prototype.put = function (data, path) {
        var options = {
            headers: new HttpHeaders({
                'Content-Type': 'application/json'
            })
        };
        return this.http.put(this.resolvePath(path), data, options).pipe(this.catchError());
    };
    DAFService.prototype.resolvePath = function (path) {
        if (this.settings && this.settings.APIRoot) {
            if (this.settings.APIRoot.endsWith('/')) {
                this.settings.APIRoot = this.settings.APIRoot.substring(this.settings.APIRoot.length - 1);
            }
            if (path.startsWith('/')) {
                path = path.substring(1);
            }
            return this.settings.APIRoot + "/" + path;
        }
        else {
            return path;
        }
    };
    return DAFService;
}());
export { DAFService };
