// @lcu/common/esm5/lib/directives/themes/theme-color.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, ElementRef, Renderer2 } from '@angular/core';
/**
 * Directive for setting theme color to a element, color comes from
 * a local stylesheet
 *
 * @use as an attribute: lcuColor="primary-A700", sets the class
 * .primary-A700 {} to the host element
 */
/**
 * SCSS class types
 */
// type ColorClasses = 'primary-500' | 'accent-500' | 'warn-500';
var ThemeColorDirective = /** @class */ (function () {
    function ThemeColorDirective(element, renderer) {
        this.element = element;
        this.renderer = renderer;
    }
    Object.defineProperty(ThemeColorDirective.prototype, "lcuThemeColor", {
        /**
         * Set the current theme class to the element
         */
        set: function (val) {
            // this.renderer.removeClass(this.element.nativeElement, `${val}`);
            this.renderer.addClass(this.element.nativeElement, "" + val);
            // this.readProperty('background-color');
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Get the value of a specific style property
     *
     * @param property style property
     */
    ThemeColorDirective.prototype.readProperty = function (property) {
        console.log('element', window.getComputedStyle(this.element.nativeElement).getPropertyValue(property));
        // console.log('element', window.getComputedStyle(this.element.nativeElement));
    };
    ThemeColorDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], ThemeColorDirective.prototype, "lcuThemeColor", null);
    ThemeColorDirective = __decorate([
        Directive({
            selector: '[lcuThemeColor]'
        }),
        __metadata("design:paramtypes", [ElementRef,
            Renderer2])
    ], ThemeColorDirective);
    return ThemeColorDirective;
}());
export { ThemeColorDirective };
