// @lcu/common/esm2015/lib/routing/router-helpers.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { ActivatedRoute, Router, NavigationCancel, NavigationEnd, NavigationError, NavigationStart } from '@angular/router';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
let RouterHelpersService = class RouterHelpersService {
    // 	Constructors
    constructor(route, router) {
        this.route = route;
        this.router = router;
    }
    // 	API Methods
    RunOnNavigation(action, navEventType = null) {
        return this.router.events.subscribe((event) => {
            if (navEventType == null || event instanceof navEventType) {
                action(event);
            }
        });
    }
    RunOnNavigationCancel(action) {
        return this.RunOnNavigation(action, NavigationCancel);
    }
    RunOnNavigationEnd(action) {
        return this.RunOnNavigation(action, NavigationEnd);
    }
    RunOnNavigationError(action) {
        return this.RunOnNavigation(action, NavigationError);
    }
    RunOnNavigationStart(action) {
        return this.RunOnNavigation(action, NavigationStart);
    }
    RunOnQueryParam(param, action, checkVal = null) {
        return this.route.queryParams.subscribe((params) => {
            if ((checkVal == null && params[param] != null) || (checkVal != null && params[param] === checkVal)) {
                action(params);
            }
        });
    }
    RunOnRouteParam(param, action, checkVal = null) {
        return this.route.params.subscribe((params) => {
            if ((checkVal == null && params[param] != null) || (checkVal != null && params[param] === checkVal)) {
                action(params);
            }
        });
    }
};
RouterHelpersService.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: Router }
];
RouterHelpersService.ɵprov = i0.ɵɵdefineInjectable({ factory: function RouterHelpersService_Factory() { return new RouterHelpersService(i0.ɵɵinject(i1.ActivatedRoute), i0.ɵɵinject(i1.Router)); }, token: RouterHelpersService, providedIn: "root" });
RouterHelpersService = __decorate([
    Injectable({
        providedIn: 'root'
    }),
    __metadata("design:paramtypes", [ActivatedRoute, Router])
], RouterHelpersService);
export { RouterHelpersService };
