// @lcu/common/esm2015/lib/hosting/fathym-shared.module.js
var FathymSharedModule_1;
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
import { LCUInterceptor } from '../api/daf/lcu.interceptor';
import { RealTimeService } from '../api/real-time/real-time.service';
import { LCUServiceSettings } from '../api/lcu-service-settings';
import { SafePipe } from '../pipes/safe-pipe';
export const winAny = window;
let FathymSharedModule = FathymSharedModule_1 = class FathymSharedModule {
    //  Fields
    //  API Methods
    static DefaultServiceSettings(env, apiRoot) {
        return {
            APIRoot: winAny.LCU && winAny.LCU.APIRoot ? winAny.LCU.APIRoot : '',
            AppConfig: {
                ID: winAny.LCU && winAny.LCU.Application && winAny.LCU.Application.ID ? winAny.LCU.Application.ID : 'test-app',
                EnterpriseAPIKey: winAny.LCU && winAny.LCU.Application.EnterpriseAPIKey ? winAny.LCU.Application.EnterpriseAPIKey : 'test-app'
            },
            StateConfig: {
                Environment: winAny.LCU.State ? winAny.LCU.State.Environment : '',
                ActionRoot: winAny.LCU.State ? winAny.LCU.State.Root : '',
                Root: winAny.LCU.State ? winAny.LCU.State.Root : '',
                UsernameMock: winAny.LCU.State ? winAny.LCU.State.UsernameMock : ''
            },
            Settings: winAny.LCU.Settings
        };
    }
    static forRoot() {
        return {
            ngModule: FathymSharedModule_1,
            providers: [
                RealTimeService,
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: LCUInterceptor,
                    multi: true,
                    deps: [LCUServiceSettings]
                }
            ]
        };
    }
};
FathymSharedModule = FathymSharedModule_1 = __decorate([
    NgModule({
        declarations: [SafePipe],
        imports: [CommonModule, FormsModule, HttpClientModule, RouterModule],
        exports: [CommonModule, FormsModule, HttpClientModule, RouterModule, SafePipe],
        providers: [],
        entryComponents: []
    })
], FathymSharedModule);
export { FathymSharedModule };
